<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Inference for High-dimensional Data: Principal Components Analysis</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/incubator/favicon-16x16.png"><link rel="manifest" href="favicons/incubator/site.webmanifest"><link rel="mask-icon" href="favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="assets/images/incubator-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/principal-component-analysis.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Inference for High-dimensional Data
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Inference for High-dimensional Data
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Inference for High-dimensional Data
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 76%" class="percentage">
    76%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 76%" aria-valuenow="76" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/principal-component-analysis.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="example-gene-expression-data.html">1. Example Gene Expression Datasets</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="inference-for-highthroughput.html">2. Basic inference for high-throughput data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="procedures-for-multiple-testing.html">3. Procedures for Multiple Comparisons</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="error-rates.html">4. Error Rates</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="bonferroni-correction.html">5. The Bonferroni Correction</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="false-discovery-rate.html">6. False Discovery Rate</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="direct-approach-false-discovery-rate.html">7. Direct Approach to FDR and q-values</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="basic-exploratory-data-analysis.html">8. Basic EDA for high-throughput data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        9. Principal Components Analysis
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#dimension-reduction-motivation">Dimension Reduction Motivation</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="statistical-models.html">10. Statistical Models</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="basic-exploratory-data-analysis.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="statistical-models.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="basic-exploratory-data-analysis.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Basic EDA for
        </a>
        <a class="chapter-link float-end" href="statistical-models.html" rel="next">
          Next: Statistical Models...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Principal Components Analysis</h1>
        <p>Last updated on 2025-08-12 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/principal-component-analysis.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can researchers simplify or streamline EDA in high-throughput
data sets?</li>
<li>What is principal component analysis (PCA) and when can it be
used?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain the purpose of dimension reduction.</li>
<li>Define a principal component.</li>
<li>Perform a principal components analysis.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="dimension-reduction-motivation">Dimension Reduction Motivation<a class="anchor" aria-label="anchor" href="#dimension-reduction-motivation"></a></h2>
<hr class="half-width"><p>Visualizing data is one of the most, if not the most, important step
in the analysis of high-throughput data. The right visualization method
may reveal problems with the experimental data that can render the
results from a standard analysis, although typically appropriate,
completely useless.</p>
<p>Recall that in recent history biologists went from using their eyes
or simple summaries to categorize results, to having thousands (and now
millions) of measurements per sample to analyze. Here we will focus on
statistical inference in the context of high-throughput measurements,
also called high-dimensional data. High-dimensional data is “wide” data,
rather than long data, which has few variables relative to the number of
observations. The number of features (variables) in high-dimensional
data is much greater than the number of observations, making simple
visualizations like scatterplots cumbersome.</p>
<p>We have shown methods for visualizing global properties of the
columns or rows, but plots that reveal relationships between columns or
between rows are more complicated due to the high dimensionality of
data. For example, to compare each of the 189 samples to each other, we
would have to create, for example, 17,766 MA-plots. Creating one single
scatterplot of the data is impossible since points are very high
dimensional.</p>
<p>We will describe powerful techniques for exploratory data analysis
based on <em>dimension reduction</em>. The general idea is to reduce the
dataset to have fewer dimensions, yet approximately preserve important
properties, such as the distance between samples. If we are able to
reduce down to, say, two dimensions, we can then easily make plots.</p>
<p>Principal component analysis (PCA) is a popular method of analyzing
high-dimensional data. Large datasets of correlated variables can be
summarized into smaller numbers of uncorrelated principal components
that explain most of the variability in the original dataset. An example
of PCA might be reducing several variables representing aspects of
patient health (blood pressure, heart rate, respiratory rate) into a
single feature.</p>
<p>PCA is a useful exploratory analysis tool. PCA allows us to reduce a
large number of variables into a few features which represent most of
the variation in the original variables. This makes exploration of the
original variables easier.</p>
<p>The first principal component (<span class="math inline">\(Z_1\)</span>) is calculated using the
equation:</p>
<p><span class="math display">\[
  Z_1 = a_{11}X_1 + a_{21}X_2 +....+a_{p1}X_p
\]</span></p>
<p><span class="math inline">\(X_1...X_p\)</span> represents variables
in the original dataset and <span class="math inline">\(a_{11}...a_p\)</span> represent principal
component loadings, which can be thought of as the degree to which each
variable contributes to the calculation of the principal component.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>Descriptions of three datasets and research questions are given
below. For which of these might PCA be considered a useful tool for
analyzing data so that the research questions may be addressed?</p>
<ol style="list-style-type: decimal"><li>An epidemiologist has data collected from different patients
admitted to hospital with infectious respiratory disease. They would
like to determine whether length of stay in hospital differs in patients
with different respiratory diseases.</li>
<li>An online retailer has collected data on user interactions with its
online app and has information on the number of times each user
interacted with the app, what products they viewed per interaction, and
the type and cost of these products. The retailer would like to use this
information to predict whether or not a user will be interested in a new
product.</li>
<li>A scientist has assayed gene expression levels in 1000 cancer
patients and has data from probes targeting different genes in tumour
samples from patients. She would like to create new variables
representing relative abundance of different groups of genes to i) find
out if genes form subgroups based on biological function and ii) use
these new variables in a linear regression examining how gene expression
varies with disease severity.</li>
<li>All of the above.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>In the first case, a regression model would be more suitable; perhaps
a survival model. In the second, again a regression model, likely linear
or logistic, would be more suitable. In the third example, PCA can help
to identify modules of correlated features that explain a large amount
of variation within the data.</p>
<p>Therefore the answer here is 3.</p>
</div>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="what-is-a-principal-component">What is a principal component?<a class="anchor" aria-label="anchor" href="#what-is-a-principal-component"></a></h1>
<p>The first principal component is the direction of the data along
which the observations vary the most. The second principal component is
the direction of the data along which the observations show the next
highest amount of variation. The second principal component is a linear
combination of the variables that is uncorrelated with the first
principal component. There are as many principal components as there are
variables in your dataset, but as we’ll see, some are more useful at
explaining your data than others. By definition, the first principal
component explains more variation than other principal components. Some
of the following content is adapted from <a href="https://carpentries-incubator.github.io/high-dimensional-stats-r/index.html" class="external-link">O’Callaghan
A, Robertson G, LLewellyn M, Becher H, Meynert A, Vallejos C, Ewing A.
(2024)</a>. <em>High dimensional statistics with R</em>. <a href="https://github.com/carpentries-incubator/high-dimensional-stats-r" class="external-link uri">https://github.com/carpentries-incubator/high-dimensional-stats-r</a>.</p>
<p>The animation below illustrates how principal components are
calculated from data. You can imagine that the black line is a rod and
each red dashed line is a spring. The energy of each spring is
proportional to its squared length. The direction of the first principal
component is the one that minimizes the total energy of all of the
springs. In the animation below, the springs pull the rod, finding the
direction of the first principal component when they reach equilibrium.
We then use the length of the springs from the rod as the first
principal component. This is explained in more detail on <a href="https://stats.stackexchange.com/questions/2691/making-sense-of-principal-component-analysis-eigenvectors-eigenvalues" class="external-link">this
Q&amp;A website</a>.</p>
<figure><img src="./fig/pendulum.gif" class="figure mx-auto d-block"></figure><div class="section level4">
<h4 id="example-reducing-two-dimensions-to-one">Example: Reducing two dimensions to one<a class="anchor" aria-label="anchor" href="#example-reducing-two-dimensions-to-one"></a></h4>
<p>We consider an example with twin heights. Here we simulate 100 two
dimensional points that represent the number of standard deviations each
individual is from the mean height. Each point is a pair of twins:</p>
<figure style="text-align: center"><img src="fig/principal-component-analysis-rendered-simulate_twin_heights-1.png" alt="Simulated twin pair heights." class="figure mx-auto d-block"><figcaption>
Simulated twin pair heights.
</figcaption></figure><p>To help with the illustration, think of this as high-throughput gene
expression data with the twin pairs representing the <span class="math inline">\(N\)</span> samples and the two heights
representing gene expression from two genes.</p>
<p>We are interested in the distance between any two samples. We can
compute this using <code>dist</code>. For example, here is the distance
between the two orange points in the figure above:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span><span class="op">=</span><span class="fu">dist</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">as.matrix</span><span class="op">(</span><span class="va">d</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.140897</code></pre>
</div>
<p>What if making two dimensional plots was too complex and we were only
able to make 1 dimensional plots. Can we, for example, reduce the data
to a one dimensional matrix that preserves distances between points?</p>
<p>If we look back at the plot, and visualize a line between any pair of
points, the length of this line is the distance between the two points.
These lines tend to go along the direction of the diagonal. We have seen
before that we can “rotate” the plot so that the diagonal is in the
x-axis by making a MA-plot instead:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z1</span> <span class="op">=</span> <span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span><span class="op">+</span><span class="va">y</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span> <span class="co">#the sum </span></span>
<span><span class="va">z2</span> <span class="op">=</span> <span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span><span class="op">-</span><span class="va">y</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span>   <span class="co">#the difference</span></span>
<span></span>
<span><span class="va">z</span> <span class="op">=</span> <span class="fu">rbind</span><span class="op">(</span> <span class="va">z1</span>, <span class="va">z2</span><span class="op">)</span> <span class="co">#matrix now same dimensions as y</span></span>
<span></span>
<span><span class="va">thelim</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>,<span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>,<span class="va">y</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>,xlab<span class="op">=</span><span class="st">"Twin 1 (standardized height)"</span>,</span>
<span>     ylab<span class="op">=</span><span class="st">"Twin 2 (standardized height)"</span>,</span>
<span>     xlim<span class="op">=</span><span class="va">thelim</span>,ylim<span class="op">=</span><span class="va">thelim</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,<span class="va">y</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,col<span class="op">=</span><span class="fl">2</span>,pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">z</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>,<span class="va">z</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>,xlim<span class="op">=</span><span class="va">thelim</span>,ylim<span class="op">=</span><span class="va">thelim</span>,xlab<span class="op">=</span><span class="st">"Average height"</span>,ylab<span class="op">=</span><span class="st">"Difference in height"</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="va">z</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,<span class="va">z</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,col<span class="op">=</span><span class="fl">2</span>,pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/principal-component-analysis-rendered-rotation-1.png" alt="Twin height scatterplot (left) and MA-plot (right)." class="figure mx-auto d-block"><figcaption>
Twin height scatterplot (left) and MA-plot (right).
</figcaption></figure><p>How do we perform a PCA?</p>
</div>
<div class="section level2">
<h2 id="a-prostate-cancer-dataset">A prostate cancer dataset<a class="anchor" aria-label="anchor" href="#a-prostate-cancer-dataset"></a></h2>
<p>The <code>Prostate</code> dataset represents data from 97 men who
have prostate cancer. The data come from a study which examined the
correlation between the level of prostate specific antigen and a number
of clinical measures in men who were about to receive a radical
prostatectomy. The data have 97 rows and 9 columns.</p>
<p>Columns include: 1. <code>lcavol</code> (log-transformed cancer
volume), 2. <code>lweight</code> (log-transformed prostate weight), 3.
<code>lbph</code> (log-transformed amount of benign prostate
enlargement), 4. <code>svi</code> (seminal vesicle invasion), 5.
<code>lcp</code> (log-transformed capsular penetration; amount of spread
of cancer in outer walls of prostate), 6. <code>gleason</code> (Gleason
score; grade of cancer cells), 7. <code>pgg45</code> (percentage Gleason
scores 4 or 5), 8. <code>lpsa</code> (log-tranformed prostate specific
antigen; level of PSA in blood). 9. <code>age</code> (patient age in
years).</p>
<p>Here we will calculate principal component scores for each of the
rows in this dataset, using five principal components (one for each
variable included in the PCA). We will include five clinical variables
in our PCA, each of the continuous variables in the prostate dataset, so
that we can create fewer variables representing clinical markers of
cancer progression. Standard PCAs are carried out using continuous
variables only.</p>
<p>First, we will examine the <code>Prostate</code> dataset in the
<code>Brq</code> package.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">Brq</span><span class="op">)</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="st">"Prostate"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">Prostate</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      lcavol  lweight age      lbph svi       lcp gleason pgg45       lpsa
1 -0.5798185 2.769459  50 -1.386294   0 -1.386294       6     0 -0.4307829
2 -0.9942523 3.319626  58 -1.386294   0 -1.386294       6     0 -0.1625189
3 -0.5108256 2.691243  74 -1.386294   0 -1.386294       7    20 -0.1625189
4 -1.2039728 3.282789  58 -1.386294   0 -1.386294       6     0 -0.1625189
5  0.7514161 3.432373  62 -1.386294   0 -1.386294       6     0  0.3715636
6 -1.0498221 3.228826  50 -1.386294   0 -1.386294       6     0  0.7654678</code></pre>
</div>
<p>Note that each row of the dataset represents a single patient.</p>
<p>We will create a subset of the data including only the clinical
variables we want to use in the PCA.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pros2</span> <span class="op">&lt;-</span> <span class="va">Prostate</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"lcavol"</span>, <span class="st">"lweight"</span>, <span class="st">"lbph"</span>, <span class="st">"lcp"</span>, <span class="st">"lpsa"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">pros2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      lcavol  lweight      lbph       lcp       lpsa
1 -0.5798185 2.769459 -1.386294 -1.386294 -0.4307829
2 -0.9942523 3.319626 -1.386294 -1.386294 -0.1625189
3 -0.5108256 2.691243 -1.386294 -1.386294 -0.1625189
4 -1.2039728 3.282789 -1.386294 -1.386294 -0.1625189
5  0.7514161 3.432373 -1.386294 -1.386294  0.3715636
6 -1.0498221 3.228826 -1.386294 -1.386294  0.7654678</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="do-we-need-to-standardise-the-data">Do we need to standardise the data?<a class="anchor" aria-label="anchor" href="#do-we-need-to-standardise-the-data"></a></h2>
<p>Now we compare the variances between variables in the dataset.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">apply</span><span class="op">(</span><span class="va">pros2</span>, <span class="fl">2</span>, <span class="va">var</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  lcavol  lweight     lbph      lcp     lpsa
1.389157 0.246642 2.104840 1.955102 1.332476 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">pros2</span><span class="op">$</span><span class="va">lweight</span>, breaks <span class="op">=</span> <span class="st">"FD"</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">pros2</span><span class="op">$</span><span class="va">lbph</span>, breaks <span class="op">=</span> <span class="st">"FD"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/principal-component-analysis-rendered-var-hist-1.png" alt="Histograms comparing variances between log-transformed prostate weight and benign enlargement." class="figure mx-auto d-block"><figcaption>
Histograms comparing variances between log-transformed prostate weight
and benign enlargement.
</figcaption></figure><p>Note that variance is greatest for <code>lbph</code> and lowest for
<code>lweight</code>. It is clear from this output that we need to scale
each of these variables before including them in a PCA analysis to
ensure that differences in variances between variables do not drive the
calculation of principal components. In this example we standardise all
five variables to have a mean of 0 and a standard deviation of 1.</p>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>Why might it be necessary to standardise variables before performing
a PCA?<br>
Can you think of datasets where it might not be necessary to standardise
variables? Discuss.</p>
<ol style="list-style-type: decimal"><li>To make the results of the PCA interesting.</li>
<li>If you want to ensure that variables with different ranges of values
contribute equally to analysis.</li>
<li>To allow the feature matrix to be calculated faster, especially in
cases where there are a lot of input variables.</li>
<li>To allow both continuous and categorical variables to be included in
the PCA.</li>
<li>All of the above.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<ol start="2" style="list-style-type: decimal"><li>Scaling the data isn’t guaranteed to make the results more
interesting. It also won’t affect how quickly the output will be
calculated, whether continuous and categorical variables are present or
not.</li>
</ol><p>It is done to ensure that all features have equal weighting in the
resulting PCs.</p>
<p>You may not want to standardise datasets which contain continuous
variables all measured on the same scale (e.g. gene expression data or
RNA sequencing data). In this case, variables with very little
sample-to-sample variability may represent only random noise, and
standardising the data would give these extra weight in the PCA.</p>
</div>
</div>
</div>
</div>
<p>Next we will carry out a PCA using the <code>prcomp()</code> function
in base R. The input data (<code>pros2</code>) is in the form of a
matrix. Note that the <code>scale = TRUE</code> argument is used to
standardise the variables to have a mean 0 and standard deviation of
1.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca.pros</span> <span class="op">&lt;-</span> <span class="fu">prcomp</span><span class="op">(</span><span class="va">pros2</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, center <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pca.pros</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Standard deviations (1, .., p=5):
[1] 1.5648756 1.1684678 0.7452990 0.6362941 0.4748755

Rotation (n x k) = (5 x 5):
              PC1         PC2         PC3         PC4         PC5
lcavol  0.5616465 -0.23664270  0.01486043  0.22708502 -0.75945046
lweight 0.2985223  0.60174151 -0.66320198 -0.32126853 -0.07577123
lbph    0.1681278  0.69638466  0.69313753  0.04517286 -0.06558369
lcp     0.4962203 -0.31092357  0.26309227 -0.72394666  0.25253840
lpsa    0.5665123 -0.01680231 -0.10141557  0.56487128  0.59111493</code></pre>
</div>
</div>
</div>
<div class="section level1">
<h1 id="how-many-principal-components-do-we-need">How many principal components do we need?<a class="anchor" aria-label="anchor" href="#how-many-principal-components-do-we-need"></a></h1>
<p>We have calculated one principal component for each variable in the
original dataset. How do we choose how many of these are necessary to
represent the true variation in the data, without having extra
components that are unnecessary?</p>
<p>Let’s look at the relative importance of each component using
<code>summary</code>.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Importance of components:
                          PC1    PC2    PC3     PC4    PC5
Standard deviation     1.5649 1.1685 0.7453 0.63629 0.4749
Proportion of Variance 0.4898 0.2731 0.1111 0.08097 0.0451
Cumulative Proportion  0.4898 0.7628 0.8739 0.95490 1.0000</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get proportions of variance explained by each PC (rounded to 2 DP)</span></span>
<span><span class="va">prop.var</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="fu">summary</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">)</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span><span class="st">"Proportion of Variance"</span>, <span class="op">]</span>, <span class="fl">2</span><span class="op">)</span> <span class="op">*</span></span>
<span>  <span class="fl">100</span></span></code></pre>
</div>
<p>This returns the proportion of variance in the data explained by each
of the (p = 5) principal components. In this example, PC1 explains
approximately 49% of variance in the data, PC2 27% of variance, PC3 a
further 11%, PC4 approximately 8% and PC5 around 5%.</p>
<p>We can use a screeplot to see how much variation in the data is
explained by each principal component. Let’s calculate the screeplot for
our PCA.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate variance explained</span></span>
<span><span class="va">varExp</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">pca.pros</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">/</span> <span class="fu">sum</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="co"># calculate percentage variance explained using output from the PCA</span></span>
<span><span class="va">varDF</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>Dimensions <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span><span class="op">(</span><span class="va">varExp</span><span class="op">)</span>, varExp <span class="op">=</span> <span class="va">varExp</span><span class="op">)</span></span>
<span><span class="co"># create new dataframe with five rows, one for each principal component</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">varDF</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/principal-component-analysis-rendered-vardf-plot-1.png" alt="Screeplot showing proportion of variance explained by each principal component." class="figure mx-auto d-block"><figcaption>
Screeplot showing proportion of variance explained by each principal
component.
</figcaption></figure><p>The screeplot shows that the first principal component explains most
of the variance in the data (&gt;50%) and each subsequent principal
component explains less and less of the total variance. The first two
principal components explain &gt;70% of variance in the data. But what
do these two principal components mean?</p>
<div class="section level2">
<h2 id="what-are-loadings-and-principal-component-scores">What are loadings and principal component scores?<a class="anchor" aria-label="anchor" href="#what-are-loadings-and-principal-component-scores"></a></h2>
<p>The output from a PCA returns a matrix of principal component
loadings in which the columns show the principal component loading
vectors. Note that the square of values in each column sums to 1 as each
loading is scaled so as to prevent a blow up in variance. Larger values
in the columns suggest a greater contribution of that variable to the
principal component.</p>
<p>We can examine the output of our PCA by writing the following in
<code>R</code>:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca.pros</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Standard deviations (1, .., p=5):
[1] 1.5648756 1.1684678 0.7452990 0.6362941 0.4748755

Rotation (n x k) = (5 x 5):
              PC1         PC2         PC3         PC4         PC5
lcavol  0.5616465 -0.23664270  0.01486043  0.22708502 -0.75945046
lweight 0.2985223  0.60174151 -0.66320198 -0.32126853 -0.07577123
lbph    0.1681278  0.69638466  0.69313753  0.04517286 -0.06558369
lcp     0.4962203 -0.31092357  0.26309227 -0.72394666  0.25253840
lpsa    0.5665123 -0.01680231 -0.10141557  0.56487128  0.59111493</code></pre>
</div>
<p>For each row in the original dataset PCA returns a principal
component score for each of the principal components (PC1 to PC5 in the
<code>Prostate</code> data example). We can see how the principal
component score (<span class="math inline">\(Z_{i1}\)</span> for rows
<span class="math inline">\(i\)</span> to <span class="math inline">\(n\)</span>) is calculated for the first principal
component using the following equation from Figure 1:</p>
<p><span class="math display">\[
  Z_{i1} = a_1 \times (lcavol_i - \overline{lcavol}) + a_2 \times
(lpsa_i - \overline{lpsa})
\]</span></p>
<p><span class="math inline">\(a_1\)</span> and <span class="math inline">\(a_2\)</span> represent principal component
loadings in this equation. A loading can be thought of as the ‘weight’
each variable has on the calculation of the principal component. Note
that in our example using the <code>Prostate</code> dataset
<code>lcavol</code> (log-transformed cancer volume) and
<code>lpsa</code> (log-tranformed prostate specific antigen in blood)
are the variables that contribute most to the first principal
component.</p>
<p>We can better understand what the principal components represent in
terms of the original variables by plotting the first two principal
components against each other and labelling points by patient number.
Clusters of points which have similar principal component scores can be
observed using a biplot and the strength and direction of influence
different variables have on the calculation of the principal component
scores can be observed by plotting arrows representing the loadings onto
the graph. A biplot of the first two principal components can be created
as follows:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">biplot</span><span class="op">(</span><span class="va">pca.pros</span>, xlim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">0.3</span>, <span class="fl">0.3</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/principal-component-analysis-rendered-stats-biplot-1.png" alt="Biplot of first two principal components." class="figure mx-auto d-block"><figcaption>
Biplot of first two principal components.
</figcaption></figure><figure><img src="./fig/stats-biplot.png" alt="Biplot of first two principal components." class="figure mx-auto d-block"><div class="figcaption">Biplot of first two principal components.</div>
</figure><p>This biplot shows the position of each patient on a 2-dimensional
plot where loadings can be observed via the red arrows associated with
each of the variables. The variables <code>lpsa</code>,
<code>lcavol</code> and <code>lcp</code> are associated with positive
values on PC1 while positive values on PC2 are associated with the
variables <code>lbph</code> and <code>lweight</code>. The length of the
arrows indicates how much each variable contributes to the calculation
of each principal component.</p>
<p>The left and bottom axes show normalised principal component scores.
The axes on the top and right of the plot are used to interpret the
loadings, where loadings are scaled by the standard deviation of the
principal components (<code>pca.pros$sdev</code>) times the square root
the number of observations.</p>
</div>
</div>
<div class="section level1">
<h1 id="using-pca-to-analyse-gene-expression-data">Using PCA to analyse gene expression data<a class="anchor" aria-label="anchor" href="#using-pca-to-analyse-gene-expression-data"></a></h1>
<p>In this section you will carry out your own PCA using the
Bioconductor package <strong><code>PCAtools</code></strong> applied to
gene expression data to explore the topics covered above.
<strong><code>PCAtools</code></strong> provides functions that can be
used to explore data via PCA and produce useful figures and analysis
tools.</p>
<div class="section level2">
<h2 id="a-gene-expression-dataset-of-cancer-patients">A gene expression dataset of cancer patients<a class="anchor" aria-label="anchor" href="#a-gene-expression-dataset-of-cancer-patients"></a></h2>
<p>The dataset we will be analysing in this lesson includes two subsets
of data: 1. a matrix of gene expression data showing microarray results
for different probes used to examine gene expression profiles in 91
different breast cancer patient samples. 2. metadata associated with the
gene expression results detailing information from patients from whom
samples were taken.</p>
<p>Let’s load the <strong><code>PCAtools</code></strong> package and the
data.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"PCAtools"</span><span class="op">)</span></span></code></pre>
</div>
<p>We will first load the microarray breast cancer gene expression data
and associated metadata, downloaded from the <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE2990" class="external-link">Gene
Expression Omnibus</a>.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"SummarizedExperiment"</span><span class="op">)</span></span>
<span><span class="va">cancer</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"./data/cancer_expression.rds"</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">cancer</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">cancer</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">View</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="co">#nrow=22215 probes</span></span>
<span><span class="co">#ncol=91 samples</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">View</span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#nrow=91</span></span>
<span><span class="co">#ncol=8</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">all</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">==</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Check that column names and row names match</span></span>
<span><span class="co">#If they do should return TRUE</span></span></code></pre>
</div>
<p>The ‘mat’ variable contains a matrix of gene expression profiles for
each sample. Rows represent gene expression measurements and columns
represent samples. The ‘metadata’ variable contains the metadata
associated with the gene expression data including the name of the study
from which data originate, the age of the patient from which the sample
was taken, whether or not an oestrogen receptor was involved in their
cancer and the grade and size of the cancer for each sample (represented
by rows).</p>
<p>Microarray data are difficult to analyse for several reasons.
Firstly, they are typically high-dimensional and therefore are subject
to the same difficulties associated with analysing high dimensional data
outlined above (i.e. <em>p</em>&gt;<em>n</em>, large numbers of rows,
multiple possible response variables, curse of dimensionality).
Secondly, formulating a research question using microarray data can be
difficult, especially if not much is known a priori about which genes
code for particular phenotypes of interest. Finally, exploratory
analysis, which can be used to help formulate research questions and
display relationships, is difficult using microarray data due to the
number of potentially interesting response variables (i.e. expression
data from probes targeting different genes).</p>
<p>If researchers hypothesise that groups of genes (e.g. biological
pathways) may be associated with different phenotypic characteristics of
cancers (e.g. histologic grade, tumour size), using statistical methods
that reduce the number of columns in the microarray matrix to a smaller
number of dimensions representing groups of genes would help visualise
the data and address research questions regarding the effect different
groups of genes have on disease progression.</p>
<p>Using the <strong><code>PCAtools</code></strong> we will apply a PCA
to the cancer gene expression data, plot the amount of variation in the
data explained by each principal component and plot the most important
principal components against each other as well as understanding what
each principal component represents.</p>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>Apply a PCA to the cancer gene expression data using the
<code>pca()</code> function from <strong><code>PCAtools</code></strong>.
You can use the help files in PCAtools to find out about the
<code>pca()</code> function (type <code>help("pca")</code> or
<code>?pca</code> in R).</p>
<p>Remove the lower 20% of principal components from your PCA using the
<code>removeVar</code> argument in the <code>pca()</code> function. As
in the example using prostate data above, examine the first 5 rows and
columns of rotated data and loadings from your PCA.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<pre><code><span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu">pca</span><span class="op">(</span><span class="va">mat</span>, metadata <span class="op">=</span> <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#Many PCs explain a very small amount of the total variance in the data</span></span>
<span><span class="co">#Remove the lower 20% of PCs with lower variance</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu">pca</span><span class="op">(</span><span class="va">mat</span>, metadata <span class="op">=</span> <span class="va">metadata</span>, removeVar <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="co">#Explore other arguments provided in pca</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">rotated</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">which.max</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span><span class="fl">49</span>, <span class="op">]</span></span>
<span></span>
<span><span class="fu">which.max</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span><span class="fl">27</span>, <span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>The function <code>pca()</code> is used to perform PCA, and uses as
inputs a matrix (<code>mat</code>) containing continuous numerical data
in which rows are data variables and columns are samples, and
<code>metadata</code> associated with the matrix in which rows represent
samples and columns represent data variables. It has options to centre
or scale the input data before a PCA is performed, although in this case
gene expression data do not need to be transformed prior to PCA being
carried out as variables are measured on a similar scale (values are
comparable between rows). The output of the <code>pca()</code> function
includes a lot of information such as loading values for each variable
(<code>loadings</code>), principal component scores
(<code>rotated</code>) and the amount of variance in the data explained
by each principal component.</p>
<p>Rotated data shows principal component scores for each sample and
each principal component. Loadings the contribution each variable makes
to each principal component.</p>
</div>
<div class="section level2">
<h2 id="scaling-variables-for-pca">Scaling variables for PCA<a class="anchor" aria-label="anchor" href="#scaling-variables-for-pca"></a></h2>
<p>When running <code>pca()</code> above, we kept the default setting,
<code>scale=FALSE</code>. That means genes with higher variation in
their expression levels should have higher loadings, which is what we
are interested in. Whether or not to scale variables for PCA will depend
on your data and research question.</p>
<p>Note that this is different from normalising gene expression data.
Gene expression data have to be normalised before donwstream analyses
can be carried out. This is to reduce to effect technical and other
potentially confounding factors. We assume that the expression data we
use had been normalised previously.</p>
</div>
<div class="section level2">
<h2 id="choosing-how-many-components-are-important-to-explain-the-variance-in-the-data">Choosing how many components are important to explain the variance
in the data<a class="anchor" aria-label="anchor" href="#choosing-how-many-components-are-important-to-explain-the-variance-in-the-data"></a></h2>
<p>As in the example using the <code>Prostate</code> dataset we can use
a screeplot to compare the proportion of variance in the data explained
by each principal component. This allows us to understand how much
information in the microarray dataset is lost by projecting the
observations onto the first few principal components and whether these
principal components represent a reasonable amount of the variation. The
proportion of variance explained should sum to one.</p>
<p>There are no clear guidelines on how many principal components should
be included in PCA: your choice depends on the total variability of the
data and the size of the dataset. We often look at the ‘elbow’ on the
screeplot as an indicator that the addition of principal components does
not drastically contribute to explain the remaining variance or choose
an arbitory cut off for proportion of variance explained.</p>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>Using the <code>screeplot()</code> function in
<strong><code>PCAtools</code></strong>, create a screeplot to show
proportion of variance explained by each principal component. Explain
the output of the screeplot in terms of proportion of variance in data
explained by each principal component.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<pre><code><span><span class="fu">screeplot</span><span class="op">(</span><span class="va">pc</span>, axisLabSize <span class="op">=</span> <span class="fl">5</span>, titleLabSize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span></code></pre>
<p>Note that first principal component (PC1) explains more variation
than other principal components (which is always the case in PCA). The
screeplot shows that the first principal component only explains ~33% of
the total variation in the micrarray data and many principal components
explain very little variation. The red line shows the cumulative
percentage of explained variation with increasing principal components.
Note that in this case 18 principal components are needed to explain
over 75% of variation in the data. This is not an unusual result for
complex biological datasets including genetic information as clear
relationships between groups are sometimes difficult to observe in the
data. The screeplot shows that using a PCA we have reduced 91 predictors
to 18 in order to explain a significant amount of variation in the data.
See additional arguments in screeplot function for improving the
appearance of the plot.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="investigating-the-principal-components">Investigating the principal components<a class="anchor" aria-label="anchor" href="#investigating-the-principal-components"></a></h2>
<p>Once the most important principal components have been identified
using <code>screeplot()</code>, these can be explored in more detail by
plotting principal components against each other and highlighting points
based on variables in the metadata. This will allow any potential
clustering of points according to demographic or phenotypic variables to
be seen.</p>
<p>We can use biplots to look for patterns in the output from the PCA.
Note that there are two functions called <code>biplot()</code>, one in
the package <strong><code>PCAtools</code></strong> and one in
<strong><code>stats</code></strong>. Both functions produce biplots but
their scales are different!</p>
<div id="exercise-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-5" class="callout-inner">
<h3 class="callout-title">Exercise 5</h3>
<div class="callout-content">
<p>Create a biplot of the first two principal components from your PCA
(using <code>biplot()</code> function in
<strong><code>PCAtools</code></strong> - see
<code>help("PCAtools::biplot")</code> for arguments) and examine whether
the data appear to form clusters. Explain your results.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<pre><code><span><span class="fu">biplot</span><span class="op">(</span><span class="va">pc</span>, lab <span class="op">=</span> <span class="cn">NULL</span>, colby <span class="op">=</span> <span class="st">'Grade'</span>, legendPosition <span class="op">=</span> <span class="st">'top'</span><span class="op">)</span></span></code></pre>
<p>The biplot shows the position of patient samples relative to PC1 and
PC2 in a 2-dimensional plot. Note that two groups are apparent along the
PC1 axis according to expressions of different genes while no separation
can be seem along the PC2 axis. Labels of patient samples are
automatically added in the biplot. Labels for each sample are added by
default, but can be removed if there is too much overlap in names. Note
that <strong><code>PCAtools</code></strong> does not scale biplot in the
same way as biplot using the stats package.</p>
</div>
</div>
</div>
</div>
<p>Let’s consider this biplot in more detail, and also display the
loadings:</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">biplot</span><span class="op">(</span><span class="va">pc</span>, lab <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">metadata</span><span class="op">)</span>, pointSize <span class="op">=</span> <span class="fl">1</span>, labSize <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pcaobj$rotated: object of type 'closure' is not subsettable</code></pre>
</div>
<figure><img src="./fig/pca-biplot.png" alt="Biplot of principal components one and two showing two groups for PC1 according to gene expression." class="figure mx-auto d-block"><div class="figcaption">Biplot of principal components one and two
showing two groups for PC1 according to gene expression.</div>
</figure><p>Sizes of labels, points and axes can be changed using arguments in
<code>biplot</code> (see <code>help("biplot")</code>). We can see from
the biplot that there appear to be two separate groups of points that
separate on the PC1 axis, but that no other grouping is apparent on
other PC axes.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotloadings</span><span class="op">(</span><span class="va">pc</span>, labSize <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pcaobj$loadings: object of type 'closure' is not subsettable</code></pre>
</div>
<figure><img src="./fig/pca-loadings.png" alt="Plot of principal component loadings show magnitude and direction of gene probes." class="figure mx-auto d-block"><div class="figcaption">Plot of principal component loadings show
magnitude and direction of gene probes.</div>
</figure><p>Plotting the loadings shows the magnitude and direction of loadings
for probes detecting genes on each principal component.</p>
<div id="exercise-6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-6" class="callout-inner">
<h3 class="callout-title">Exercise 6</h3>
<div class="callout-content">
<p>Use <code>colby</code> and <code>lab</code> arguments in
<code>biplot()</code> to explore whether these two groups may cluster by
patient age or by whether or not the sample expresses the oestrogen
receptor gene (ER+ or ER-).</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<pre><code><span>  <span class="fu">biplot</span><span class="op">(</span><span class="va">pc</span>,</span>
<span>    lab <span class="op">=</span> <span class="fu">paste0</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">metadata</span><span class="op">$</span><span class="va">Age</span>,<span class="st">'years'</span><span class="op">)</span>,</span>
<span>    colby <span class="op">=</span> <span class="st">'ER'</span>,</span>
<span>    hline <span class="op">=</span> <span class="fl">0</span>, vline <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    legendPosition <span class="op">=</span> <span class="st">'right'</span><span class="op">)</span></span></code></pre>
<p>{: .language-r} It appears that one cluster has more ER+ samples than
the other group.</p>
</div>
</div>
</div>
</div>
<p>So far we have only looked at a biplot of PC1 versus PC2 which only
gives part of the picture. The <code>pairplots()</code> function in
<strong><code>PCAtools</code></strong> can be used to create multiple
biplots including different principal components.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pairsplot</span><span class="op">(</span><span class="va">pc</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pcaobj$components: object of type 'closure' is not subsettable</code></pre>
</div>
<figure><img src="./fig/pairsplot.png" alt="Pairsplot of principal components show clusters on PC1 only." class="figure mx-auto d-block"><div class="figcaption">Pairsplot of principal components show clusters
on PC1 only.</div>
</figure><p>The plots show two apparent clusters involving the first principal
component only. No other clusters are found involving other principal
components.</p>
</div>
</div>
<div class="section level1">
<h1 id="using-pca-output-in-further-analysis">Using PCA output in further analysis<a class="anchor" aria-label="anchor" href="#using-pca-output-in-further-analysis"></a></h1>
<p>The output of PCA can be used to interpret data or can be used in
further analyses. For example, the PCA outputs new variables (principal
components) which represent several variables in the original dataset.
These new variables are useful for further exploring data, for example,
comparing principal component scores between groups or including the new
variables in linear regressions. Because the principal components are
uncorrelated (and independent) they can be included together in a single
linear regression.</p>
<div class="section level2">
<h2 id="principal-component-regression">Principal component regression<a class="anchor" aria-label="anchor" href="#principal-component-regression"></a></h2>
<p>PCA is often used to reduce large numbers of correlated variables
into fewer uncorrelated variables that can then be included in linear
regression or other models. This technique is called principal component
regression (PCR) and it allows researchers to examine the effect of
several correlated explanatory variables on a single response variable
in cases where a high degree of correlation initially prevents them from
being included in the same model. This is called principal componenet
regression (PCR) and is just one example of how principal components can
be used in further analysis of data. When carrying out PCR, the variable
of interest (response/dependent variable) is regressed against the
principal components calculated using PCA, rather than against each
individual explanatory variable from the original dataset. As there as
many principal components created from PCA as there are variables in the
dataset, we must select which principal components to include in PCR.
This can be done by examining the amount of variation in the data
explained by each principal component (see above).</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>Visualizing data with thousands or tens of thousands of measurements
is impossible using standard techniques.</li>
<li>Dimension reduction techniques coupled with visualization can reveal
relationships between dimensions (rows or columns) in the data.</li>
<li>Principal components analysis is a dimension reduction technique
that can reduce and summarize large datasets.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="basic-exploratory-data-analysis.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="statistical-models.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="basic-exploratory-data-analysis.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Basic EDA for
        </a>
        <a class="chapter-link float-end" href="statistical-models.html" rel="next">
          Next: Statistical Models...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/principal-component-analysis.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/smcclatchy/inference-high-dimensional-data/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/smcclatchy/inference-high-dimensional-data/" class="external-link">Source</a></p>
				<p><a href="https://github.com/smcclatchy/inference-high-dimensional-data/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:susan.mcclatchy@jax.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://smcclatchy.github.io/inference-high-dimensional-data/principal-component-analysis.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries, inference, high-dimensional data, biomedical science",
  "name": "Principal Components Analysis",
  "creativeWorkStatus": "active",
  "url": "https://smcclatchy.github.io/inference-high-dimensional-data/principal-component-analysis.html",
  "identifier": "https://smcclatchy.github.io/inference-high-dimensional-data/principal-component-analysis.html",
  "dateCreated": "2025-07-14",
  "dateModified": "2025-08-12",
  "datePublished": "2025-08-12"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

