<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Inference for High-dimensional Data: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png">
<link rel="manifest" href="../favicons/incubator/site.webmanifest">
<link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-warning">
          <abbr title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="border-radius: 5px"></i>
              Alpha
            </a>
            <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Inference for High-dimensional Data
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Inference for High-dimensional Data
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Inference for High-dimensional Data
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Using RMarkdown</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="example-gene-expression-data.html">2. Example Gene Expression Datasets</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="inference-for-highthroughput.html">3. Basic inference for high-throughput data</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="procedures-for-multiple-testing.html">4. Procedures for Multiple Comparisons</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="error-rates.html">5. Error Rates</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="bonferroni-correction.html">6. The Bonferroni Correction</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="false-discovery-rate.html">7. False Discovery Rate</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="direct-approach-false-discovery-rate.html">8. Direct Approach to FDR and q-values</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="basic-exploratory-data-analysis.html">9. Basic EDA for high-throughput data</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="principal-component-analysis.html">10. Principal Components Analysis</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="statistical-models.html">11. Statistical Models</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-introduction"><p>Content from <a href="introduction.html">Using RMarkdown</a></p>
<hr>
<p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/introduction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 12 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you write a lesson using R Markdown and
<a href="https://carpentries.github.io/sandpaper/" class="external-link">sandpaper</a>?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain how to use markdown with the new lesson template</li>
<li>Demonstrate how to include pieces of code, figures, and nested
challenge blocks</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>This is a lesson created via The Carpentries Workbench. It is written
in <a href="https://pandoc.org/MANUAL.txt" class="external-link">Pandoc-flavored Markdown</a>
for static files and <a href="https://rmarkdown.rstudio.com/" class="external-link">R
Markdown</a> for dynamic files that can render code into output. Please
refer to the <a href="https://carpentries.github.io/sandpaper-docs/" class="external-link">Introduction to The
Carpentries Workbench</a> for full documentation.</p>
<p>What you need to know is that there are three sections required for a
valid Carpentries lesson template:</p>
<ol style="list-style-type: decimal">
<li>
<code>questions</code> are displayed at the beginning of the episode
to prime the learner for the content.</li>
<li>
<code>objectives</code> are the learning objectives for an episode
displayed with the questions.</li>
<li>
<code>keypoints</code> are displayed at the end of the episode to
reinforce the objectives.</li>
</ol>
<div id="accordionInstructor1" class="accordion instructor-note accordion-flush">
<div class="accordion-item">
<button class="accordion-button instructor-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseInstructor1" aria-expanded="false" aria-controls="collapseInstructor1">
  <h3 class="accordion-header" id="headingInstructor1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="edit-2"></i></div> Instructor Note </h3>
</button>
<div id="collapseInstructor1" class="accordion-collapse collapse" data-bs-parent="#accordionInstructor1" aria-labelledby="headingInstructor1">
<div class="accordion-body">
<p>Inline instructor notes can help inform instructors of timing
challenges associated with the lessons. They appear in the “Instructor
View”</p>
</div>
</div>
</div>
</div>
<div id="challenge-1-can-you-do-it" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-1-can-you-do-it" class="callout-inner">
<h3 class="callout-title">Challenge 1: Can you do it?</h3>
<div class="callout-content">
<p>What is the output of this command?</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">paste</span><span class="op">(</span><span class="st">"This"</span>, <span class="st">"new"</span>, <span class="st">"lesson"</span>, <span class="st">"looks"</span>, <span class="st">"good"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Output </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "This new lesson looks good"</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-2-how-do-you-nest-solutions-within-challenge-blocks" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-2-how-do-you-nest-solutions-within-challenge-blocks" class="callout-inner">
<h3 class="callout-title">Challenge 2: how do you nest solutions within
challenge blocks?</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>You can add a line with at least three colons and a
<code>solution</code> tag.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="figures">Figures<a class="anchor" aria-label="anchor" href="#figures"></a>
</h2>
<hr class="half-width">
<p>You can also include figures generated from R Markdown:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pie</span><span class="op">(</span></span>
<span>  <span class="fu">c</span><span class="op">(</span>Sky <span class="op">=</span> <span class="fl">78</span>, <span class="st">"Sunny side of pyramid"</span> <span class="op">=</span> <span class="fl">17</span>, <span class="st">"Shady side of pyramid"</span> <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, </span>
<span>  init.angle <span class="op">=</span> <span class="fl">315</span>, </span>
<span>  col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"deepskyblue"</span>, <span class="st">"yellow"</span>, <span class="st">"yellow3"</span><span class="op">)</span>, </span>
<span>  border <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/introduction-rendered-pyramid-1.png" alt="pie chart illusion of a pyramid" class="figure mx-auto d-block"><figcaption>
Sun arise each and every morning
</figcaption></figure><p>Or you can use standard markdown for static figures with the
following syntax:</p>
<p><code>![optional caption that appears below the figure](figure url){alt='alt text for accessibility purposes'}</code></p>
<figure><img src="https://raw.githubusercontent.com/carpentries/logo/master/Badge_Carpentries.svg" alt="Blue Carpentries hex person logo with no text." class="figure mx-auto d-block"><div class="figcaption">You belong in The Carpentries!</div>
</figure><div id="callout1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div class="callout-inner">
<div class="callout-content">
<p>Callout sections can highlight information.</p>
<p>They are sometimes used to emphasise particularly important points
but are also used in some lessons to present “asides”: content that is
not central to the narrative of the lesson, e.g. by providing the answer
to a commonly-asked question.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="math">Math<a class="anchor" aria-label="anchor" href="#math"></a>
</h2>
<hr class="half-width">
<p>One of our episodes contains <span class="math inline">\(\LaTeX\)</span> equations when describing how to
create dynamic reports with {knitr}, so we now use mathjax to describe
this:</p>
<p><code>$\alpha = \dfrac{1}{(1 - \beta)^2}$</code> becomes: <span class="math inline">\(\alpha = \dfrac{1}{(1 - \beta)^2}\)</span></p>
<p>Cool, right?</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Use <code>.md</code> files for episodes when you want static
content</li>
<li>Use <code>.Rmd</code> files for episodes when you need to generate
output</li>
<li>Run <code>sandpaper::check_lesson()</code> to identify any issues
with your lesson</li>
<li>Run <code>sandpaper::build_lesson()</code> to preview your lesson
locally</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-example-gene-expression-data"><p>Content from <a href="example-gene-expression-data.html">Example Gene Expression Datasets</a></p>
<hr>
<p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/example-gene-expression-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 35 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What data will be be using for our analyses?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explore a high-throughput dataset composed of three tables.</li>
<li>Examine the features (high-throughput measurements) of the data you
explored.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="explore-a-gene-expression-dataset">Explore a Gene Expression Dataset<a class="anchor" aria-label="anchor" href="#explore-a-gene-expression-dataset"></a>
</h1>
<p>Since there is a vast number of available public datasets, we use
several gene expression examples. Nonetheless, the statistical
techniques you will learn have also proven useful in other fields that
make use of high-throughput technologies. Technologies such as
microarrays, next generation sequencing, fMRI, and mass spectrometry all
produce data to answer questions for which what we learn here will be
indispensable.</p>
<div class="section level4">
<h4 id="the-three-tables">The three tables<a class="anchor" aria-label="anchor" href="#the-three-tables"></a>
</h4>
<p>Most of the data we use as examples in this book are created with
high-throughput technologies. These technologies measure thousands of
<em>features</em>. Examples of features are genes, single base locations
of the genome, genomic regions, or image pixel intensities. Each
specific measurement product is defined by a specific set of features.
For example, a specific gene expression microarray product is defined by
the set of genes that it measures.</p>
<p>A specific study will typically use one product to make measurements
on several experimental units, such as individuals. The most common
experimental unit will be the individual, but they can also be defined
by other entities, for example different parts of a tumor. We often call
the experimental units <em>samples</em> following experimental jargon.
It is important that these are not confused with samples as referred to
in previous chapters, for example “random sample”.</p>
<p>So a high-throughput experiment is usually defined by three tables:
one with the high-throughput measurements and two tables with
information about the columns and rows of this first table
respectively.</p>
<p>Because a dataset is typically defined by a set of experimental units
and a product defines a fixed set of features, the high-throughput
measurements can be stored in an <i>n x m</i> matrix, with <i>n</i> the
number of units and <i>m</i> the number of features. In R, the
convention has been to store the transpose of these matrices, in which
all the rows become columns and the columns become the rows.</p>
<p>Here is an example from a gene expression dataset:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">load</span><span class="op">(</span><span class="st">"./data/GSE5859Subset.rda"</span><span class="op">)</span> <span class="co">## this loads the three tables</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 8793   24</code></pre>
</div>
<p>We have RNA expression measurements for 8793 genes from blood taken
from 24 individuals (the experimental units). For most statistical
analyses, we will also need information about the individuals. For
example, in this case the data was originally collected to compare gene
expression across ethnic groups. However, we have created a subset of
this dataset for illustration and separated the data into two
groups:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 24  4</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    ethnicity       date         filename group
107       ASN 2005-06-23 GSM136508.CEL.gz     1
122       ASN 2005-06-27 GSM136530.CEL.gz     1
113       ASN 2005-06-27 GSM136517.CEL.gz     1
163       ASN 2005-10-28 GSM136576.CEL.gz     1
153       ASN 2005-10-07 GSM136566.CEL.gz     1
161       ASN 2005-10-07 GSM136574.CEL.gz     1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">group</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0</code></pre>
</div>
<p>One of the columns, filenames, permits us to connect the rows of this
table to the columns of the measurement table.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">match</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">filename</span>, <span class="fu">colnames</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24</code></pre>
</div>
<p>Finally, we have a table describing the features:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">geneAnnotation</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 8793    4</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">geneAnnotation</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     PROBEID  CHR     CHRLOC SYMBOL
1  1007_s_at chr6   30852327   DDR1
30   1053_at chr7  -73645832   RFC2
31    117_at chr1  161494036  HSPA6
32    121_at chr2 -113973574   PAX8
33 1255_g_at chr6   42123144 GUCA1A
34   1294_at chr3  -49842638   UBA7</code></pre>
</div>
<p>The table includes an ID that permits us to connect the rows of this
table with the rows of the measurement table:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">match</span><span class="op">(</span><span class="va">geneAnnotation</span><span class="op">$</span><span class="va">PROBEID</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 3 4 5 6</code></pre>
</div>
<p>The table also includes biological information about the features,
namely chromosome location and the gene “name” used by biologists.</p>
<div id="exercise-1-how-many-samples-were-processed-on-2005-06-27" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1-how-many-samples-were-processed-on-2005-06-27" class="callout-inner">
<h3 class="callout-title">Exercise 1: How many samples were processed on
2005-06-27?</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<pre><code><span><span class="fu">unique</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span><span class="op">)</span> <span class="co"># check date format</span></span>
<span><span class="va">sampleInfo</span><span class="op">[</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span> <span class="op">==</span> <span class="st">"2005-06-27"</span>,<span class="op">]</span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span> <span class="op">==</span> <span class="st">"2005-06-27"</span><span class="op">)</span> <span class="co"># sum of TRUEs     </span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="exercise-2-how-many-of-the-genes-represented-in-this-particular-technology-are-on-chromosome-y" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2-how-many-of-the-genes-represented-in-this-particular-technology-are-on-chromosome-y" class="callout-inner">
<h3 class="callout-title">Exercise 2: How many of the genes represented
in this particular technology are on chromosome Y?</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<pre><code><span><span class="fu">unique</span><span class="op">(</span><span class="va">geneAnnotation</span><span class="op">$</span><span class="va">CHR</span><span class="op">)</span> <span class="co"># check chromosome spelling</span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="va">geneAnnotation</span><span class="op">$</span><span class="va">CHR</span> <span class="op">==</span> <span class="st">"chrY"</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># remove missing values # (NAs) to sum TRUEs</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="exercise-3-what-is-the-log-expression-value-for-gene-arpc1a-on-the-one-subject-that-we-measured-on-2005-06-10" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3-what-is-the-log-expression-value-for-gene-arpc1a-on-the-one-subject-that-we-measured-on-2005-06-10" class="callout-inner">
<h3 class="callout-title">Exercise 3: What is the log expression value
for gene ARPC1A on the one subject that we measured on 2005-06-10?</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<pre><code><span><span class="va">sampleInfo</span><span class="op">[</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span> <span class="op">==</span> <span class="st">"2005-06-10"</span>,<span class="op">]</span> <span class="co"># June 10 sample</span></span>
<span><span class="va">sampleFileName</span> <span class="op">&lt;-</span> <span class="va">sampleInfo</span><span class="op">[</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span> <span class="op">==</span> <span class="st">"2005-06-10"</span>, <span class="st">"filename"</span><span class="op">]</span> <span class="co"># save file name</span></span>
<span><span class="va">sampleProbeID</span> <span class="op">&lt;-</span> <span class="va">geneAnnotation</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="va">geneAnnotation</span><span class="op">$</span><span class="va">SYMBOL</span> <span class="op">==</span> <span class="st">"ARPC1A"</span><span class="op">)</span>, <span class="st">"PROBEID"</span><span class="op">]</span> <span class="co"># save probe ID</span></span>
<span><span class="va">geneExpression</span><span class="op">[</span><span class="va">sampleProbeID</span>, <span class="va">sampleFileName</span><span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="discussion" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="discussion" class="callout-inner">
<h3 class="callout-title">Discussion</h3>
<div class="callout-content">
<p>What kinds of research questions might you ask of this data? What are
the dependent (response) and independent variables? Turn to a partner
and discuss, then share with the group in the collaborative
document.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>High-throughput data measures thousands of features.</li>
<li>High-throughput data is typically composed of multiple tables.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</div></section><section id="aio-inference-for-highthroughput"><p>Content from <a href="inference-for-highthroughput.html">Basic inference for high-throughput data</a></p>
<hr>
<p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/inference-for-highthroughput.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How are inferences from high-throughput data different from
inferences from smaller samples?</li>
<li>How is the interpretation of p-values affected by high-throughput
data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Demonstrate that p-values are random variables.</li>
<li>Simulate p-values from multiple analyses of high-throughput
data.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="inference-in-practice">Inference in Practice<a class="anchor" aria-label="anchor" href="#inference-in-practice"></a>
</h2>
<hr class="half-width">
<p>Suppose we were given high-throughput gene expression data that was
measured for several individuals in two populations. We are asked to
report which genes have different average expression levels in the two
populations. If instead of thousands of genes, we were handed data from
just one gene, we could simply apply the inference techniques that we
have learned before. We could, for example, use a t-test or some other
test. Here we review what changes when we consider high-throughput
data.</p>
<div class="section level4">
<h4 id="p-values-are-random-variables">p-values are random variables<a class="anchor" aria-label="anchor" href="#p-values-are-random-variables"></a>
</h4>
<p>An important concept to remember in order to understand the concepts
presented in this chapter is that p-values are random variables. Let’s
revisit random variables as addressed in <a href="https://smcclatchy.github.io/statistical-inference-for-biology/02-inference-rv-dists/index.html" class="external-link">this
Statistical Inference for Biology episode</a>. We have a dataset that we
imagine contains the weights of all control female mice. We refer to
this as the population, even though it is impossible to have data for
all mice in population. For illustrative purposes we imagine having
access to an entire population, though in practice we cannot.</p>
<p>To illustrate a random variable, sample 12 mice from the population
three times and watch how the average changes.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">population</span> <span class="op">&lt;-</span> <span class="fu">unlist</span><span class="op">(</span><span class="fu">read.csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"./data/femaleControlsPopulation.csv"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23.185</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 22.25583</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 24.85</code></pre>
</div>
<p>Notice that the mean is a random variable. To explore p-values as
random variables, consider the example in which we define a p-value from
a t-test with a large enough sample size to use the <a href="https://smcclatchy.github.io/statistical-inference-for-biology/04-inference-clt/index.html" class="external-link">Central
Limit Theorem (CLT) approximation</a>. The CLT says that when the sample
size is large, the average Ȳ of a random sample follows a normal
distribution centered at the population average μ<sub>Y</sub> and with
standard deviation equal to the population standard deviation
σ<sub>Y</sub>, divided by the square root of the sample size
<em>N</em>.Then our p-value is defined as the probability that a
normally distributed random variable is larger, in absolute value, <a href="https://smcclatchy.github.io/statistical-inference-for-biology/06-inference-ttests-practice/index.html" class="external-link">than
the observed t-test</a>, call it <i>Z</i>. Recall that a t-statistic is
the ratio of the observed effect size and the standard error. Because
it’s the ratio of two random variables, it too is a random variable. So
for a two sided test the p-value is:</p>
<figure><img src=".././fig/pequals.png" class="figure mx-auto d-block"></figure><p>In R, we write:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">2</span> <span class="op">*</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu">pnorm</span><span class="op">(</span><span class="fu">abs</span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Now because <i>Z</i> is a random variable and <i>Φ</i> is a
deterministic function, <i>p</i> is also a random variable. We will
create a Monte Carlo simulation showing how the values of <i>p</i>
change.</p>
<p>We use <code>replicate</code> to repeatedly create p-values.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="op">{</span></span>
<span>  <span class="va">control</span> <span class="op">=</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span> <span class="op">=</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>, <span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">p.val</span> </span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span></span></code></pre>
</div>
<p>As implied by the histogram, in this case the distribution of the
p-value is uniformly distributed. In fact, we can show theoretically
that when the null hypothesis is true, this is always the case. For the
case in which we use the CLT, we have that the null hypothesis
<i>H<sub>0</sub></i> implies that our test statistic <i>Z</i> follows a
normal distribution with mean 0 and SD 1 thus:</p>
<p><img src=".././fig/psubaequals.png" class="figure"> This implies that:</p>
<figure><img src=".././fig/probpltpsuba.png" class="figure mx-auto d-block"></figure><p>which is the definition of a uniform distribution.</p>
</div>
<div class="section level4">
<h4 id="thousands-of-tests">Thousands of tests<a class="anchor" aria-label="anchor" href="#thousands-of-tests"></a>
</h4>
<p>In this data we have two groups denoted with 0 and 1:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">load</span><span class="op">(</span><span class="st">"data/GSE5859Subset.rda"</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">sampleInfo</span><span class="op">$</span><span class="va">group</span></span>
<span><span class="va">g</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0</code></pre>
</div>
<p>If we were interested in a particular gene, let’s arbitrarily pick
the one on the 25th row, we would simply compute a t-test. To compute a
p-value, we will use the t-distribution approximation and therefore we
need the population data to be approximately normal. Recall that when
the CLT does not apply (e.g. when sample sizes aren’t large enough), we
can <a href="https://smcclatchy.github.io/statistical-inference-for-biology/06-inference-ttests-practice/index.html" class="external-link">use
the t-distribution</a>.</p>
<p>We check our assumption that the population is normal with a
qq-plot:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">e</span> <span class="op">&lt;-</span> <span class="va">geneExpression</span><span class="op">[</span><span class="fl">25</span>, <span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="va">g</span><span class="op">==</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="va">g</span><span class="op">==</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="va">g</span><span class="op">==</span><span class="fl">0</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="va">g</span><span class="op">==</span><span class="fl">0</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src=".././fig/qqplots_for_one_gene-1.png" alt="Normal qq-plots for one gene. Left plot shows first group and right plot shows second group." class="figure mx-auto d-block"><div class="figcaption">Normal qq-plots for one gene. Left plot shows
first group and right plot shows second group.</div>
</figure><p>The qq-plots show that the data is well approximated by the normal
approximation. The t-test does not find this gene to be statistically
significant:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">t.test</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="va">g</span><span class="op">==</span><span class="fl">1</span><span class="op">]</span>, <span class="va">e</span><span class="op">[</span><span class="va">g</span><span class="op">==</span><span class="fl">0</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.779303</code></pre>
</div>
<p>To answer the question for each gene, we simply repeat the above for
each gene. Here we will define our own function and use
<code>apply</code>:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myttest</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> </span>
<span>  <span class="fu">t.test</span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">g</span><span class="op">==</span><span class="fl">1</span><span class="op">]</span>, </span>
<span>         <span class="va">x</span><span class="op">[</span><span class="va">g</span><span class="op">==</span><span class="fl">0</span><span class="op">]</span>, </span>
<span>         var.equal<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">apply</span><span class="op">(</span><span class="va">geneExpression</span>, <span class="fl">1</span>, <span class="va">myttest</span><span class="op">)</span></span></code></pre>
</div>
<p>We can now see which genes have p-values less than, say, 0.05. For
example, right away we see that…</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sum</span><span class="op">(</span><span class="va">pvals</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1383</code></pre>
</div>
<p>… genes had p-values less than 0.05.</p>
<p>However, as we will describe in more detail below, we have to be
careful in interpreting this result because we have performed over 8,000
tests. If we performed the same procedure on random data, for which the
null hypothesis is true for all features, we obtain the following
results:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span></span>
<span><span class="va">randomData</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">n</span> <span class="op">*</span> <span class="va">m</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span></span>
<span><span class="va">nullpvals</span> <span class="op">&lt;-</span> <span class="fu">apply</span><span class="op">(</span><span class="va">randomData</span>, <span class="fl">1</span>, <span class="va">myttest</span><span class="op">)</span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="va">nullpvals</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 419</code></pre>
</div>
<div id="discussion" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="discussion" class="callout-inner">
<h3 class="callout-title">Discussion</h3>
<div class="callout-content">
<p>Turn to a partner and explain what you did and found in the previous
two analyses. What do you think the results mean? Then, share your
responses with the group through the collaborative document.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<p>As we will explain later in the chapter, this is to be expected: 419
is roughly 0.05 * 8192 and we will describe the theory that tells us why
this prediction works.</p>
</div>
<div class="section level4">
<h4 id="faster-t-test-implementation">Faster t-test implementation<a class="anchor" aria-label="anchor" href="#faster-t-test-implementation"></a>
</h4>
<p>Before we continue, we should point out that the above implementation
is very inefficient. There are several faster implementations that
perform t-test for high-throughput data. We make use of a function that
is not available from CRAN, but rather from the Bioconductor project.
Now we can show that this function is much faster than our code above
and produce practically the same answer:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">genefilter</span><span class="op">)</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">geneExpression</span>, <span class="fu">factor</span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">max</span><span class="op">(</span><span class="fu">abs</span><span class="op">(</span><span class="va">pvals</span> <span class="op">-</span> <span class="va">results</span><span class="op">$</span><span class="va">p</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 6.528111e-14</code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<hr class="half-width">
<p>These exercises will help clarify that p-values are random variables
and some of the properties of these p-values. Note that just like the
sample average is a random variable because it is based on a random
sample, the p-values are based on random variables (sample mean and
sample standard deviation for example) and thus it is also a random
variable.<br>
To see this, let’s see how p-values change when we take different
samples.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="fl">1000</span>, <span class="op">{</span> <span class="co"># recreate p-values as from above</span></span>
<span>  <span class="va">control</span> <span class="op">=</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="fl">12</span><span class="op">)</span> </span>
<span>  <span class="va">treatment</span> <span class="op">=</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="fl">12</span><span class="op">)</span> </span>
<span>  <span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>,<span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">p.val</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.3191557945 0.2683723148 0.0003358878 0.0312671917 0.1410320545
[6] 0.9478677657</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">hist</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span></span></code></pre>
</div>
<div id="exercise-1-what-proportion-of-the-p-values-is-below-0.05" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1-what-proportion-of-the-p-values-is-below-0.05" class="callout-inner">
<h3 class="callout-title">Exercise 1: What proportion of the p-values is
below 0.05?</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p><code>sum(pvals &lt; 0.05)/length(pvals)</code><br>
50 of the p-values are less than 0.05 of a total<br>
1000 p-values</p>
</div>
</div>
</div>
</div>
<div id="exercise-2-what-proportion-of-the-p-values-is-below-0.01" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2-what-proportion-of-the-p-values-is-below-0.01" class="callout-inner">
<h3 class="callout-title">Exercise 2: What proportion of the p-values is
below 0.01?</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p><code>sum(pvals &lt; 0.01)/length(pvals)</code><br>
11 of the p-values are less than 0.01 of a total<br>
1000 p-values</p>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>Assume you are testing the effectiveness of 20 diets on mice weight.
For each of the 20 diets, you run an experiment with 10 control mice and
10 treated mice. Assume the null hypothesis, that the diet has no
effect, is true for all 20 diets and that mice weights follow a normal
distribution, with mean 30 grams and a standard deviation of 2 grams.
Run a Monte Carlo simulation for one of these studies:</p>
<pre><code><span><span class="va">cases</span> <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">30</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">controls</span> <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">30</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">t.test</span><span class="op">(</span><span class="va">cases</span>, <span class="va">controls</span><span class="op">)</span>     </span></code></pre>
<p>Now run a Monte Carlo simulation imitating the results for the
experiment for all 20 diets. If you set the seed at 100, set.seed(100),
how many of p-values are below 0.05?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p><code>set.seed(100)</code><br><code>n &lt;- 20</code><br><code>null &lt;- vector("numeric", n)</code><br><code>for (i in 1:n) {</code><br><code>cases = rnorm(10, 30, 2)</code><br><code>controls = rnorm(10, 30, 2)</code><br><code>null[i] &lt;- t.test(cases, controls)$p.val</code><br><code>}</code><br><code>sum(null &lt; 0.05)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>Now create a simulation to learn about the distribution of the number
of p-values that are less than 0.05. In question 3, we ran the 20 diet
experiment once. Now we will run the experiment 1,000 times and each
time save the number of p-values that are less than 0.05. Set the seed
at 100, set.seed(100), run the code from Question 3 1,000 times, and
save the number of times the p-value is less than 0.05 for each of the
1,000 instances. What is the average of these numbers? This is the
expected number of tests (out of the 20 we run) that we will reject when
the null is true.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<p><code>set.seed(100)</code><br><code>res &lt;- replicate(1000, {</code><br><code>randomData &lt;- matrix(rnorm(20*20,30,2),20,20)</code><br><code>pvals &lt;- rowttests(randomData, factor(g))$p.value</code><br><code>return(sum(pvals&lt;0.05))</code> # total number of false
positives per replication<br><code>})</code><br><code>mean(res)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-5" class="callout-inner">
<h3 class="callout-title">Exercise 5</h3>
<div class="callout-content">
<p>What this says is that on average, we expect some p-value to be 0.05
even when the null is true for all diets. Use the same simulation data
and report for what percent of the 1,000 replications did we make more
than 0 false positives?</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<p><code>mean(res 0)</code></p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>P-values are random variables.</li>
<li>Very small p-values can occur by random chance when analyzing
high-throughput data.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-procedures-for-multiple-testing"><p>Content from <a href="procedures-for-multiple-testing.html">Procedures for Multiple Comparisons</a></p>
<hr>
<p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/procedures-for-multiple-testing.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 30 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why are p-values not a useful quantity when dealing with
high-dimensional data?</li>
<li>What are error rates and how are they calculated?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Define multiple comparisons and the resulting problems.</li>
<li>Explain terms associated with error rates.</li>
<li>Explain the relationship between error rates and multiple
comparisons.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="procedures">Procedures<a class="anchor" aria-label="anchor" href="#procedures"></a>
</h2>
<hr class="half-width">
<p>In the previous section we learned how p-values are no longer a
useful quantity to interpret when dealing with high-dimensional data.
This is because we are testing many <em>features</em> at the same time.
We refer to this as the <em>multiple comparison</em> or <em>multiple
testing</em> or <em>multiplicity</em> problem. The definition of a
p-value does not provide a useful quantification here. Again, because
when we test many hypotheses simultaneously, a list based simply on a
small p-value cut-off of, say 0.01, can result in many false positives
with high probability. Here we define terms that are more appropriate in
the context of high-throughput data.</p>
<p>The most widely used approach to the multiplicity problem is to
define a <em>procedure</em> and then estimate or <em>control</em> an
informative <em>error rate</em> for this procedure. What we mean by
<em>control</em> here is that we adapt the procedure to guarantee an
<em>error rate</em> below a predefined value. The procedures are
typically flexible through parameters or cutoffs that let us control
specificity and sensitivity. An example of a procedure is:</p>
<ul>
<li>Compute a p-value for each gene.</li>
<li>Call significant all genes with p-values smaller than <span class="math inline">\(\alpha\)</span>.</li>
</ul>
<p>Note that changing the <span class="math inline">\(\alpha\)</span>
permits us to adjust specificity and sensitivity.</p>
<p>Next we define the <em>error rates</em> that we will try to estimate
and control.</p>
<p>Review the matrix below to understand sensitivity, specificity, and
error rates.</p>
<figure><img src=".././fig/confusion-matrix-1.png" alt="confusion matrix showing specificity, sensitivity and Type I and Type II errors" class="figure mx-auto d-block"><div class="figcaption">confusion matrix showing specificity,
sensitivity and Type I and Type II errors</div>
</figure><div id="discussion" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="discussion" class="callout-inner">
<h3 class="callout-title">Discussion</h3>
<div class="callout-content">
<p>Turn to a partner and explain the following:<br>
1. Type I error<br>
2. Type II error<br>
3. sensitivity<br>
4. specificity<br>
When you are finished discussing, share with the group in the
collaborative document.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<p>For more on this, see <a href="https://www.nature.com/articles/nmeth.3945" class="external-link">Classification
evaluation</a> by J. Lever, M. Krzywinski and N. Altman in Nature
Methods 13, 603-604 (2016).</p>
</section><section><h2 class="section-heading" id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<hr class="half-width">
<p>With these exercises we hope to help you further grasp the concept
that p-values are random variables and start laying the ground work for
the development of procedures that control error rates. The calculations
to compute error rates require us to understand the random behavior of
p-values. We are going to ask you to perform some calculations related
to introductory probability theory. One particular concept you need to
understand is statistical independence. You also will need to know that
the probability of two random e vents that are statistically independent
occurring is P (A and B) = P (A)P (B). This is a consequence of the more
general formula P (A and B) = P (A)P (B|A).</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>Assume the null is true and denote the p-value you would get if you
ran a test as P. Define the function f (x) = Pr(P x) . What does f (x)
look like?<br>
A) A uniform distribution.<br>
B) The identity line.<br>
C) A constant at 0.05.<br>
D) P is not a random value.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>In the previous exercises, we saw how the probability of incorrectly
rejecting the null for at least one of 20 experiments for which the null
is true, is well over 5%. Now let’s consider a case in which we run
thousands of tests, as we would do in a high throughput experiment. We
previously learned that under the null, the probability of a p-value
&lt; p is p. If we run 8,793 independent tests, what it the probability
of incorrectly rejecting at least one of the null hypothesis?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p><code>1 - 0.95^8793</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>Suppose we need to run 8,793 statistical tests and we want to make
the probability of a mistake very small, say 5%. Use the answer to
Exercise 2 (Sidak’s procedure) to determine how small we have to change
the cutoff, previously 0.05, to lower our probability of at least one
mistake to be 5%.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p><code>m &lt;- 8793</code><br><code>1 - 0.95^(1/m)</code></p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>We want our inferential analyses to maximize the percentage of true
positives (sensitivity) and true negatives (specificity)</li>
<li>p values are random variables, conducting multiple comparisons on
high throughput data can produce a large number of false positives (Type
1 errors) simply by chance.</li>
<li>There are procedures for improving sensitivity and specificity by
controlling error rates below a predefined value.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-error-rates"><p>Content from <a href="error-rates.html">Error Rates</a></p>
<hr>
<p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/error-rates.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 30 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why are type I and II error rates a problem in inferential
statistics?</li>
<li>What is Family Wise Error Rate, and why is it a concern in high
throughput data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Calculate error rates in simulated data.</li>
<li>Calculate the number of false positives that occur in a given data
set.</li>
<li>Define family wise error rates.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="error-rates">Error Rates<a class="anchor" aria-label="anchor" href="#error-rates"></a>
</h2>
<hr class="half-width">
<p>Throughout this section we will be using the type I error and type II
error terminology. We will also refer to them as false positives and
false negatives respectively. We also use the more general terms
specificity, which relates to type I error, and sensitivity, which
relates to type II errors.</p>
<div class="section level4">
<h4 id="types-of-error">Types of Error<a class="anchor" aria-label="anchor" href="#types-of-error"></a>
</h4>
<p>Whenever we perform a statistical test, we are aware that we may make
a mistake. This is why our p-values are not 0. Under the null, there is
always a positive, perhaps very small, but still positive chance that we
will reject the null when it is true. If the p-value is 0.05, it will
happen 1 out of 20 times. This <em>error</em> is called <em>type I
error</em> by statisticians.</p>
<figure><img src=".././fig/null-hypothesis.png" alt="Null distribution showing type I error as alpha" class="figure mx-auto d-block"><div class="figcaption">Null distribution showing type I error as
alpha</div>
</figure><p>A type I error is defined as rejecting the null when we should not.
This is also referred to as a false positive. So why do we then use
0.05? Shouldn’t we use 0.000001 to be really sure? The reason we don’t
use infinitesimal cut-offs to avoid type I errors at all cost is that
there is another error we can commit: to not reject the null when we
should. This is called a <em>type II error</em> or a false negative.</p>
<figure><img src=".././fig/alternative-hypothesis.png" alt="Alternative hypothesis showing type II error as beta" class="figure mx-auto d-block"><div class="figcaption">Alternative hypothesis showing type II error as
beta</div>
</figure><p>In the context of high-throughput data we can make several type I
errors and several type II errors in one experiment, as opposed to one
or the other as seen in Chapter 1. In this table, we summarize the
possibilities using the notation from the seminal paper by
Benjamini-Hochberg:</p>
<table class="table">
<thead><tr class="header">
<th></th>
<th>Called significant</th>
<th>Not called significant</th>
<th>Total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Null True</td>
<td><span class="math inline">\(V\)</span></td>
<td><span class="math inline">\(m_0-V\)</span></td>
<td><span class="math inline">\(m_0\)</span></td>
</tr>
<tr class="even">
<td>Alternative True</td>
<td><span class="math inline">\(S\)</span></td>
<td><span class="math inline">\(m_1-S\)</span></td>
<td><span class="math inline">\(m_1\)</span></td>
</tr>
<tr class="odd">
<td>True</td>
<td><span class="math inline">\(R\)</span></td>
<td><span class="math inline">\(m-R\)</span></td>
<td><span class="math inline">\(m\)</span></td>
</tr>
</tbody>
</table>
<p>To describe the entries in the table, let’s use as an example a
dataset representing measurements from 10,000 genes, which means that
the total number of tests that we are conducting is: <span class="math inline">\(m=10,000\)</span>. The number of genes for which
the null hypothesis is true, which in most cases represent the
“non-interesting” genes, is <span class="math inline">\(m_0\)</span>,
while the number of genes for which the null hypothesis is false is
<span class="math inline">\(m_1\)</span>. For this we can also say that
the <em>alternative hypothesis</em> is true. In general, we are
interested in <em>detecting</em> as many as possible of the cases for
which the alternative hypothesis is true (true positives), without
incorrectly detecting cases for which the null hypothesis is true (false
positives). For most high-throughput experiments, we assume that <span class="math inline">\(m_0\)</span> is much greater than <span class="math inline">\(m_1\)</span>. For example, we test 10,000
expecting 100 genes or less to be <em>interesting</em>. This would imply
that <span class="math inline">\(m_1 \leq 100\)</span> and <span class="math inline">\(m_0 \geq 19,900\)</span>.</p>
<p>Throughout this chapter we refer to <em>features</em> as the units
being tested. In genomics, examples of features are genes, transcripts,
binding sites, CpG sites, and SNPs. In the table, <span class="math inline">\(R\)</span> represents the total number of features
that we call significant after applying our procedure, while <span class="math inline">\(m-R\)</span> is the total number of genes we don’t
call significant. The rest of the table contains important quantities
that are unknown in practice.</p>
<ul>
<li>
<span class="math inline">\(V\)</span> represents the number of type
I errors or false positives. Specifically, <span class="math inline">\(V\)</span> is the number of features for which the
null hypothesis is true, that we call significant.</li>
<li>
<span class="math inline">\(S\)</span> represents the number of true
positives. Specifically, <span class="math inline">\(S\)</span> is the
number of features for which the alternative is true, that we call
significant.</li>
</ul>
<p>This implies that there are <span class="math inline">\(m_1-S\)</span> type II errors or <em>false
negatives</em> and <span class="math inline">\(m_0-V\)</span> true
negatives. Keep in mind that if we only ran one test, a p-value is
simply the probability that <span class="math inline">\(V=1\)</span>
when <span class="math inline">\(m=m_0=1\)</span>. Power is the
probability of <span class="math inline">\(S=1\)</span> when <span class="math inline">\(m=m_1=1\)</span>. In this very simple case, we
wouldn’t bother making the table above, but now we show how defining the
terms in the table helps for the high-dimensional setting.</p>
</div>
<div class="section level4">
<h4 id="data-example">Data example<a class="anchor" aria-label="anchor" href="#data-example"></a>
</h4>
<p>Let’s compute these quantities with a data example. We will use a
Monte Carlo simulation using our mice data to imitate a situation in
which we perform tests for 10,000 different fad diets, none of them
having an effect on weight. This implies that the null hypothesis is
true for diets and thus <span class="math inline">\(m=m_0=10,000\)</span> and <span class="math inline">\(m_1=0\)</span>. Let’s run the tests with a sample
size of <span class="math inline">\(N=12\)</span> and compute <span class="math inline">\(R\)</span>. Our procedure will declare any diet
achieving a p-value smaller than <span class="math inline">\(\alpha=0.05\)</span> as significant.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">population</span> <span class="op">=</span> <span class="fu">unlist</span><span class="op">(</span> <span class="fu">read.csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"./data/femaleControlsPopulation.csv"</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">m</span>,<span class="op">{</span></span>
<span>  <span class="va">control</span> <span class="op">=</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>,<span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span> <span class="op">=</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>,<span class="va">N</span><span class="op">)</span></span>
<span>  <span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>,<span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<p>Although in practice we do not know the fact that no diet works, in
this simulation we do, and therefore we can actually compute <span class="math inline">\(V\)</span> and <span class="math inline">\(S\)</span>. Because all null hypotheses are true,
we know, in this specific simulation, that <span class="math inline">\(V=R\)</span>. Of course, in practice we can
compute <span class="math inline">\(R\)</span> but not <span class="math inline">\(V\)</span>.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sum</span><span class="op">(</span><span class="va">pvals</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 409</code></pre>
</div>
<p>These many false positives are not acceptable in most contexts.</p>
<p>Here is more complicated code showing results where 10% of the diets
are effective with an average effect size of <span class="math inline">\(\Delta= 3\)</span> ounces. Studying this code
carefully will help us understand the meaning of the table above. First
let’s define <em>the truth</em>:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="fl">0.90</span> <span class="co">##10% of diets work, 90% don't</span></span>
<span><span class="va">m0</span> <span class="op">&lt;-</span> <span class="va">m</span> <span class="op">*</span> <span class="va">p0</span></span>
<span><span class="va">m1</span> <span class="op">&lt;-</span> <span class="va">m</span> <span class="op">-</span> <span class="va">m0</span></span>
<span><span class="va">nullHypothesis</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span> <span class="fu">rep</span><span class="op">(</span><span class="cn">TRUE</span>, <span class="va">m0</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="cn">FALSE</span>, <span class="va">m1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">delta</span> <span class="op">&lt;-</span> <span class="fl">3</span></span></code></pre>
</div>
<p>Now we are ready to simulate 10,000 tests, perform a t-test on each,
and record if we rejected the null hypothesis or not:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">calls</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">m</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span> <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="va">treatment</span> <span class="op">+</span> <span class="va">delta</span></span>
<span>  <span class="fu">ifelse</span><span class="op">(</span> <span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>, <span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span> <span class="op">&lt;</span> <span class="va">alpha</span>, </span>
<span>          <span class="st">"Called Significant"</span>,</span>
<span>          <span class="st">"Not Called Significant"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<p>Because in this simulation we know the truth (saved in
<code>nullHypothesis</code>), we can compute the entries of the
table:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">null_hypothesis</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">nullHypothesis</span>, levels<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"FALSE"</span>, <span class="st">"TRUE"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">table</span><span class="op">(</span><span class="va">null_hypothesis</span>, <span class="va">calls</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               calls
null_hypothesis Called Significant Not Called Significant
          FALSE                553                    447
          TRUE                 361                   8639</code></pre>
</div>
<p>The first column of the table above shows us <span class="math inline">\(V\)</span> and <span class="math inline">\(S\)</span>. Note that <span class="math inline">\(V\)</span> and <span class="math inline">\(S\)</span> are random variables. If we run the
simulation repeatedly, these values change. Here is a quick example:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">10</span> <span class="co">##number of simulations</span></span>
<span><span class="va">VandS</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>,<span class="op">{</span></span>
<span>  <span class="va">calls</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">m</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">)</span></span>
<span>    <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">)</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span> <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="va">treatment</span> <span class="op">+</span> <span class="va">delta</span></span>
<span>    <span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>, <span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">p.val</span> <span class="op">&lt;</span> <span class="va">alpha</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="fu">cat</span><span class="op">(</span><span class="st">"V ="</span>,<span class="fu">sum</span><span class="op">(</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span>, <span class="st">"S ="</span>,<span class="fu">sum</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span>,<span class="st">"\n"</span><span class="op">)</span></span>
<span>  <span class="fu">c</span><span class="op">(</span><span class="fu">sum</span><span class="op">(</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span>,<span class="fu">sum</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>V = 373 S = 540
V = 375 S = 553
V = 386 S = 545
V = 396 S = 565
V = 373 S = 532
V = 355 S = 541
V = 372 S = 501
V = 376 S = 522
V = 355 S = 570
V = 388 S = 506 </code></pre>
</div>
<p>This motivates the definition of error rates. We can, for example,
estimate probability that <span class="math inline">\(V\)</span> is
larger than 0. This is interpreted as the probability of making at least
one type I error among the 10,000 tests. In the simulation above, <span class="math inline">\(V\)</span> was much larger than 1 in every single
simulation, so we suspect this probability is very practically 1. When
<span class="math inline">\(m=1\)</span>, this probability is equivalent
to the p-value. When we have a multiple tests situation, we call it the
Family Wise Error Rate (FWER) and it relates to a technique that is
widely used: The Bonferroni Correction.</p>
<div id="discussion" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="discussion" class="callout-inner">
<h3 class="callout-title">Discussion</h3>
<div class="callout-content">
<p>Refer to the confusion matrix below showing false positive and false
negative rates, along with specificity and sensitivity. Turn to a
partner and explain the following:<br>
How is specificity related to Type I error and false positive
rates?<br>
How is sensitivity related to Type II error and false negative
rates?<br>
When you are finished discussing, share with the group in the
collaborative document.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<figure><img src=".././fig/confusion-matrix-2.png" alt="confusion matrix showing error rates" class="figure mx-auto d-block"><div class="figcaption">confusion matrix showing error rates</div>
</figure><p>For more on this, see <a href="https://www.nature.com/articles/nmeth.3945" class="external-link"><i>Classification
evaluation</i></a> by J. Lever, M. Krzywinski and N. Altman in <i>Nature
Methods</i> <b>13</b>, 603-604 (2016).</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Type 1 and Type II error are complementary concerns in data
analysis. The reason we don’t have extremely strict cutoffs for alpha is
because we don’t want to miss finding true positive results.</li>
<li>It is possible to calculate the probability of finding at least one
false positive result when conducting multiple inferential tests. This
is the Family Wise Error Rate.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-bonferroni-correction"><p>Content from <a href="bonferroni-correction.html">The Bonferroni Correction</a></p>
<hr>
<p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/bonferroni-correction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 45 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is one way to control family wise error rate?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Apply the Bonferroni procedure to identify the true positives.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="the-bonferroni-correction">The Bonferroni Correction<a class="anchor" aria-label="anchor" href="#the-bonferroni-correction"></a>
</h2>
<hr class="half-width">
<p>Now that we have learned about the Family Wise Error Rate (FWER), we
describe what we can actually do to control it. In practice, we want to
choose a <em>procedure</em> that guarantees the FWER is smaller than a
predetermined value such as 0.05. We can keep it general and instead of
0.05, use <span class="math inline">\(\alpha\)</span> in our
derivations.</p>
<p>Since we are now describing what we do in practice, we no longer have
the advantage of knowing <em>the truth</em>. Instead, we pose a
procedure and try to estimate the FWER. Let’s consider the naive
procedure: “reject all the hypotheses with p-value &lt;0.01”. For
illustrative purposes we will assume all the tests are independent (in
the case of testing diets this is a safe assumption; in the case of
genes it is not so safe since some groups of genes act together). Let
<span class="math inline">\(p_1,\dots,p_{10000}\)</span> be the the
p-values we get from each test. These are independent random variables
so:</p>
<p><span class="math display">\[
\begin{aligned}
\mbox{Pr}(\mbox{at least one rejection}) &amp;= 1 -\mbox{Pr}(\mbox{no
rejections}) \\
&amp;= 1 - \prod_{i=1}^{1000} \mbox{Pr}(p_i&gt;0.01) \\
&amp;= 1-0.99^{1000} \approx 1
\end{aligned}
\]</span></p>
<p>Or if you want to use simulations:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span><span class="va">minpval</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="fu">min</span><span class="op">(</span><span class="fu">runif</span><span class="op">(</span><span class="fl">10000</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">minpval</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1</code></pre>
</div>
<p>So our FWER is 1! This is not what we were hoping for. If we wanted
it to be lower than <span class="math inline">\(\alpha=0.05\)</span>, we
failed miserably.</p>
<p>So what do we do to make the probability of a mistake lower than
<span class="math inline">\(\alpha\)</span> ? Using the derivation above
we can change the procedure by selecting a more stringent cutoff,
previously 0.01, to lower our probability of at least one mistake to be
5%. Namely, by noting that:</p>
<p><span class="math inline">\(\mbox{Pr}(\mbox{at least one rejection})
=  1-(1-k)^{10000}\)</span></p>
<p>and solving for <span class="math inline">\(k\)</span>, we get <span class="math inline">\(1-(1-k)^{10000}=0.01 \implies k = 1-0.99^{1/10000}
\approx 1e-6\)</span></p>
<p>This now gives a specific example of a <em>procedure</em>. This one
is actually called Sidak’s procedure. Specifically, we define a set of
instructions, such as “reject all the null hypothesis for which p-values
&lt; 1e-6. Then, knowing the p-values are random variables, we use
statistical theory to compute how many mistakes, on average, we are
expected to make if we follow this procedure. More precisely, we compute
bounds on these rates; that is, we show that they are smaller than some
predetermined value. There is a preference in the life sciences to err
on the side of being conservative.</p>
<p>A problem with Sidak’s procedure is that it assumes the tests are
independent. It therefore only controls FWER when this assumption holds.
The Bonferroni correction is more general in that it controls FWER even
if the tests are not independent. As with Sidak’s procedure we start by
noting that:</p>
<p><span class="math inline">\(FWER = \mbox{Pr}(V&gt;0) \leq
\mbox{Pr}(V&gt;0 \mid \mbox{all nulls are true})\)</span></p>
<p>or using the notation from the table above:</p>
<p><span class="math inline">\(\mbox{Pr}(V&gt;0) \leq \mbox{Pr}(V&gt;0
\mid m_1=0)\)</span></p>
<p>The Bonferroni procedure sets <span class="math inline">\(k=\alpha/m\)</span> since we can show that:</p>
<p><span class="math display">\[
\begin{align*}
\mbox{Pr}(V&gt;0 \,\mid \, m_1=0) &amp;= \mbox{Pr}\left( \min_i \{p_i\}
\leq \frac{\alpha}{m} \mid m_1=0 \right)\\
&amp;\leq \sum_{i=1}^m \mbox{Pr}\left(p_i \leq \frac{\alpha}{m}
\right)\\
&amp;= m \frac{\alpha}{m}=\alpha
\end{align*}
\]</span></p>
<p>Controlling the FWER at 0.05 is a very conservative approach. Using
the p-values computed in the previous section…</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">population</span> <span class="op">&lt;-</span> <span class="fu">unlist</span><span class="op">(</span><span class="fu">read.csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"./data/femaleControlsPopulation.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="fl">0.90</span> <span class="co">##10% of diets work, 90% don't</span></span>
<span><span class="va">m0</span> <span class="op">&lt;-</span> <span class="va">m</span> <span class="op">*</span> <span class="va">p0</span></span>
<span><span class="va">m1</span> <span class="op">&lt;-</span> <span class="va">m</span> <span class="op">-</span> <span class="va">m0</span></span>
<span><span class="va">nullHypothesis</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span> <span class="fu">rep</span><span class="op">(</span><span class="cn">TRUE</span>,<span class="va">m0</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="cn">FALSE</span>,<span class="va">m1</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">delta</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">m</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span> <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="va">treatment</span> <span class="op">+</span> <span class="va">delta</span></span>
<span>  <span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>, <span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<p>…we note that only:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sum</span><span class="op">(</span><span class="va">pvals</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">/</span><span class="fl">10000</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0</code></pre>
</div>
<p>are called significant after applying the Bonferroni procedure,
despite having 1,000 diets that work.</p>
</section><section><h2 class="section-heading" id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<hr class="half-width">
<p>The following exercises should help you understand the concept of an
error controlling procedure. You can think of it as defining a set of
instructions, such as “reject all the null hypothesis for which p-values
&lt; 0.0001” or “reject the null hypothesis for the 10 features with
smallest p-values”. Then, knowing the p-values are random variables, we
use statistical theory to compute how many mistakes, on average, we will
make if we follow this procedure. More precisely, we commonly find
bounds on these rates, meaning that we show that they are smaller than
some predetermined value. As described in the text, we can compute
different error rates. The FWER tells us the probability of having at
least one false positive. The FDR is the expected rate of rejected null
hypothesis.</p>
<p>Note 1: the FWER and FDR are not procedures, but error rates. We will
review procedures here and use Monte Carlo simulations to estimate their
error rates.</p>
<p>Note 2: We sometimes use the colloquial term “pick genes that”
meaning “reject the null hypothesis for genes that”.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>We have learned about the family wide error rate FWER. This is the
probability of incorrectly rejecting the null at least once. Using the
notation in the video, this probability is written like this: Pr(V 0).
What we want to do in practice is choose a procedure that guarantees
this probability is smaller than a predetermined value such as 0.05.
Here we keep it general and, instead of 0.05, we use α. We have already
learned that the procedure “pick all the genes with p-value &lt; 0.05”
fails miserably as we have seen that Pr(V 0) ≈ 1. So what else can we
do? The Bonferroni procedure assumes we have computed p-values for each
test and asks what constant k should we pick so that the procedure “pick
all genes with p-value less than k “ has Pr(V 0) = 0.05. Furthermore, we
typically want to be conservative rather than lenient, so we accept a
procedure that has Pr(V 0) ≤ 0.05. So the first result we rely on is
that this probability is largest when all the null hypotheses are true:
Pr(V 0) ≤ Pr(V 0|all nulls are true) or: Pr(V 0) ≤ Pr(V 0 | m1 = 0) We
showed that if the tests are independent then: Pr(V 0|m1) = 1−(1−k)m And
we pick k so that 1 − (1 − k)m = α =⇒ k=1−(1−α)1/m Now this requires the
tests to be independent. The Bonferroni procedure does not make this
assumption and, as we previously saw, sets k = α/m and shows that with
this choice of k this procedure results in P r(V 0) ≤ α. In R define
alphas &lt;- seq(0,0.25,0.01) Make a plot of α/m and 1 − (1 − α)1/m for
various values of m 1.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p><code>plot(alphas/m, (1-(1-alphas)^(1/m)), xlab = 'bonf', ylab = 'sidak',</code>
<code>main = 'p-val cutoff')</code><br><code>abline(0,1)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>Which procedure is more conservative Bonferroni’s or Sidak’s?<br>
A) They are the same.<br>
B) Bonferroni’s.<br>
C) Depends on m.<br>
D) Sidak’s.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>Bonferroni’s procedure is more conservative (choice B). Conservative
refers to strictness. The p-value cutoff for significance is lower in
Bonferroni’s procedure, and therefore more conservative.</p>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>To simulate the p-value results of, say 8,793 t-tests for which the
null is true, we don’t actually have to generate the original data. We
can generate p-values for a uniform distribution like this:
<code>pvals &lt;- runif(8793,0,1)</code>. Using what we have learned,
set the cutoff using the Bonferroni correction and report back the FWER.
Set the seed at 1 and run 10,000 simulation.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p><code>set.seed(1)</code><br><code>bonf_res &lt;- replicate(10000, {</code><br><code>pvals &lt;- runif(8793,0,1)</code><br><code>bonfcall &lt;- sum((pvals * m) &lt; 0.05)</code><br><code>return(bonfcall)</code><br><code>})</code><br><code>sum(bonf_res&gt;0)/length(bonf_res)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>Using the same seed, repeat exercise 5, but for Sidak’s cutoff.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p><code>set.seed(1)</code><br><code>sidak_res &lt;- replicate(10000, {</code><br><code>pvals &lt;- runif(8793,0,1)</code><br><code>sidakcall &lt;- sum((1-(1-pvals)^m) &lt; 0.05)</code><br><code>return(sidakcall)</code><br><code>sum(sidak_res&gt;0)/length(sidak_res)</code><br><code>})</code></p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>The Bonferroni technique controls FWER by dividing a predetermined
alpha rate (e.g. alpha = .05) by the number of inferential tests
performed.</li>
<li>The Bonferroni correction is very strict and conservative.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-false-discovery-rate"><p>Content from <a href="false-discovery-rate.html">False Discovery Rate</a></p>
<hr>
<p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/false-discovery-rate.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 40 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are False Discovery Rates, and when are they a concern in data
analysis?</li>
<li>How can you control false discovery rates?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Calculate the false discovery rate.</li>
<li>Explain the limitations of restricting family wise error rates in a
study.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="false-discovery-rate">False Discovery Rate<a class="anchor" aria-label="anchor" href="#false-discovery-rate"></a>
</h2>
<hr class="half-width">
<p>There are many situations for which requiring an FWER of 0.05 does
not make sense as it is much too strict. For example, consider the very
common exercise of running a preliminary small study to determine a
handful of candidate genes. This is referred to as a <em>discovery</em>
driven project or experiment. We may be in search of an unknown
causative gene and more than willing to perform follow-up studies with
many more samples on just the candidates. If we develop a procedure that
produces, for example, a list of 10 genes of which 1 or 2 pan out as
important, the experiment is a resounding success. With a small sample
size, the only way to achieve a FWER <span class="math inline">\(\leq\)</span> 0.05 is with an empty list of genes.
We already saw in the previous section that despite 1,000 diets being
effective, we ended up with a list with just 2. Change the sample size
to 6 and you very likely get 0:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">population</span> <span class="op">&lt;-</span> <span class="fu">unlist</span><span class="op">(</span><span class="fu">read.csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"./data/femaleControlsPopulation.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="fl">0.90</span></span>
<span><span class="va">m0</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">*</span><span class="va">p0</span></span>
<span><span class="va">m1</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">-</span><span class="va">m0</span></span>
<span><span class="va">nullHypothesis</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span> <span class="fu">rep</span><span class="op">(</span><span class="cn">TRUE</span>,<span class="va">m0</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="cn">FALSE</span>,<span class="va">m1</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">delta</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">m</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">control</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="fl">6</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="fl">6</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span> <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="va">treatment</span> <span class="op">+</span> <span class="va">delta</span></span>
<span>  <span class="fu">t.test</span><span class="op">(</span><span class="va">treatment</span>, <span class="va">control</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="va">pvals</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">/</span><span class="fl">10000</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0</code></pre>
</div>
<p>By requiring a FWER <span class="math inline">\(\leq\)</span> 0.05,
we are practically assuring 0 power (sensitivity). In many applications,
this specificity requirement is over-kill. A widely used alternative to
the FWER is the false discovery rate (FDR). The idea behind FDR is to
focus on the random variable <span class="math inline">\(Q \equiv
V/R\)</span> with <span class="math inline">\(Q=0\)</span> when <span class="math inline">\(R=0\)</span> and <span class="math inline">\(V=0\)</span>. Note that <span class="math inline">\(R=0\)</span> (nothing called significant) implies
<span class="math inline">\(V=0\)</span> (no false positives). So <span class="math inline">\(Q\)</span> is a random variable that can take
values between 0 and 1 and we can define a rate by considering the
average of <span class="math inline">\(Q\)</span>. To better understand
this concept here, we compute <span class="math inline">\(Q\)</span> for
the procedure: call everything p-value &lt; 0.05 significant.</p>
<div id="discussion" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="discussion" class="callout-inner">
<h3 class="callout-title">Discussion</h3>
<div class="callout-content">
<p>Refer to the confusion matrix below showing false positive, false
negative, and false discovery rates, along with specificity and
sensitivity. Turn to a partner and explain the following:<br>
How is the false discovery rate related to Type I error and false
positive rates?<br>
How is sensitivity related to the false discovery rate?<br>
When you are finished discussing, share with the group in the
collaborative document.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">

</div>
</div>
</div>
</div>
<figure><img src=".././fig/confusion-matrix-3.png" alt="confusion matrix showing error rates" class="figure mx-auto d-block"><div class="figcaption">confusion matrix showing error rates</div>
</figure><p>For more on this, see <a href="https://www.nature.com/articles/nmeth.3945" class="external-link"><i>Classification
evaluation</i></a> by J. Lever, M. Krzywinski and N. Altman in <i>Nature
Methods</i> <b>13</b>, 603-604 (2016).</p>
<div class="section level4">
<h4 id="vectorizing-code">Vectorizing code<a class="anchor" aria-label="anchor" href="#vectorizing-code"></a>
</h4>
<p>Before running the simulation, we are going to <em>vectorize</em> the
code. This means that instead of using <code>sapply</code> to run
<code>m</code> tests, we will create a matrix with all data in one call
to sample. This code runs several times faster than the code above,
which is necessary here due to the fact that we will be generating
several simulations. Understanding this chunk of code and how it is
equivalent to the code above using <code>sapply</code> will take a you
long way in helping you code efficiently in R.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">genefilter</span><span class="op">)</span> <span class="co">##rowttests is here</span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">##Define groups to be used with rowttests</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>, <span class="va">N</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>, <span class="va">N</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">1000</span> <span class="co">##number of simulations</span></span>
<span><span class="va">Qs</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="op">{</span></span>
<span>  <span class="co">##matrix with control data (rows are tests, columns are mice)</span></span>
<span>  <span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">*</span><span class="va">m</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">m</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">##matrix with control data (rows are tests, columns are mice)</span></span>
<span>  <span class="va">treatments</span> <span class="op">&lt;-</span>  <span class="fu">matrix</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">*</span><span class="va">m</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">m</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">##add effect to 10% of them</span></span>
<span>  <span class="va">treatments</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">)</span>,<span class="op">]</span> <span class="op">&lt;-</span> <span class="va">treatments</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">)</span>,<span class="op">]</span> <span class="op">+</span> <span class="va">delta</span></span>
<span>  </span>
<span>  <span class="co">##combine to form one matrix</span></span>
<span>  <span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">controls</span>, <span class="va">treatments</span><span class="op">)</span></span>
<span>  </span>
<span> <span class="va">calls</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">dat</span>, <span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span> <span class="op">&lt;</span> <span class="va">alpha</span></span>
<span> <span class="va">R</span><span class="op">=</span><span class="fu">sum</span><span class="op">(</span><span class="va">calls</span><span class="op">)</span></span>
<span> <span class="va">Q</span><span class="op">=</span><span class="fu">ifelse</span><span class="op">(</span><span class="va">R</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="fu">sum</span><span class="op">(</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span><span class="op">/</span><span class="va">R</span>, <span class="fl">0</span><span class="op">)</span></span>
<span> <span class="kw">return</span><span class="op">(</span><span class="va">Q</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="controlling-fdr">Controlling FDR<a class="anchor" aria-label="anchor" href="#controlling-fdr"></a>
</h4>
<p>The code above is a Monte Carlo simulation that generates 10,000
experiments 1,000 times, each time saving the observed <span class="math inline">\(Q\)</span>. Here is a histogram of these
values:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">Qs</span><span class="op">)</span> <span class="co">##Q is a random variable, this is its distribution</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/false-discovery-rate-rendered-Q_distribution-1.png" alt="Q (false positives divided by number of features called significant) is a random variable. Here we generated a distribution with a Monte Carlo simulation." class="figure mx-auto d-block"><figcaption>
Q (false positives divided by number of features called significant) is
a random variable. Here we generated a distribution with a Monte Carlo
simulation.
</figcaption></figure><p>The FDR is the average value of <span class="math inline">\(Q\)</span></p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">FDR</span><span class="op">=</span><span class="fu">mean</span><span class="op">(</span><span class="va">Qs</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">FDR</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.4465395</code></pre>
</div>
<p>The FDR is relatively high here. This is because for 90% of the
tests, the null hypotheses is true. This implies that with a 0.05
p-value cut-off, out of 100 tests we incorrectly call between 4 and 5
significant on average. This combined with the fact that we don’t
“catch” all the cases where the alternative is true, gives us a
relatively high FDR. So how can we control this? What if we want lower
FDR, say 5%?</p>
<p>To visually see why the FDR is high, we can make a histogram of the
p-values. We use a higher value of <code>m</code> to have more data from
the histogram. We draw a horizontal line representing the uniform
distribution one gets for the <code>m0</code> cases for which the null
is true.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">*</span><span class="va">m</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="va">treatments</span> <span class="op">&lt;-</span>  <span class="fu">matrix</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">*</span><span class="va">m</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="va">treatments</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">)</span>,<span class="op">]</span> <span class="op">&lt;-</span> <span class="va">treatments</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">)</span>,<span class="op">]</span> <span class="op">+</span> <span class="va">delta</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">controls</span>, <span class="va">treatments</span><span class="op">)</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">dat</span>, <span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span> </span>
<span></span>
<span><span class="va">h</span> <span class="op">&lt;-</span> <span class="fu">hist</span><span class="op">(</span><span class="va">pvals</span>, breaks<span class="op">=</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">polygon</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.05</span>,<span class="fl">0.05</span>,<span class="fl">0</span><span class="op">)</span>, <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="va">h</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="va">h</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>, col<span class="op">=</span><span class="st">"lightsteelblue1"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="va">m0</span><span class="op">/</span><span class="fl">20</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/false-discovery-rate-rendered-pval_hist-1.png" alt="Histogram of p-values. Monte Carlo simulation was used to generate data with m_1 genes having differences between groups." class="figure mx-auto d-block"><figcaption>
Histogram of p-values. Monte Carlo simulation was used to generate data
with m_1 genes having differences between groups.
</figcaption></figure><p>The first bar (grey) on the left represents cases with p-values
smaller than 0.05. From the horizontal line we can infer that about 1/2
are false positives. This is in agreement with an FDR of 0.50. If we
look at the bar for 0.01, we can see a lower FDR, as expected, but would
call fewer features significant.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h</span> <span class="op">&lt;-</span> <span class="fu">hist</span><span class="op">(</span><span class="va">pvals</span>,breaks<span class="op">=</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">polygon</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.01</span>,<span class="fl">0.01</span>,<span class="fl">0</span><span class="op">)</span>,<span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="va">h</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="va">h</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,col<span class="op">=</span><span class="st">"lightsteelblue1"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="va">m0</span><span class="op">/</span><span class="fl">100</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/false-discovery-rate-rendered-pval_hist2-1.png" alt="Histogram of p-values with breaks at every 0.01. Monte Carlo simulation was used to generate data with m_1 genes having differences between groups." class="figure mx-auto d-block"><figcaption>
Histogram of p-values with breaks at every 0.01. Monte Carlo simulation
was used to generate data with m_1 genes having differences between
groups.
</figcaption></figure><p>As we consider a lower and lower p-value cut-off, the number of
features detected decreases (loss of sensitivity), but our FDR also
decreases (gain of specificity). So how do we decide on this cut-off?
One approach is to set a desired FDR level <span class="math inline">\(\alpha\)</span>, and then develop procedures that
control the error rate: FDR <span class="math inline">\(\leq
\alpha\)</span>.</p>
</div>
<div class="section level4">
<h4 id="benjamini-hochberg-advanced">Benjamini-Hochberg (Advanced)<a class="anchor" aria-label="anchor" href="#benjamini-hochberg-advanced"></a>
</h4>
<p>We want to construct a procedure that guarantees the FDR to be below
a certain level <span class="math inline">\(\alpha\)</span>. For any
given <span class="math inline">\(\alpha\)</span>, the
Benjamini-Hochberg (1995) procedure is very practical because it simply
requires that we are able to compute p-values for each of the individual
tests and this permits a procedure to be defined.</p>
<p>For this procedure, order the p-values in increasing order: <span class="math inline">\(p_{(1)},\dots,p_{(m)}\)</span>. Then define <span class="math inline">\(k\)</span> to be the largest <span class="math inline">\(i\)</span> for which</p>
<p><span class="math display">\[p_{(i)} \leq
\frac{i}{m}\alpha\]</span></p>
<p>The procedure is to reject tests with p-values smaller or equal to
<span class="math inline">\(p_{(k)}\)</span>. Here is an example of how
we would select the <span class="math inline">\(k\)</span> with code
using the p-values computed above:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">i</span> <span class="op">=</span> <span class="fu">seq</span><span class="op">(</span>along<span class="op">=</span><span class="va">pvals</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">i</span>,<span class="fu">sort</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="va">i</span><span class="op">/</span><span class="va">m</span><span class="op">*</span><span class="va">alpha</span><span class="op">)</span></span>
<span><span class="co">##close-up</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">i</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span>,<span class="fu">sort</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span>,main<span class="op">=</span><span class="st">"Close-up"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="va">i</span><span class="op">/</span><span class="va">m</span><span class="op">*</span><span class="va">alpha</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/false-discovery-rate-rendered-pvalue_vs_rank_plot-1.png" alt="Plotting p-values plotted against their rank illustrates the Benjamini-Hochberg procedure. The plot on the right is a close-up of the plot on the left." class="figure mx-auto d-block"><figcaption>
Plotting p-values plotted against their rank illustrates the
Benjamini-Hochberg procedure. The plot on the right is a close-up of the
plot on the left.
</figcaption></figure><div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu">max</span><span class="op">(</span> <span class="fu">which</span><span class="op">(</span> <span class="fu">sort</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">i</span><span class="op">/</span><span class="va">m</span><span class="op">*</span><span class="va">alpha</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">cutoff</span> <span class="op">&lt;-</span> <span class="fu">sort</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span><span class="op">[</span><span class="va">k</span><span class="op">]</span></span>
<span><span class="fu">cat</span><span class="op">(</span><span class="st">"k ="</span>,<span class="va">k</span>,<span class="st">"p-value cutoff="</span>,<span class="va">cutoff</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>k = 24 p-value cutoff= 0.0001197627</code></pre>
</div>
<p>We can show mathematically that this procedure has FDR lower than 5%.
Please see Benjamini-Hochberg (1995) for details. An important outcome
is that we now have selected 11 tests instead of just 2. If we are
willing to set an FDR of 50% (this means we expect at least 1/2 our
genes to be hits), then this list grows to 1063. The FWER does not
provide this flexibility since any list of substantial size will result
in an FWER of 1.</p>
<p>Keep in mind that we don’t have to run the complicated code above as
we have functions to do this. For example, using the p-values
<code>pvals</code> computed above, we simply type the following:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fdr</span> <span class="op">&lt;-</span> <span class="fu">p.adjust</span><span class="op">(</span><span class="va">pvals</span>, method<span class="op">=</span><span class="st">"fdr"</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">pvals</span>,<span class="va">fdr</span>,log<span class="op">=</span><span class="st">"xy"</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="va">alpha</span>,v<span class="op">=</span><span class="va">cutoff</span><span class="op">)</span> <span class="co">##cutoff was computed above</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/false-discovery-rate-rendered-fdr-versus-pval-1.png" alt="FDR estimates plotted against p-value." class="figure mx-auto d-block"><figcaption>
FDR estimates plotted against p-value.
</figcaption></figure><p>We can run a Monte-Carlo simulation to confirm that the FDR is in
fact lower than .05. We compute all p-values first, and then use these
to decide which get called.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">1000</span> <span class="co">##number of simulations. We should increase for more precision</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>,<span class="op">{</span></span>
<span>  <span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">*</span><span class="va">m</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,nrow<span class="op">=</span><span class="va">m</span><span class="op">)</span></span>
<span>  <span class="va">treatments</span> <span class="op">&lt;-</span>  <span class="fu">matrix</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">*</span><span class="va">m</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,nrow<span class="op">=</span><span class="va">m</span><span class="op">)</span></span>
<span>  <span class="va">treatments</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">)</span>,<span class="op">]</span><span class="op">&lt;-</span><span class="va">treatments</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">)</span>,<span class="op">]</span><span class="op">+</span><span class="va">delta</span></span>
<span>  <span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">controls</span>,<span class="va">treatments</span><span class="op">)</span></span>
<span>  <span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">dat</span>,<span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span> </span>
<span>  <span class="co">##then the FDR</span></span>
<span>  <span class="va">calls</span> <span class="op">&lt;-</span> <span class="fu">p.adjust</span><span class="op">(</span><span class="va">pvals</span>,method<span class="op">=</span><span class="st">"fdr"</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">alpha</span></span>
<span>  <span class="va">R</span><span class="op">=</span><span class="fu">sum</span><span class="op">(</span><span class="va">calls</span><span class="op">)</span></span>
<span>  <span class="va">Q</span><span class="op">=</span><span class="fu">ifelse</span><span class="op">(</span><span class="va">R</span><span class="op">&gt;</span><span class="fl">0</span>,<span class="fu">sum</span><span class="op">(</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span><span class="op">/</span><span class="va">R</span>,<span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">R</span>,<span class="va">Q</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">Qs</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">Qs</span><span class="op">)</span> <span class="co">##Q is a random variable, this is its distribution</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/false-discovery-rate-rendered-Q_distribution2-1.png" alt="Histogram of Q (false positives divided by number of features called significant) when the alternative hypothesis is true for some features." class="figure mx-auto d-block"><figcaption>
Histogram of Q (false positives divided by number of features called
significant) when the alternative hypothesis is true for some features.
</figcaption></figure><div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">FDR</span><span class="op">=</span><span class="fu">mean</span><span class="op">(</span><span class="va">Qs</span><span class="op">)</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="va">FDR</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.03641142</code></pre>
</div>
<p>The FDR is lower than 0.05. This is to be expected because we need to
be conservative to ensure the FDR <span class="math inline">\(\leq\)</span> 0.05 for any value of <span class="math inline">\(m_0\)</span>, such as for the extreme case where
every hypothesis tested is null: <span class="math inline">\(m=m_0\)</span>. If you re-do the simulation above
for this case, you will find that the FDR increases.</p>
<p>We should also note that in …</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Rs</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">Rs</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.3</code></pre>
</div>
<p>… percent of the simulations, we did not call any genes
significant.</p>
<p>Finally, note that the <code>p.adjust</code> function has several
options for error rate controlling procedures:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p.adjust.methods</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "holm"       "hochberg"   "hommel"     "bonferroni" "BH"
[6] "BY"         "fdr"        "none"      </code></pre>
</div>
<p>It is important to remember that these options offer not just
different approaches to estimating error rates, but also that different
error rates are estimated: namely FWER and FDR. This is an important
distinction. More information is available from:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">p.adjust</span></span></code></pre>
</div>
<p>In summary, requiring that FDR <span class="math inline">\(\leq\)</span> 0.05 is a much more lenient
requirement FWER <span class="math inline">\(\leq\)</span> 0.05.
Although we will end up with more false positives, FDR gives us much
more power. This makes it particularly appropriate for discovery phase
experiments where we may accept FDR levels much higher than 0.05.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Restricting FWER too much can cause researchers to reject the null
hypothesis when it’s actually true. This is especially likely in the
small samples used in discovery phase experiments.</li>
<li>The Benjamini-Hochberg correction controls FDR by guaranteeing it to
be below a desired alpha level.</li>
<li>FDR is a more liberal correction than Bonferroni. While it generates
more false positives, it also provides more statistical power.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-direct-approach-false-discovery-rate"><p>Content from <a href="direct-approach-false-discovery-rate.html">Direct Approach to FDR and q-values</a></p>
<hr>
<p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/direct-approach-false-discovery-rate.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can you control false discovery rates when you don’t have an a
priori error rate?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Apply the Storey correction to identify true negatives.</li>
<li>Explain the difference between the Storey correction and
Benjamini-Hochberg approach.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="direct-approach-to-fdr-and-q-values-advanced">Direct Approach to FDR and q-values (Advanced)<a class="anchor" aria-label="anchor" href="#direct-approach-to-fdr-and-q-values-advanced"></a>
</h2>
<hr class="half-width">
<p>Here we review the results described by John D. Storey in J. R.
Statist. Soc. B (2002). One major distinction between Storey’s approach
and Benjamini and Hochberg’s is that we are no longer going to set a
<span class="math inline">\(\alpha\)</span> level a priori. Because in
many high-throughput experiments we are interested in obtaining some
list for validation, we can instead decide beforehand that we will
consider all tests with p-values smaller than 0.01. We then want to
attach an estimate of an error rate. Using this approach, we are
guaranteed to have <span class="math inline">\(R&gt;0\)</span>. Note
that in the FDR definition above we assigned <span class="math inline">\(Q=0\)</span> in the case that <span class="math inline">\(R=V=0\)</span>. We were therefore computing:</p>
<p><span class="math display">\[
\mbox{FDR} = E\left( \frac{V}{R} \mid R&gt;0\right) \mbox{Pr}(R&gt;0)
\]</span></p>
<p>In the approach proposed by Storey, we condition on having a
non-empty list, which implies <span class="math inline">\(R&gt;0\)</span>, and we instead compute the
<em>positive FDR</em></p>
<p><span class="math display">\[
\mbox{pFDR} = E\left( \frac{V}{R} \mid R&gt;0\right)
\]</span></p>
<p>A second distinction is that while Benjamini and Hochberg’s procedure
controls under the worst case scenario, in which all null hypotheses are
true ( <span class="math inline">\(m=m_0\)</span> ), Storey proposes
that we actually try to estimate <span class="math inline">\(m_0\)</span> from the data. Because in
high-throughput experiments we have so much data, this is certainly
possible. The general idea is to pick a relatively high value p-value
cut-off, call it <span class="math inline">\(\lambda\)</span>, and
assume that tests obtaining p-values &gt; <span class="math inline">\(\lambda\)</span> are mostly from cases in which
the null hypothesis holds. We can then estimate <span class="math inline">\(\pi_0 = m_0/m\)</span> as:</p>
<p><span class="math display">\[
\hat{\pi}_0 = \frac{\#\left\{p_i &gt; \lambda \right\} }{ (1-\lambda) m
}
\]</span></p>
<p>There are more sophisticated procedures than this, but they follow
the same general idea. Here is an example setting <span class="math inline">\(\lambda=0.1\)</span>. Using the p-values computed
above we have:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">genefilter</span><span class="op">)</span> <span class="co">##rowttests is here</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span></span>
<span><span class="co">##Define groups to be used with rowttests</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>, <span class="va">N</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>, <span class="va">N</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># re-create p-values from earlier if needed</span></span>
<span><span class="va">population</span> <span class="op">&lt;-</span> <span class="fu">unlist</span><span class="op">(</span><span class="fu">read.csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"./data/femaleControlsPopulation.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="fl">0.90</span></span>
<span><span class="va">m0</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">*</span><span class="va">p0</span></span>
<span><span class="va">m1</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">-</span><span class="va">m0</span></span>
<span><span class="va">nullHypothesis</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span> <span class="fu">rep</span><span class="op">(</span><span class="cn">TRUE</span>,<span class="va">m0</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="cn">FALSE</span>,<span class="va">m1</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">delta</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">1000</span> <span class="co">##number of simulations. We should increase for more precision </span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="op">{</span></span>
<span>  <span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">*</span><span class="va">m</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,nrow<span class="op">=</span><span class="va">m</span><span class="op">)</span></span>
<span>  <span class="va">treatments</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">*</span><span class="va">m</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,nrow<span class="op">=</span><span class="va">m</span><span class="op">)</span></span>
<span>  <span class="va">treatments</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">)</span>,<span class="op">]</span><span class="op">&lt;-</span><span class="va">treatments</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">)</span>,<span class="op">]</span><span class="op">+</span><span class="va">delta</span> </span>
<span>  <span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">controls</span>,<span class="va">treatments</span><span class="op">)</span></span>
<span>  <span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">dat</span>,<span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>  <span class="co">##then the FDR</span></span>
<span>  <span class="va">calls</span> <span class="op">&lt;-</span> <span class="fu">p.adjust</span><span class="op">(</span><span class="va">pvals</span>,method<span class="op">=</span><span class="st">"fdr"</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">alpha</span></span>
<span>  <span class="va">R</span><span class="op">=</span><span class="fu">sum</span><span class="op">(</span><span class="va">calls</span><span class="op">)</span></span>
<span>  <span class="va">Q</span><span class="op">=</span><span class="fu">ifelse</span><span class="op">(</span><span class="va">R</span><span class="op">&gt;</span><span class="fl">0</span>,<span class="fu">sum</span><span class="op">(</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span><span class="op">/</span><span class="va">R</span>,<span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">R</span>,<span class="va">Q</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">Qs</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">*</span><span class="va">m</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,nrow<span class="op">=</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="va">treatments</span> <span class="op">&lt;-</span>  <span class="fu">matrix</span><span class="op">(</span><span class="fu">sample</span><span class="op">(</span><span class="va">population</span>, <span class="va">N</span><span class="op">*</span><span class="va">m</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,nrow<span class="op">=</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="va">treatments</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">)</span>,<span class="op">]</span><span class="op">&lt;-</span><span class="va">treatments</span><span class="op">[</span><span class="fu">which</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span><span class="op">)</span>,<span class="op">]</span><span class="op">+</span><span class="va">delta</span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">controls</span>,<span class="va">treatments</span><span class="op">)</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">dat</span>,<span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span> </span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">pvals</span>,breaks<span class="op">=</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">0.05</span><span class="op">)</span>,freq<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">lambda</span> <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="va">pi0</span><span class="op">=</span><span class="fu">sum</span><span class="op">(</span><span class="va">pvals</span><span class="op">&gt;</span> <span class="va">lambda</span><span class="op">)</span> <span class="op">/</span><span class="op">(</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">lambda</span><span class="op">)</span><span class="op">*</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span> <span class="va">pi0</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/direct-approach-false-discovery-rate-rendered-pi0_estimate-1.png" alt="p-value histogram with pi0 estimate." class="figure mx-auto d-block"><figcaption>
p-value histogram with pi0 estimate.
</figcaption></figure><div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">print</span><span class="op">(</span><span class="va">pi0</span><span class="op">)</span> <span class="co">##this is close to the try pi0=0.9</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.9326667</code></pre>
</div>
<p>With this estimate in place we can, for example, alter the Benjamini
and Hochberg procedures to select the <span class="math inline">\(k\)</span> to be the largest value so that:</p>
<p><span class="math display">\[\hat{\pi}_0 p_{(i)} \leq
\frac{i}{m}\alpha\]</span></p>
<p>However, instead of doing this, we compute a <em>q-value</em> for
each test. If a feature resulted in a p-value of <span class="math inline">\(p\)</span>, the q-value is the estimated pFDR for
a list of all the features with a p-value at least as small as <span class="math inline">\(p\)</span>.</p>
<p>In R, this can be computed with the <code>qvalue</code> function in
the <code>qvalue</code> package:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">qvalue</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">qvalue</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span></span>
<span><span class="va">qvals</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">qvalues</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">pvals</span>, <span class="va">qvals</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/direct-approach-false-discovery-rate-rendered-qval_vs_pval-1.png" alt="q-values versus p-values." class="figure mx-auto d-block"><figcaption>
q-values versus p-values.
</figcaption></figure><p>we also obtain the estimate of <span class="math inline">\(\hat{\pi}_0\)</span>:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span><span class="op">$</span><span class="va">pi0</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.9117251</code></pre>
</div>
<p>This function uses a more sophisticated approach at estimating <span class="math inline">\(\pi_0\)</span> than what is described above.</p>
<div class="section level4">
<h4 id="note-on-estimating-pi_0">Note on estimating <span class="math inline">\(\pi_0\)</span>
<a class="anchor" aria-label="anchor" href="#note-on-estimating-pi_0"></a>
</h4>
<p>In our experience the estimation of <span class="math inline">\(\pi_0\)</span> can be unstable and adds a step of
uncertainty to the data analysis pipeline. Although more conservative,
the Benjamini-Hochberg procedure is computationally more stable.</p>
</div>
</section><section><h2 class="section-heading" id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<hr class="half-width">
<p>In the following exercises, we will define error controlling
procedures for experimental data. We will make a list of genes based on
q-values. We will also assess your understanding of false positives
rates and false negative rates by asking you to create a Monte Carlo
simulation.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>Load the gene expression data:<br><code>load("episodes/data/GSE5859Subset.rda")</code><br>
We are interested in comparing gene expression between the two groups
defined in the sampleInfo table. Compute a p-value for each gene using
the function <code>rowttests</code> from the genefilter package.</p>
<p><code>library(genefilter)</code><br><code>?rowttests</code></p>
<p>How many genes have p-values smaller than 0.05?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p><code>g &lt;- sampleInfo$group</code><br><code>pvals &lt;- rowttests(geneExpression, factor(g))$p.value</code><br><code>sum(pvals &lt; 0.05)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-2-apply-the-bonferroni-correction-to-achieve-a-fwer-of-0.05.-how-many-genes-are-called-significant-under-this-procedure" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2-apply-the-bonferroni-correction-to-achieve-a-fwer-of-0.05.-how-many-genes-are-called-significant-under-this-procedure" class="callout-inner">
<h3 class="callout-title">Exercise 2: Apply the Bonferroni correction to
achieve a FWER of 0.05. How many genes are called significant under this
procedure?</h3>
<div class="callout-content">

</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p><code>m &lt;- 8793</code><br><code>sum(pvals &lt; (0.05/m))</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>The FDR is a property of a list of features, not each specific
feature. The q-value relates FDR to individual features. To define the
q-value, we order features we tested by p-value, then compute the FDRs
for a list with the most significant, the two most significant, the
three most significant, etc. The FDR of the list with the, say, m most
significant tests is defined as the q-value of the m-th most significant
feature. In other words, the q-value of a feature, is the FDR of the
biggest list that includes that gene. In R, we can compute q-values
using the p.adjust function with the FDR option. Read the help file for
p.adjust and compute how many genes achieve a q-value &lt; 0.05 for our
gene expression dataset.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p><code>pvals_adjust &lt;- p.adjust(pvals, method = 'fdr')</code><br><code>sum(pvals_adjust &lt; 0.05)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>Now use the qvalue function, in the Bioconductor qvalue package, to
estimate q-values using the procedure described by Storey. How many
genes have q-values below 0.05?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p><code>res &lt;- qvalue(pvals)</code><br><code>sum(res$qvalues &lt; 0.05)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-5" class="callout-inner">
<h3 class="callout-title">Exercise 5</h3>
<div class="callout-content">
<p>Read the help file for qvalue and report the estimated proportion of
genes for which the null hypothesis is true π0 = m0/m</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<p><code>res$pi0</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-6" class="callout-inner">
<h3 class="callout-title">Exercise 6</h3>
<div class="callout-content">
<p>The number of genes passing the q-value &lt; 0.05 threshold is larger
with the q-value function than the p.adjust difference. Why is this the
case? Make a plot of the ratio of these two estimates to help answer the
question.<br>
A) One of the two procedures is flawed.<br>
B) The two functions are estimating different things.<br>
C) The qvalue function estimates the proportion of genes for which the
null hypothesis is true and provides a less conservative estimate.<br>
D) The qvalue function estimates the proportion of genes for which the
null hypothesis is true and provides a more conservative estimate.</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<p><code>plot(pvals_adjust, res$qvalues, xlab = 'fdr', ylab = 'qval')</code><br><code>abline(0,1)</code><br>
The qvalue function estimates the proportion of genes for which the null
hypothesis is true and provides a less conservative estimate (choice
C).</p>
</div>
</div>
</div>
</div>
<div id="exercise-7" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-7" class="callout-inner">
<h3 class="callout-title">Exercise 7</h3>
<div class="callout-content">
<p>This exercise and the remaining ones are more advanced. Create a
Monte Carlo Simulation in which you simulate measurements from 8,793
genes for 24 samples, 12 cases and 12 controls. Then for 100 genes
create a difference of 1 between cases and controls. You can use the
code provided below. Run this experiment 1,000 times with a Monte Carlo
simulation. For each instance, compute p-values using a t-test and keep
track of the number of false positives and false negatives. Compute the
false positive rate and false negative rates if we use Bonferroni,
q-values from p.adjust, and q-values from qvalue function. Set the seed
to 1 for all three simulations. What is the false positive rate for
Bonferroni? What are the false negative rates for Bonferroni?</p>
<pre><code><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">24</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">8793</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">n</span><span class="op">*</span><span class="va">m</span><span class="op">)</span>,<span class="va">m</span>,<span class="va">n</span><span class="op">)</span></span>
<span><span class="va">delta</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">positives</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">positives</span>,<span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="va">n</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">positives</span>,<span class="fl">1</span><span class="op">:</span><span class="op">(</span><span class="va">n</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span><span class="op">+</span><span class="va">delta</span>  </span></code></pre>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<pre><code><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">12</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">8793</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span><span class="va">m1</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">m0</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">-</span><span class="va">m1</span></span>
<span><span class="va">nullHypothesis</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="cn">TRUE</span>,<span class="va">m0</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="cn">FALSE</span>,<span class="va">m1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">delta</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="op">{</span></span>
<span>  <span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">N</span><span class="op">*</span><span class="va">m</span><span class="op">)</span>,nrow <span class="op">=</span> <span class="va">m</span>, ncol <span class="op">=</span> <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">N</span><span class="op">*</span><span class="va">m</span><span class="op">)</span>,nrow <span class="op">=</span> <span class="va">m</span>, ncol <span class="op">=</span> <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span><span class="op">[</span><span class="op">!</span><span class="va">nullHypothesis</span>,<span class="op">]</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">treatment</span><span class="op">[</span><span class="op">!</span><span class="va">nullHypothesis</span>,<span class="op">]</span> <span class="op">+</span> <span class="va">delta</span></span>
<span>  <span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">controls</span>, <span class="va">treatment</span><span class="op">)</span></span>
<span>  <span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">dat</span>, <span class="fu">factor</span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>  <span class="va">calls</span> <span class="op">&lt;-</span> <span class="va">pvals</span> <span class="op">&lt;</span> <span class="op">(</span><span class="fl">0.05</span><span class="op">/</span><span class="va">m</span><span class="op">)</span></span>
<span>  <span class="va">R</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">calls</span><span class="op">)</span></span>
<span>  <span class="va">V</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span></span>
<span>  <span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span><span class="op">/</span><span class="va">m0</span> <span class="co"># false positive</span></span>
<span>  <span class="va">fn</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">calls</span><span class="op">)</span><span class="op">/</span><span class="va">m1</span> <span class="co"># false negative</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">fp</span>,<span class="va">fn</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">res</span><span class="op">&lt;-</span><span class="fu">t</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="co"># false positive rate</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="co"># false negative rate</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="exercise-8" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-8" class="callout-inner">
<h3 class="callout-title">Exercise 8</h3>
<div class="callout-content">
<p>What are the false positive rates for p.adjust?<br>
What are the false negative rates for p.adjust?</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<pre><code><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">12</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">8793</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span><span class="va">m1</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">m0</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">-</span><span class="va">m1</span></span>
<span><span class="va">nullHypothesis</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="cn">TRUE</span>,<span class="va">m0</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="cn">FALSE</span>,<span class="va">m1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">delta</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="op">{</span></span>
<span>  <span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">N</span><span class="op">*</span><span class="va">m</span><span class="op">)</span>,nrow <span class="op">=</span> <span class="va">m</span>, ncol <span class="op">=</span> <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">N</span><span class="op">*</span><span class="va">m</span><span class="op">)</span>,nrow <span class="op">=</span> <span class="va">m</span>, ncol <span class="op">=</span> <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span><span class="op">[</span><span class="op">!</span><span class="va">nullHypothesis</span>,<span class="op">]</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">treatment</span><span class="op">[</span><span class="op">!</span><span class="va">nullHypothesis</span>,<span class="op">]</span> <span class="op">+</span> <span class="va">delta</span></span>
<span>  <span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">controls</span>, <span class="va">treatment</span><span class="op">)</span></span>
<span>  <span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">dat</span>, <span class="fu">factor</span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>  <span class="va">pvals_adjust</span> <span class="op">&lt;-</span> <span class="fu">p.adjust</span><span class="op">(</span><span class="va">pvals</span>, method <span class="op">=</span> <span class="st">'fdr'</span><span class="op">)</span></span>
<span>  <span class="va">calls</span> <span class="op">&lt;-</span> <span class="va">pvals_adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span></span>
<span>  <span class="va">R</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">calls</span><span class="op">)</span></span>
<span>  <span class="va">V</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span></span>
<span>  <span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span><span class="op">/</span><span class="va">m0</span> <span class="co"># false positive</span></span>
<span>  <span class="va">fn</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">calls</span><span class="op">)</span><span class="op">/</span><span class="va">m1</span> <span class="co"># false negative</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">fp</span>,<span class="va">fn</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">t</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="co"># false positive rate</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="co"># false negative rate</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="exercise-9" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-9" class="callout-inner">
<h3 class="callout-title">Exercise 9</h3>
<div class="callout-content">
<p>What are the false positive rates for qvalues?<br>
What are the false negative rates for qvalues?</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" aria-labelledby="headingSolution9" data-bs-parent="#accordionSolution9">
<div class="accordion-body">
<pre><code><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">12</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">8793</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span><span class="va">m1</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">m0</span> <span class="op">&lt;-</span> <span class="va">m</span><span class="op">-</span><span class="va">m1</span></span>
<span><span class="va">nullHypothesis</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="cn">TRUE</span>,<span class="va">m0</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="cn">FALSE</span>,<span class="va">m1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">delta</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="op">{</span></span>
<span>  <span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">N</span><span class="op">*</span><span class="va">m</span><span class="op">)</span>,nrow <span class="op">=</span> <span class="va">m</span>, ncol <span class="op">=</span> <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">N</span><span class="op">*</span><span class="va">m</span><span class="op">)</span>,nrow <span class="op">=</span> <span class="va">m</span>, ncol <span class="op">=</span> <span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">treatment</span><span class="op">[</span><span class="op">!</span><span class="va">nullHypothesis</span>,<span class="op">]</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">treatment</span><span class="op">[</span><span class="op">!</span><span class="va">nullHypothesis</span>,<span class="op">]</span> <span class="op">+</span> <span class="va">delta</span></span>
<span>  <span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">controls</span>, <span class="va">treatment</span><span class="op">)</span></span>
<span>  <span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">dat</span>, <span class="fu">factor</span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>  <span class="va">qvals</span> <span class="op">&lt;-</span> <span class="fu">qvalue</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span><span class="op">$</span><span class="va">qvalue</span></span>
<span>  <span class="va">calls</span> <span class="op">&lt;-</span> <span class="va">qvals</span> <span class="op">&lt;</span> <span class="fl">0.05</span></span>
<span>  <span class="va">R</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">calls</span><span class="op">)</span></span>
<span>  <span class="va">V</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span></span>
<span>  <span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="va">calls</span><span class="op">)</span><span class="op">/</span><span class="va">m0</span> <span class="co"># false positive</span></span>
<span>  <span class="va">fn</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="op">!</span><span class="va">nullHypothesis</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">calls</span><span class="op">)</span><span class="op">/</span><span class="va">m1</span> <span class="co"># false negative</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">fp</span>,<span class="va">fn</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">t</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="co"># false positive rate</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="co"># false negative rate  </span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>The Storey correction makes different assumptions that
Benjamini-Hochberg. It does not set a priori alpha levels, but instead
estimates the number of true null hypotheses from a given data set.</li>
<li>The Storey correction is less computationally stable than
Benjamini-Hochberg.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-basic-exploratory-data-analysis"><p>Content from <a href="basic-exploratory-data-analysis.html">Basic EDA for high-throughput data</a></p>
<hr>
<p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/basic-exploratory-data-analysis.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 75 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What problems can different kinds of exploratory plots detect in
high-throughput data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Create four different types of plots for high-throughput data.</li>
<li>Explain how to use them to detect problems.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="basic-exploratory-data-analysis">Basic Exploratory Data Analysis<a class="anchor" aria-label="anchor" href="#basic-exploratory-data-analysis"></a>
</h2>
<hr class="half-width">
<p>An under-appreciated advantage of working with high-throughput data
is that problems with the data are sometimes more easily exposed than
with low-throughput data. The fact that we have thousands of
measurements permits us to see problems that are not apparent when only
a few measurements are available. A powerful way to detect these
problems is with exploratory data analysis (EDA). Here we review some of
the plots that allow us to detect quality problems.</p>
<div class="section level4">
<h4 id="volcano-plots">Volcano plots<a class="anchor" aria-label="anchor" href="#volcano-plots"></a>
</h4>
<p>Here we will use the results obtained from applying t-test to data
from a gene expression dataset:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">genefilter</span><span class="op">)</span></span>
<span><span class="fu">load</span><span class="op">(</span><span class="st">"data/GSE5859Subset.rda"</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">geneExpression</span>, <span class="va">g</span><span class="op">)</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">$</span><span class="va">p.value</span></span></code></pre>
</div>
<p>And we also generate p-values from a dataset for which we know the
null is true:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span></span>
<span><span class="va">randomData</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">n</span> <span class="op">*</span> <span class="va">m</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span></span>
<span><span class="va">nullpvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">randomData</span>, <span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span></code></pre>
</div>
<p>As we described earlier, reporting only p-values is a mistake when we
can also report effect sizes. With high-throughput data, we can
visualize the results by making a <em>volcano plot</em>. The idea behind
a volcano plot is to show these for all features. In the y-axis we plot
-log (base 10) p-values and on the x-axis we plot the effect size. By
using -log (base 10), the “highly significant” features appear at the
top of the plot. Using log also permits us to better distinguish between
small and very small p-values, for example 0.01 and <span class="math inline">\(10^6\)</span>. Here is the volcano plot for our
results above:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">dm</span>,<span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span>,</span>
<span>     xlab<span class="op">=</span><span class="st">"Effect size"</span>,ylab<span class="op">=</span><span class="st">"- log (base 10) p-values"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/basic-exploratory-data-analysis-rendered-volcano_plot-1.png" class="figure mx-auto d-block"></figure><p>Many features with very small p-values, but small effect sizes as we
see here, are sometimes indicative of problematic data.</p>
</div>
<div class="section level4">
<h4 id="p-value-histograms">p-value Histograms<a class="anchor" aria-label="anchor" href="#p-value-histograms"></a>
</h4>
<p>Another plot we can create to get an overall idea of the results is
to make histograms of p-values. When we generate completely null data
the histogram follows a uniform distribution. With our original dataset
we see a higher frequency of smaller p-values.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">nullpvals</span>,ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1400</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">pvals</span>,ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1400</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/basic-exploratory-data-analysis-rendered-pval-hist-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>When we expect most hypotheses to be null and don’t see a uniform
p-value distribution, it might be indicative of unexpected properties,
such as correlated samples.</p>
<p>If we permute the outcomes and calculate p-values then, if the
samples are independent, we should see a uniform distribution. With
these data we do not:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">permg</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="va">permresults</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">geneExpression</span>,<span class="va">permg</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">permresults</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/basic-exploratory-data-analysis-rendered-pval-hist2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In a later chapter we will see that the columns in this dataset are
not independent and thus the assumptions used to compute the p-values
here are incorrect.</p>
</div>
<div class="section level4">
<h4 id="data-boxplots-and-histograms">Data boxplots and histograms<a class="anchor" aria-label="anchor" href="#data-boxplots-and-histograms"></a>
</h4>
<p>With high-throughput data, we have thousands of measurements for each
experimental unit. As mentioned earlier, this can help us detect quality
issues. For example, if one sample has a completely different
distribution than the rest, we might suspect there are problems.
Although a complete change in distribution could be due to real
biological differences, more often than not it is due to a technical
problem. Here we load a large gene expression experiment available on
Github. We “accidentally” use log instead of log2 on one of the
samples.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">Biobase</span><span class="op">)</span></span>
<span><span class="fu">load</span><span class="op">(</span><span class="st">"data/GSE5859.rda"</span><span class="op">)</span></span>
<span><span class="va">ge</span> <span class="op">&lt;-</span> <span class="fu">exprs</span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="co">##ge for gene expression</span></span>
<span><span class="va">ge</span><span class="op">[</span>,<span class="fl">49</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">ge</span><span class="op">[</span>,<span class="fl">49</span><span class="op">]</span><span class="op">/</span><span class="fu">log2</span><span class="op">(</span><span class="fu">exp</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="co">##imitate error</span></span></code></pre>
</div>
<p>A quick look at a summary of the distribution using boxplots
immediately highlights the mistake:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">boxplot</span><span class="op">(</span><span class="va">ge</span>,range<span class="op">=</span><span class="fl">0</span>,names<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu">ncol</span><span class="op">(</span><span class="va">e</span><span class="op">)</span>,col<span class="op">=</span><span class="fu">ifelse</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">ncol</span><span class="op">(</span><span class="va">ge</span><span class="op">)</span><span class="op">==</span><span class="fl">49</span>,<span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/basic-exploratory-data-analysis-rendered-boxplots-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Note that the number of samples is a bit too large here, making it
hard to see the boxes. One can instead simply show the boxplot summaries
without the boxes:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qs</span> <span class="op">&lt;-</span> <span class="fu">t</span><span class="op">(</span><span class="fu">apply</span><span class="op">(</span><span class="va">ge</span>,<span class="fl">2</span>,<span class="va">quantile</span>,prob<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0.05</span>,<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="fl">0.75</span>,<span class="fl">0.95</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">matplot</span><span class="op">(</span><span class="va">qs</span>,type<span class="op">=</span><span class="st">"l"</span>,lty<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/basic-exploratory-data-analysis-rendered-kaboxplot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We refer to this figure as a <em>kaboxplot</em> because Karl Broman
was the first we saw use it as an alternative to boxplots.</p>
<p>We can also plot all the histograms. Because we have so much data, we
create histograms using small bins, then smooth the heights of the bars
and then plot <em>smooth histograms</em>. We re-calibrate the height of
these smooth curves so that if a bar is made with base of size “unit”
and height given by the curve at <span class="math inline">\(x_0\)</span>, the area approximates the number of
points in region of size “unit” centered at <span class="math inline">\(x_0\)</span>:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">shist</span><span class="op">(</span><span class="va">ge</span>,unit<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/basic-exploratory-data-analysis-rendered-shist-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
<div class="section level4">
<h4 id="ma-plot">MA plot<a class="anchor" aria-label="anchor" href="#ma-plot"></a>
</h4>
<p>Scatterplots and correlation are not the best tools to detect
replication problems. A better measure of replication can be obtained
from examining the differences between the values that should be the
same. Therefore, a better plot is a rotation of the scatterplot
containing the differences on the y-axis and the averages on the x-axis.
This plot was originally named a Bland-Altman plot, but in genomics it
is commonly referred to as an MA-plot. The name MA comes from plots of
red log intensity minus (M) green intensities versus average (A) log
intensities used with microarrays (MA) data.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">ge</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">ge</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="op">(</span><span class="va">x</span> <span class="op">+</span> <span class="va">y</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, <span class="va">x</span> <span class="op">-</span> <span class="va">y</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/basic-exploratory-data-analysis-rendered-maplot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Note that once we rotate the plot, the fact that these data have
differences of about:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sd</span><span class="op">(</span><span class="va">y</span> <span class="op">-</span> <span class="va">x</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.2025465</code></pre>
</div>
<p>becomes immediate. The scatterplot shows very strong correlation,
which is not necessarily informative here.</p>
<p>We will later introduce dendograms, heatmaps, and multi-dimensional
scaling plots.</p>
</div>
</section><section><h2 class="section-heading" id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<hr class="half-width">
<p>We will be using a handful of Bioconductor packages. These are
installed using the function biocLite which you can source from the web:
source(“<a href="http://www.bioconductor.org/biocLite.R" class="external-link uri">http://www.bioconductor.org/biocLite.R</a>”) or you can run
the bioc_install in the rafalib package.<br><code>library(rafalib)</code><br><code>bioc_install()</code></p>
<p>Download and install the Bioconductor package SpikeInSubset and then
load the library and the mas133 data:<br><code>library(rafalib)</code>
<code>install_bioc("SpikeInSubset")</code></p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">SpikeInSubset</span><span class="op">)</span>  </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: affy</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data</span><span class="op">(</span><span class="va">mas133</span><span class="op">)</span></span></code></pre>
</div>
<p>Now make the following plot of the first two samples and compute the
correlation:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">e</span> <span class="op">&lt;-</span> <span class="fu">exprs</span><span class="op">(</span><span class="va">mas133</span><span class="op">)</span>  </span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">e</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,<span class="va">e</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>,main<span class="op">=</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"corr="</span>,<span class="fu">signif</span><span class="op">(</span><span class="fu">cor</span><span class="op">(</span><span class="va">e</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,<span class="va">e</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span>,cex<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span>  </span>
<span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fl">3000</span>  </span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fl">1000</span> <span class="co">#a buffer   </span></span>
<span><span class="fu">polygon</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="va">b</span>,<span class="va">k</span>,<span class="va">k</span>,<span class="op">-</span><span class="va">b</span><span class="op">)</span>,<span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="va">b</span>,<span class="op">-</span><span class="va">b</span>,<span class="va">k</span>,<span class="va">k</span><span class="op">)</span>,col<span class="op">=</span><span class="st">"red"</span>,density<span class="op">=</span><span class="fl">0</span>,border<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/basic-exploratory-data-analysis-rendered-corr_two_samples-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>What proportion of the points are inside the box?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p><code>length(which(e[,1] &lt;= 3000 &amp; e[,2]&lt;= 3000)) / dim(e)[1]</code><br><code>sum(e[,1] &lt;= 3000 &amp; e[,2] &lt;= 3000) / dim(e)[1]</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>Now make the sample plot with log:<br><code>plot(log2(e[,1]),log2(e[,2]))</code><br><code>k &lt;- log2(3000)</code><br><code>b &lt;- log2(0.5)</code><br><code>polygon(c(b,k,k,b),c(b,b,k,k),col="red",density=0,border="red")</code><br>
What is an advantage of taking the log?<br>
A) The tails do not dominate the plot: 95% of data is not in a tiny
section of plot.<br>
B) There are less points.<br>
C) There is exponential growth.<br>
D) We always take logs.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>The answer choice is A: The tails do not dominate the plot, 95% of
data are not in a tiny section of the plot.</p>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>Make an MA-plot:<br><code>e &lt;- log2(exprs(mas133))</code><br><code>plot((e[,1]+e[,2])/2,e[,2]-e[,1],cex=0.5)</code><br>
The two samples we are plotting are replicates (they are random samples
from the same batch of RNA). The correlation of the data was 0.997 in
the original scale and 0.96 in the log-scale. High correlations are
sometimes confused with evidence of replication. However, replication
implies we get very small differences between the observations, which is
better measured with distance or differences.<br>
What is the standard deviation of the log ratios for this
comparison?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p><code>e &lt;- log2(exprs(mas133))</code><br><code>plot((e[,1]+e[,2])/2,e[,2]-e[,1],cex=0.5)</code><br><code>sd(e[,2]-e[,1])</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>How many fold changes above 2 do we see?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p><code>sum(abs(e[,2]-e[,1])1)</code></p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>While it is tempting to jump straight in to inferential analyses,
it’s very important to run EDA first. Visualizing high-throughput data
with EDA enables researchers to detect biological and technical issues
with data. Plots can show at a glance where errors lie, making
inferential analyses more accurate, efficient, and replicable.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-principal-component-analysis"><p>Content from <a href="principal-component-analysis.html">Principal Components Analysis</a></p>
<hr>
<p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/principal-component-analysis.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can researchers simplify or streamline EDA in high-throughput
data sets?</li>
<li>What is principal component analysis (PCA) and when can it be
used?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the purpose of dimension reduction.</li>
<li>Define a principal component.</li>
<li>Perform a principal components analysis.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="dimension-reduction-motivation">Dimension Reduction Motivation<a class="anchor" aria-label="anchor" href="#dimension-reduction-motivation"></a>
</h2>
<hr class="half-width">
<p>Visualizing data is one of the most, if not the most, important step
in the analysis of high-throughput data. The right visualization method
may reveal problems with the experimental data that can render the
results from a standard analysis, although typically appropriate,
completely useless.</p>
<p>Recall that in recent history biologists went from using their eyes
or simple summaries to categorize results, to having thousands (and now
millions) of measurements per sample to analyze. Here we will focus on
statistical inference in the context of high-throughput measurements,
also called high-dimensional data. High-dimensional data is “wide” data,
rather than long data, which has few variables relative to the number of
observations. The number of features (variables) in high-dimensional
data is much greater than the number of observations, making simple
visualizations like scatterplots cumbersome.</p>
<p>We have shown methods for visualizing global properties of the
columns or rows, but plots that reveal relationships between columns or
between rows are more complicated due to the high dimensionality of
data. For example, to compare each of the 189 samples to each other, we
would have to create, for example, 17,766 MA-plots. Creating one single
scatterplot of the data is impossible since points are very high
dimensional.</p>
<p>We will describe powerful techniques for exploratory data analysis
based on <em>dimension reduction</em>. The general idea is to reduce the
dataset to have fewer dimensions, yet approximately preserve important
properties, such as the distance between samples. If we are able to
reduce down to, say, two dimensions, we can then easily make plots.</p>
<p>Principal component analysis (PCA) is a popular method of analyzing
high-dimensional data. Large datasets of correlated variables can be
summarized into smaller numbers of uncorrelated principal components
that explain most of the variability in the original dataset. An example
of PCA might be reducing several variables representing aspects of
patient health (blood pressure, heart rate, respiratory rate) into a
single feature.</p>
<p>PCA is a useful exploratory analysis tool. PCA allows us to reduce a
large number of variables into a few features which represent most of
the variation in the original variables. This makes exploration of the
original variables easier.</p>
<p>The first principal component (<span class="math inline">\(Z_1\)</span>) is calculated using the
equation:</p>
<p><span class="math display">\[
  Z_1 = a_{11}X_1 + a_{21}X_2 +....+a_{p1}X_p
\]</span></p>
<p><span class="math inline">\(X_1...X_p\)</span> represents variables
in the original dataset and <span class="math inline">\(a_{11}...a_p\)</span> represent principal
component loadings, which can be thought of as the degree to which each
variable contributes to the calculation of the principal component.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>Descriptions of three datasets and research questions are given
below. For which of these might PCA be considered a useful tool for
analyzing data so that the research questions may be addressed?</p>
<ol style="list-style-type: decimal">
<li>An epidemiologist has data collected from different patients
admitted to hospital with infectious respiratory disease. They would
like to determine whether length of stay in hospital differs in patients
with different respiratory diseases.</li>
<li>An online retailer has collected data on user interactions with its
online app and has information on the number of times each user
interacted with the app, what products they viewed per interaction, and
the type and cost of these products. The retailer would like to use this
information to predict whether or not a user will be interested in a new
product.</li>
<li>A scientist has assayed gene expression levels in 1000 cancer
patients and has data from probes targeting different genes in tumour
samples from patients. She would like to create new variables
representing relative abundance of different groups of genes to i) find
out if genes form subgroups based on biological function and ii) use
these new variables in a linear regression examining how gene expression
varies with disease severity.</li>
<li>All of the above.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>In the first case, a regression model would be more suitable; perhaps
a survival model. In the second, again a regression model, likely linear
or logistic, would be more suitable. In the third example, PCA can help
to identify modules of correlated features that explain a large amount
of variation within the data.</p>
<p>Therefore the answer here is 3.</p>
</div>
</div>
</div>
</div>
</section><div class="section level1">
<h1 id="what-is-a-principal-component">What is a principal component?<a class="anchor" aria-label="anchor" href="#what-is-a-principal-component"></a>
</h1>
<p>The first principal component is the direction of the data along
which the observations vary the most. The second principal component is
the direction of the data along which the observations show the next
highest amount of variation. The second principal component is a linear
combination of the variables that is uncorrelated with the first
principal component. There are as many principal components as there are
variables in your dataset, but as we’ll see, some are more useful at
explaining your data than others. By definition, the first principal
component explains more variation than other principal components.</p>
<p>The animation below illustrates how principal components are
calculated from data. You can imagine that the black line is a rod and
each red dashed line is a spring. The energy of each spring is
proportional to its squared length. The direction of the first principal
component is the one that minimizes the total energy of all of the
springs. In the animation below, the springs pull the rod, finding the
direction of the first principal component when they reach equilibrium.
We then use the length of the springs from the rod as the first
principal component. This is explained in more detail on <a href="https://stats.stackexchange.com/questions/2691/making-sense-of-principal-component-analysis-eigenvectors-eigenvalues" class="external-link">this
Q&amp;A website</a>.</p>
<figure><img src=".././fig/pendulum.gif" class="figure mx-auto d-block"></figure><div class="section level4">
<h4 id="example-reducing-two-dimensions-to-one">Example: Reducing two dimensions to one<a class="anchor" aria-label="anchor" href="#example-reducing-two-dimensions-to-one"></a>
</h4>
<p>We consider an example with twin heights. Here we simulate 100 two
dimensional points that represent the number of standard deviations each
individual is from the mean height. Each point is a pair of twins:</p>
<figure style="text-align: center"><img src="../fig/principal-component-analysis-rendered-simulate_twin_heights-1.png" alt="Simulated twin pair heights." class="figure mx-auto d-block"><figcaption>
Simulated twin pair heights.
</figcaption></figure><p>To help with the illustration, think of this as high-throughput gene
expression data with the twin pairs representing the <span class="math inline">\(N\)</span> samples and the two heights
representing gene expression from two genes.</p>
<p>We are interested in the distance between any two samples. We can
compute this using <code>dist</code>. For example, here is the distance
between the two orange points in the figure above:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span><span class="op">=</span><span class="fu">dist</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">as.matrix</span><span class="op">(</span><span class="va">d</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1.140897</code></pre>
</div>
<p>What if making two dimensional plots was too complex and we were only
able to make 1 dimensional plots. Can we, for example, reduce the data
to a one dimensional matrix that preserves distances between points?</p>
<p>If we look back at the plot, and visualize a line between any pair of
points, the length of this line is the distance between the two points.
These lines tend to go along the direction of the diagonal. We have seen
before that we can “rotate” the plot so that the diagonal is in the
x-axis by making a MA-plot instead:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z1</span> <span class="op">=</span> <span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span><span class="op">+</span><span class="va">y</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span> <span class="co">#the sum </span></span>
<span><span class="va">z2</span> <span class="op">=</span> <span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span><span class="op">-</span><span class="va">y</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span>   <span class="co">#the difference</span></span>
<span></span>
<span><span class="va">z</span> <span class="op">=</span> <span class="fu">rbind</span><span class="op">(</span> <span class="va">z1</span>, <span class="va">z2</span><span class="op">)</span> <span class="co">#matrix now same dimensions as y</span></span>
<span></span>
<span><span class="va">thelim</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>,<span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>,<span class="va">y</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>,xlab<span class="op">=</span><span class="st">"Twin 1 (standardized height)"</span>,</span>
<span>     ylab<span class="op">=</span><span class="st">"Twin 2 (standardized height)"</span>,</span>
<span>     xlim<span class="op">=</span><span class="va">thelim</span>,ylim<span class="op">=</span><span class="va">thelim</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,<span class="va">y</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,col<span class="op">=</span><span class="fl">2</span>,pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">z</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>,<span class="va">z</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>,xlim<span class="op">=</span><span class="va">thelim</span>,ylim<span class="op">=</span><span class="va">thelim</span>,xlab<span class="op">=</span><span class="st">"Average height"</span>,ylab<span class="op">=</span><span class="st">"Difference in height"</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="va">z</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,<span class="va">z</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,col<span class="op">=</span><span class="fl">2</span>,pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/principal-component-analysis-rendered-rotation-1.png" alt="Twin height scatterplot (left) and MA-plot (right)." class="figure mx-auto d-block"><figcaption>
Twin height scatterplot (left) and MA-plot (right).
</figcaption></figure><p>How do we perform a PCA?</p>
</div>
<div class="section level2">
<h2 id="a-prostate-cancer-dataset">A prostate cancer dataset<a class="anchor" aria-label="anchor" href="#a-prostate-cancer-dataset"></a>
</h2>
<p>The <code>Prostate</code> dataset is freely available online and
represents data from 97 men who have prostate cancer. The data come from
a study which examined the correlation between the level of prostate
specific antigen and a number of clinical measures in men who were about
to receive a radical prostatectomy. The data have 97 rows and 9
columns.</p>
<p>Columns include: - <code>lcavol</code> (log-transformed cancer
volume), - <code>lweight</code> (log-transformed prostate weight), -
<code>lbph</code> (log-transformed amount of benign prostate
enlargement), - <code>svi</code> (seminal vesicle invasion), -
<code>lcp</code> (log-transformed capsular penetration; amount of spread
of cancer in outer walls of prostate), - <code>gleason</code> (Gleason
score; grade of cancer cells), - <code>pgg45</code> (percentage Gleason
scores 4 or 5), - <code>lpsa</code> (log-tranformed prostate specific
antigen; level of PSA in blood). - <code>age</code> (patient age in
years).</p>
<p>Here we will calculate principal component scores for each of the
rows in this dataset, using five principal components (one for each
variable included in the PCA). We will include five clinical variables
in our PCA, each of the continuous variables in the prostate dataset, so
that we can create fewer variables representing clinical markers of
cancer progression. Standard PCAs are carried out using continuous
variables only.</p>
<p>First, we will examine the <code>Prostate</code> dataset.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">Brq</span><span class="op">)</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="st">"Prostate"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">Prostate</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      lcavol  lweight age      lbph svi       lcp gleason pgg45       lpsa
1 -0.5798185 2.769459  50 -1.386294   0 -1.386294       6     0 -0.4307829
2 -0.9942523 3.319626  58 -1.386294   0 -1.386294       6     0 -0.1625189
3 -0.5108256 2.691243  74 -1.386294   0 -1.386294       7    20 -0.1625189
4 -1.2039728 3.282789  58 -1.386294   0 -1.386294       6     0 -0.1625189
5  0.7514161 3.432373  62 -1.386294   0 -1.386294       6     0  0.3715636
6 -1.0498221 3.228826  50 -1.386294   0 -1.386294       6     0  0.7654678</code></pre>
</div>
<p>Note that each row of the dataset represents a single patient.</p>
<p>We will create a subset of the data including only the clinical
variables we want to use in the PCA.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pros2</span> <span class="op">&lt;-</span> <span class="va">Prostate</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"lcavol"</span>, <span class="st">"lweight"</span>, <span class="st">"lbph"</span>, <span class="st">"lcp"</span>, <span class="st">"lpsa"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">pros2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      lcavol  lweight      lbph       lcp       lpsa
1 -0.5798185 2.769459 -1.386294 -1.386294 -0.4307829
2 -0.9942523 3.319626 -1.386294 -1.386294 -0.1625189
3 -0.5108256 2.691243 -1.386294 -1.386294 -0.1625189
4 -1.2039728 3.282789 -1.386294 -1.386294 -0.1625189
5  0.7514161 3.432373 -1.386294 -1.386294  0.3715636
6 -1.0498221 3.228826 -1.386294 -1.386294  0.7654678</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="do-we-need-to-standardise-the-data">Do we need to standardise the data?<a class="anchor" aria-label="anchor" href="#do-we-need-to-standardise-the-data"></a>
</h2>
<p>Now we compare the variances between variables in the dataset.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">apply</span><span class="op">(</span><span class="va">pros2</span>, <span class="fl">2</span>, <span class="va">var</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  lcavol  lweight     lbph      lcp     lpsa
1.389157 0.246642 2.104840 1.955102 1.332476 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">par</span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">pros2</span><span class="op">$</span><span class="va">lweight</span>, breaks <span class="op">=</span> <span class="st">"FD"</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">pros2</span><span class="op">$</span><span class="va">lbph</span>, breaks <span class="op">=</span> <span class="st">"FD"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/principal-component-analysis-rendered-var-hist-1.png" alt="Histograms comparing variances between log-transformed prostate weight and benign enlargement." class="figure mx-auto d-block"><figcaption>
Histograms comparing variances between log-transformed prostate weight
and benign enlargement.
</figcaption></figure><p>Note that variance is greatest for <code>lbph</code> and lowest for
<code>lweight</code>. It is clear from this output that we need to scale
each of these variables before including them in a PCA analysis to
ensure that differences in variances between variables do not drive the
calculation of principal components. In this example we standardise all
five variables to have a mean of 0 and a standard deviation of 1.</p>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>Why might it be necessary to standardise variables before performing
a PCA?<br>
Can you think of datasets where it might not be necessary to standardise
variables? Discuss.</p>
<ol style="list-style-type: decimal">
<li>To make the results of the PCA interesting.</li>
<li>If you want to ensure that variables with different ranges of values
contribute equally to analysis.</li>
<li>To allow the feature matrix to be calculated faster, especially in
cases where there are a lot of input variables.</li>
<li>To allow both continuous and categorical variables to be included in
the PCA.</li>
<li>All of the above.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<ol start="2" style="list-style-type: decimal">
<li>Scaling the data isn’t guaranteed to make the results more
interesting. It also won’t affect how quickly the output will be
calculated, whether continuous and categorical variables are present or
not.</li>
</ol>
<p>It is done to ensure that all features have equal weighting in the
resulting PCs.</p>
<p>You may not want to standardise datasets which contain continuous
variables all measured on the same scale (e.g. gene expression data or
RNA sequencing data). In this case, variables with very little
sample-to-sample variability may represent only random noise, and
standardising the data would give these extra weight in the PCA.</p>
</div>
</div>
</div>
</div>
<p>Next we will carry out a PCA using the <code>prcomp()</code> function
in base R. The input data (<code>pros2</code>) is in the form of a
matrix. Note that the <code>scale = TRUE</code> argument is used to
standardise the variables to have a mean 0 and standard deviation of
1.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca.pros</span> <span class="op">&lt;-</span> <span class="fu">prcomp</span><span class="op">(</span><span class="va">pros2</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, center <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pca.pros</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Standard deviations (1, .., p=5):
[1] 1.5648756 1.1684678 0.7452990 0.6362941 0.4748755

Rotation (n x k) = (5 x 5):
              PC1         PC2         PC3         PC4         PC5
lcavol  0.5616465 -0.23664270  0.01486043  0.22708502 -0.75945046
lweight 0.2985223  0.60174151 -0.66320198 -0.32126853 -0.07577123
lbph    0.1681278  0.69638466  0.69313753  0.04517286 -0.06558369
lcp     0.4962203 -0.31092357  0.26309227 -0.72394666  0.25253840
lpsa    0.5665123 -0.01680231 -0.10141557  0.56487128  0.59111493</code></pre>
</div>
</div>
</div>
<div class="section level1">
<h1 id="how-many-principal-components-do-we-need">How many principal components do we need?<a class="anchor" aria-label="anchor" href="#how-many-principal-components-do-we-need"></a>
</h1>
<p>We have calculated one principal component for each variable in the
original dataset. How do we choose how many of these are necessary to
represent the true variation in the data, without having extra
components that are unnecessary?</p>
<p>Let’s look at the relative importance of each component using
<code>summary</code>.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Importance of components:
                          PC1    PC2    PC3     PC4    PC5
Standard deviation     1.5649 1.1685 0.7453 0.63629 0.4749
Proportion of Variance 0.4898 0.2731 0.1111 0.08097 0.0451
Cumulative Proportion  0.4898 0.7628 0.8739 0.95490 1.0000</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get proportions of variance explained by each PC (rounded to 2 DP)</span></span>
<span><span class="va">prop.var</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="fu">summary</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">)</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span><span class="st">"Proportion of Variance"</span>, <span class="op">]</span>, <span class="fl">2</span><span class="op">)</span> <span class="op">*</span></span>
<span>  <span class="fl">100</span></span></code></pre>
</div>
<p>This returns the proportion of variance in the data explained by each
of the (p = 5) principal components. In this example, PC1 explains
approximately 49% of variance in the data, PC2 27% of variance, PC3 a
further 11%, PC4 approximately 8% and PC5 around 5%.</p>
<p>We can use a screeplot to see how much variation in the data is
explained by each principal component. Let’s calculate the screeplot for
our PCA.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate variance explained</span></span>
<span><span class="va">varExp</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">pca.pros</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">/</span> <span class="fu">sum</span><span class="op">(</span><span class="va">pca.pros</span><span class="op">$</span><span class="va">sdev</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span><span class="co"># calculate percentage variance explained using output from the PCA</span></span>
<span><span class="va">varDF</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>Dimensions <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu">length</span><span class="op">(</span><span class="va">varExp</span><span class="op">)</span>, varExp <span class="op">=</span> <span class="va">varExp</span><span class="op">)</span></span>
<span><span class="co"># create new dataframe with five rows, one for each principal component</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">varDF</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/principal-component-analysis-rendered-vardf-plot-1.png" alt="Screeplot showing proportion of variance explained by each principal component." class="figure mx-auto d-block"><figcaption>
Screeplot showing proportion of variance explained by each principal
component.
</figcaption></figure><p>The screeplot shows that the first principal component explains most
of the variance in the data (&gt;50%) and each subsequent principal
component explains less and less of the total variance. The first two
principal components explain &gt;70% of variance in the data. But what
do these two principal components mean?</p>
<div class="section level2">
<h2 id="what-are-loadings-and-principal-component-scores">What are loadings and principal component scores?<a class="anchor" aria-label="anchor" href="#what-are-loadings-and-principal-component-scores"></a>
</h2>
<p>The output from a PCA returns a matrix of principal component
loadings in which the columns show the principal component loading
vectors. Note that the square of values in each column sums to 1 as each
loading is scaled so as to prevent a blow up in variance. Larger values
in the columns suggest a greater contribution of that variable to the
principal component.</p>
<p>We can examine the output of our PCA by writing the following in
<code>R</code>:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca.pros</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Standard deviations (1, .., p=5):
[1] 1.5648756 1.1684678 0.7452990 0.6362941 0.4748755

Rotation (n x k) = (5 x 5):
              PC1         PC2         PC3         PC4         PC5
lcavol  0.5616465 -0.23664270  0.01486043  0.22708502 -0.75945046
lweight 0.2985223  0.60174151 -0.66320198 -0.32126853 -0.07577123
lbph    0.1681278  0.69638466  0.69313753  0.04517286 -0.06558369
lcp     0.4962203 -0.31092357  0.26309227 -0.72394666  0.25253840
lpsa    0.5665123 -0.01680231 -0.10141557  0.56487128  0.59111493</code></pre>
</div>
<p>For each row in the original dataset PCA returns a principal
component score for each of the principal components (PC1 to PC5 in the
<code>Prostate</code> data example). We can see how the principal
component score (<span class="math inline">\(Z_{i1}\)</span> for rows
<span class="math inline">\(i\)</span> to <span class="math inline">\(n\)</span>) is calculated for the first principal
component using the following equation from Figure 1:</p>
<p><span class="math display">\[
  Z_{i1} = a_1 \times (lcavol_i - \overline{lcavol}) + a_2 \times
(lpsa_i - \overline{lpsa})
\]</span></p>
<p><span class="math inline">\(a_1\)</span> and <span class="math inline">\(a_2\)</span> represent principal component
loadings in this equation. A loading can be thought of as the ‘weight’
each variable has on the calculation of the principal component. Note
that in our example using the <code>Prostate</code> dataset
<code>lcavol</code> (log-transformed cancer volume) and
<code>lpsa</code> (log-tranformed prostate specific antigen in blood)
are the variables that contribute most to the first principal
component.</p>
<p>We can better understand what the principal components represent in
terms of the original variables by plotting the first two principal
components against each other and labelling points by patient number.
Clusters of points which have similar principal component scores can be
observed using a biplot and the strength and direction of influence
different variables have on the calculation of the principal component
scores can be observed by plotting arrows representing the loadings onto
the graph. A biplot of the first two principal components can be created
as follows:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">biplot</span><span class="op">(</span><span class="va">pca.pros</span>, xlim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">0.3</span>, <span class="fl">0.3</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/principal-component-analysis-rendered-stats-biplot-1.png" alt="Biplot of first two principal components." class="figure mx-auto d-block"><figcaption>
Biplot of first two principal components.
</figcaption></figure><figure><img src=".././fig/stats-biplot.png" alt="Biplot of first two principal components." class="figure mx-auto d-block"><div class="figcaption">Biplot of first two principal components.</div>
</figure><p>This biplot shows the position of each patient on a 2-dimensional
plot where loadings can be observed via the red arrows associated with
each of the variables. The variables <code>lpsa</code>,
<code>lcavol</code> and <code>lcp</code> are associated with positive
values on PC1 while positive values on PC2 are associated with the
variables <code>lbph</code> and <code>lweight</code>. The length of the
arrows indicates how much each variable contributes to the calculation
of each principal component.</p>
<p>The left and bottom axes show normalised principal component scores.
The axes on the top and right of the plot are used to interpret the
loadings, where loadings are scaled by the standard deviation of the
principal components (<code>pca.pros$sdev</code>) times the square root
the number of observations.</p>
</div>
</div>
<div class="section level1">
<h1 id="using-pca-to-analyse-gene-expression-data">Using PCA to analyse gene expression data<a class="anchor" aria-label="anchor" href="#using-pca-to-analyse-gene-expression-data"></a>
</h1>
<p>In this section you will carry out your own PCA using the
Bioconductor package <strong><code>PCAtools</code></strong> applied to
gene expression data to explore the topics covered above.
<strong><code>PCAtools</code></strong> provides functions that can be
used to explore data via PCA and produce useful figures and analysis
tools.</p>
<div class="section level2">
<h2 id="a-gene-expression-dataset-of-cancer-patients">A gene expression dataset of cancer patients<a class="anchor" aria-label="anchor" href="#a-gene-expression-dataset-of-cancer-patients"></a>
</h2>
<p>The dataset we will be analysing in this lesson includes two subsets
of data: * a matrix of gene expression data showing microarray results
for different probes used to examine gene expression profiles in 91
different breast cancer patient samples. * metadata associated with the
gene expression results detailing information from patients from whom
samples were taken.</p>
<p>Let’s load the <strong><code>PCAtools</code></strong> package and the
data.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"PCAtools"</span><span class="op">)</span></span></code></pre>
</div>
<p>We will first load the microarray breast cancer gene expression data
and associated metadata, downloaded from the <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE2990" class="external-link">Gene
Expression Omnibus</a>.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"SummarizedExperiment"</span><span class="op">)</span></span>
<span><span class="va">cancer</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"./data/cancer_expression.rds"</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">cancer</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">cancer</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">View</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="co">#nrow=22215 probes</span></span>
<span><span class="co">#ncol=91 samples</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">View</span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#nrow=91</span></span>
<span><span class="co">#ncol=8</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">all</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">==</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] TRUE</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Check that column names and row names match</span></span>
<span><span class="co">#If they do should return TRUE</span></span></code></pre>
</div>
<p>The ‘mat’ variable contains a matrix of gene expression profiles for
each sample. Rows represent gene expression measurements and columns
represent samples. The ‘metadata’ variable contains the metadata
associated with the gene expression data including the name of the study
from which data originate, the age of the patient from which the sample
was taken, whether or not an oestrogen receptor was involved in their
cancer and the grade and size of the cancer for each sample (represented
by rows).</p>
<p>Microarray data are difficult to analyse for several reasons.
Firstly, they are typically high-dimensional and therefore are subject
to the same difficulties associated with analysing high dimensional data
outlined above (i.e. <em>p</em>&gt;<em>n</em>, large numbers of rows,
multiple possible response variables, curse of dimensionality).
Secondly, formulating a research question using microarray data can be
difficult, especially if not much is known a priori about which genes
code for particular phenotypes of interest. Finally, exploratory
analysis, which can be used to help formulate research questions and
display relationships, is difficult using microarray data due to the
number of potentially interesting response variables (i.e. expression
data from probes targeting different genes).</p>
<p>If researchers hypothesise that groups of genes (e.g. biological
pathways) may be associated with different phenotypic characteristics of
cancers (e.g. histologic grade, tumour size), using statistical methods
that reduce the number of columns in the microarray matrix to a smaller
number of dimensions representing groups of genes would help visualise
the data and address research questions regarding the effect different
groups of genes have on disease progression.</p>
<p>Using the <strong><code>PCAtools</code></strong> we will apply a PCA
to the cancer gene expression data, plot the amount of variation in the
data explained by each principal component and plot the most important
principal components against each other as well as understanding what
each principal component represents.</p>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>Apply a PCA to the cancer gene expression data using the
<code>pca()</code> function from <strong><code>PCAtools</code></strong>.
You can use the help files in PCAtools to find out about the
<code>pca()</code> function (type <code>help("pca")</code> or
<code>?pca</code> in R).</p>
<p>Remove the lower 20% of principal components from your PCA using the
<code>removeVar</code> argument in the <code>pca()</code> function. As
in the example using prostate data above, examine the first 5 rows and
columns of rotated data and loadings from your PCA.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<pre><code><span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu">pca</span><span class="op">(</span><span class="va">mat</span>, metadata <span class="op">=</span> <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#Many PCs explain a very small amount of the total variance in the data</span></span>
<span><span class="co">#Remove the lower 20% of PCs with lower variance</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu">pca</span><span class="op">(</span><span class="va">mat</span>, metadata <span class="op">=</span> <span class="va">metadata</span>, removeVar <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="co">#Explore other arguments provided in pca</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">rotated</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">which.max</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span><span class="fl">49</span>, <span class="op">]</span></span>
<span></span>
<span><span class="fu">which.max</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">pc</span><span class="op">$</span><span class="va">loadings</span><span class="op">[</span><span class="fl">27</span>, <span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>The function <code>pca()</code> is used to perform PCA, and uses as
inputs a matrix (<code>mat</code>) containing continuous numerical data
in which rows are data variables and columns are samples, and
<code>metadata</code> associated with the matrix in which rows represent
samples and columns represent data variables. It has options to centre
or scale the input data before a PCA is performed, although in this case
gene expression data do not need to be transformed prior to PCA being
carried out as variables are measured on a similar scale (values are
comparable between rows). The output of the <code>pca()</code> function
includes a lot of information such as loading values for each variable
(<code>loadings</code>), principal component scores
(<code>rotated</code>) and the amount of variance in the data explained
by each principal component.</p>
<p>Rotated data shows principal component scores for each sample and
each principal component. Loadings the contribution each variable makes
to each principal component.</p>
</div>
<div class="section level2">
<h2 id="scaling-variables-for-pca">Scaling variables for PCA<a class="anchor" aria-label="anchor" href="#scaling-variables-for-pca"></a>
</h2>
<p>When running <code>pca()</code> above, we kept the default setting,
<code>scale=FALSE</code>. That means genes with higher variation in
their expression levels should have higher loadings, which is what we
are interested in. Whether or not to scale variables for PCA will depend
on your data and research question.</p>
<p>Note that this is different from normalising gene expression data.
Gene expression data have to be normalised before donwstream analyses
can be carried out. This is to reduce to effect technical and other
potentially confounding factors. We assume that the expression data we
use had been normalised previously.</p>
</div>
<div class="section level2">
<h2 id="choosing-how-many-components-are-important-to-explain-the-variance-in-the-data">Choosing how many components are important to explain the variance
in the data<a class="anchor" aria-label="anchor" href="#choosing-how-many-components-are-important-to-explain-the-variance-in-the-data"></a>
</h2>
<p>As in the example using the <code>Prostate</code> dataset we can use
a screeplot to compare the proportion of variance in the data explained
by each principal component. This allows us to understand how much
information in the microarray dataset is lost by projecting the
observations onto the first few principal components and whether these
principal components represent a reasonable amount of the variation. The
proportion of variance explained should sum to one.</p>
<p>There are no clear guidelines on how many principal components should
be included in PCA: your choice depends on the total variability of the
data and the size of the dataset. We often look at the ‘elbow’ on the
screeplot as an indicator that the addition of principal components does
not drastically contribute to explain the remaining variance or choose
an arbitory cut off for proportion of variance explained.</p>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>Using the <code>screeplot()</code> function in
<strong><code>PCAtools</code></strong>, create a screeplot to show
proportion of variance explained by each principal component. Explain
the output of the screeplot in terms of proportion of variance in data
explained by each principal component.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<pre><code><span><span class="fu">screeplot</span><span class="op">(</span><span class="va">pc</span>, axisLabSize <span class="op">=</span> <span class="fl">5</span>, titleLabSize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span></code></pre>
<p>Note that first principal component (PC1) explains more variation
than other principal components (which is always the case in PCA). The
screeplot shows that the first principal component only explains ~33% of
the total variation in the micrarray data and many principal components
explain very little variation. The red line shows the cumulative
percentage of explained variation with increasing principal components.
Note that in this case 18 principal components are needed to explain
over 75% of variation in the data. This is not an unusual result for
complex biological datasets including genetic information as clear
relationships between groups are sometimes difficult to observe in the
data. The screeplot shows that using a PCA we have reduced 91 predictors
to 18 in order to explain a significant amount of variation in the data.
See additional arguments in screeplot function for improving the
appearance of the plot.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="investigating-the-principal-components">Investigating the principal components<a class="anchor" aria-label="anchor" href="#investigating-the-principal-components"></a>
</h2>
<p>Once the most important principal components have been identified
using <code>screeplot()</code>, these can be explored in more detail by
plotting principal components against each other and highlighting points
based on variables in the metadata. This will allow any potential
clustering of points according to demographic or phenotypic variables to
be seen.</p>
<p>We can use biplots to look for patterns in the output from the PCA.
Note that there are two functions called <code>biplot()</code>, one in
the package <strong><code>PCAtools</code></strong> and one in
<strong><code>stats</code></strong>. Both functions produce biplots but
their scales are different!</p>
<div id="exercise-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-5" class="callout-inner">
<h3 class="callout-title">Exercise 5</h3>
<div class="callout-content">
<p>Create a biplot of the first two principal components from your PCA
(using <code>biplot()</code> function in
<strong><code>PCAtools</code></strong> - see
<code>help("PCAtools::biplot")</code> for arguments) and examine whether
the data appear to form clusters. Explain your results.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<pre><code><span><span class="fu">biplot</span><span class="op">(</span><span class="va">pc</span>, lab <span class="op">=</span> <span class="cn">NULL</span>, colby <span class="op">=</span> <span class="st">'Grade'</span>, legendPosition <span class="op">=</span> <span class="st">'top'</span><span class="op">)</span></span></code></pre>
<p>The biplot shows the position of patient samples relative to PC1 and
PC2 in a 2-dimensional plot. Note that two groups are apparent along the
PC1 axis according to expressions of different genes while no separation
can be seem along the PC2 axis. Labels of patient samples are
automatically added in the biplot. Labels for each sample are added by
default, but can be removed if there is too much overlap in names. Note
that <strong><code>PCAtools</code></strong> does not scale biplot in the
same way as biplot using the stats package.</p>
</div>
</div>
</div>
</div>
<p>Let’s consider this biplot in more detail, and also display the
loadings:</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">biplot</span><span class="op">(</span><span class="va">pc</span>, lab <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">metadata</span><span class="op">)</span>, pointSize <span class="op">=</span> <span class="fl">1</span>, labSize <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pcaobj$rotated: object of type 'closure' is not subsettable</code></pre>
</div>
<figure><img src=".././fig/pca-biplot.png" alt="Biplot of principal components one and two showing two groups for PC1 according to gene expression." class="figure mx-auto d-block"><div class="figcaption">Biplot of principal components one and two
showing two groups for PC1 according to gene expression.</div>
</figure><p>Sizes of labels, points and axes can be changed using arguments in
<code>biplot</code> (see <code>help("biplot")</code>). We can see from
the biplot that there appear to be two separate groups of points that
separate on the PC1 axis, but that no other grouping is apparent on
other PC axes.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotloadings</span><span class="op">(</span><span class="va">pc</span>, labSize <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pcaobj$loadings: object of type 'closure' is not subsettable</code></pre>
</div>
<figure><img src=".././fig/pca-loadings.png" alt="Plot of principal component loadings show magnitude and direction of gene probes." class="figure mx-auto d-block"><div class="figcaption">Plot of principal component loadings show
magnitude and direction of gene probes.</div>
</figure><p>Plotting the loadings shows the magnitude and direction of loadings
for probes detecting genes on each principal component.</p>
<div id="exercise-6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-6" class="callout-inner">
<h3 class="callout-title">Exercise 6</h3>
<div class="callout-content">
<p>Use <code>colby</code> and <code>lab</code> arguments in
<code>biplot()</code> to explore whether these two groups may cluster by
patient age or by whether or not the sample expresses the oestrogen
receptor gene (ER+ or ER-).</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<pre><code><span>  <span class="fu">biplot</span><span class="op">(</span><span class="va">pc</span>,</span>
<span>    lab <span class="op">=</span> <span class="fu">paste0</span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">metadata</span><span class="op">$</span><span class="va">Age</span>,<span class="st">'years'</span><span class="op">)</span>,</span>
<span>    colby <span class="op">=</span> <span class="st">'ER'</span>,</span>
<span>    hline <span class="op">=</span> <span class="fl">0</span>, vline <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    legendPosition <span class="op">=</span> <span class="st">'right'</span><span class="op">)</span></span></code></pre>
<p>{: .language-r} It appears that one cluster has more ER+ samples than
the other group.</p>
</div>
</div>
</div>
</div>
<p>So far we have only looked at a biplot of PC1 versus PC2 which only
gives part of the picture. The <code>pairplots()</code> function in
<strong><code>PCAtools</code></strong> can be used to create multiple
biplots including different principal components.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pairsplot</span><span class="op">(</span><span class="va">pc</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pcaobj$components: object of type 'closure' is not subsettable</code></pre>
</div>
<figure><img src=".././fig/pairsplot.png" alt="Pairsplot of principal components show clusters on PC1 only." class="figure mx-auto d-block"><div class="figcaption">Pairsplot of principal components show clusters
on PC1 only.</div>
</figure><p>The plots show two apparent clusters involving the first principal
component only. No other clusters are found involving other principal
components.</p>
</div>
</div>
<div class="section level1">
<h1 id="using-pca-output-in-further-analysis">Using PCA output in further analysis<a class="anchor" aria-label="anchor" href="#using-pca-output-in-further-analysis"></a>
</h1>
<p>The output of PCA can be used to interpret data or can be used in
further analyses. For example, the PCA outputs new variables (principal
components) which represent several variables in the original dataset.
These new variables are useful for further exploring data, for example,
comparing principal component scores between groups or including the new
variables in linear regressions. Because the principal components are
uncorrelated (and independent) they can be included together in a single
linear regression.</p>
<div class="section level2">
<h2 id="principal-component-regression">Principal component regression<a class="anchor" aria-label="anchor" href="#principal-component-regression"></a>
</h2>
<p>PCA is often used to reduce large numbers of correlated variables
into fewer uncorrelated variables that can then be included in linear
regression or other models. This technique is called principal component
regression (PCR) and it allows researchers to examine the effect of
several correlated explanatory variables on a single response variable
in cases where a high degree of correlation initially prevents them from
being included in the same model. This is called principal componenet
regression (PCR) and is just one example of how principal components can
be used in further analysis of data. When carrying out PCR, the variable
of interest (response/dependent variable) is regressed against the
principal components calculated using PCA, rather than against each
individual explanatory variable from the original dataset. As there as
many principal components created from PCA as there are variables in the
dataset, we must select which principal components to include in PCR.
This can be done by examining the amount of variation in the data
explained by each principal component (see above).</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Visualizing data with thousands or tens of thousands of measurements
is impossible using standard techniques.</li>
<li>Dimension reduction techniques coupled with visualization can reveal
relationships between dimensions (rows or columns) in the data.</li>
<li>Principal components analysis is a dimension reduction technique
that can reduce and summarize large datasets.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</div></section><section id="aio-statistical-models"><p>Content from <a href="statistical-models.html">Statistical Models</a></p>
<hr>
<p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/statistical-models.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are some of the most widely used parametric distributions used
in the life sciences besides the normal?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Define the Poisson distribution.</li>
<li>Explain its relevance in RNA sequence analysis.</li>
<li>Explain the purpose of maximum likelihood estimation.</li>
<li>Generate some research questions that maximum likelihood estimation
can address.</li>
<li>Model distributions of gene variances.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="statistical-models">Statistical Models<a class="anchor" aria-label="anchor" href="#statistical-models"></a>
</h1>
<blockquote>
<p>All models are wrong, but some are useful. -George E. P. Box</p>
</blockquote>
<p>When we see a p-value in the literature, it means a probability
distribution of some sort was used to quantify the null hypothesis. Many
times deciding which probability distribution to use is relatively
straightforward. For example, in the tea tasting challenge we can use
simple probability calculations to determine the null distribution. Most
p-values in the scientific literature are based on sample averages or
least squares estimates from a linear model and make use of the CLT to
approximate the null distribution of their statistic as normal.</p>
<p>The CLT is backed by theoretical results that guarantee that the
approximation is accurate. However, we cannot always use this
approximation, such as when our sample size is too small. Previously, we
described how the sample average can be approximated as t-distributed
when the population data is approximately normal. However, there is no
theoretical backing for this assumption. We are now <em>modeling</em>.
In the case of height, we know from experience that this turns out to be
a very good model.</p>
<p>But this does not imply that every dataset we collect will follow a
normal distribution. Some examples are: coin tosses, the number of
people who win the lottery, and US incomes. The normal distribution is
not the only parametric distribution that is available for modeling.
Here we provide a very brief introduction to some of the most widely
used parametric distributions and some of their uses in the life
sciences. We focus on the models and concepts needed to understand the
techniques currently used to perform statistical inference on
high-throughput data. To do this we also need to introduce the basics of
Bayesian statistics. For more in-depth description of probability models
and parametric distributions please consult a Statistics textbook such
as <a href="https://www.stat.berkeley.edu/~rice/Book3ed/index.html" class="external-link">this
one</a>.</p>
<div class="section level2">
<h2 id="the-binomial-distribution">The Binomial Distribution<a class="anchor" aria-label="anchor" href="#the-binomial-distribution"></a>
</h2>
<p>The first distribution we will describe is the binomial distribution.
It reports the probability of observing <span class="math inline">\(S=k\)</span> successes in <span class="math inline">\(N\)</span> trials as</p>
<p><span class="math display">\[
\mbox{Pr}(S=k) = {N \choose k}p^k (1-p)^{N-k}
\]</span></p>
<p>with <span class="math inline">\(p\)</span> the probability of
success. The best known example is coin tosses with <span class="math inline">\(S\)</span> the number of heads when tossing <span class="math inline">\(N\)</span> coins. In this example <span class="math inline">\(p=0.5\)</span>.</p>
<p>Note that <span class="math inline">\(S/N\)</span> is the average of
independent random variables and thus the CLT tells us that <span class="math inline">\(S\)</span> is approximately normal when <span class="math inline">\(N\)</span> is large. This distribution has many
applications in the life sciences. Recently, it has been used by the
variant callers and genotypers applied to next generation sequencing. A
special case of this distribution is approximated by the Poisson
distribution which we describe next.</p>
</div>
<div class="section level2">
<h2 id="the-poisson-distribution">The Poisson Distribution<a class="anchor" aria-label="anchor" href="#the-poisson-distribution"></a>
</h2>
<p>Since it is the sum of binary outcomes, the number of people that win
the lottery follows a binomial distribution (we assume each person buys
one ticket). The number of trials <span class="math inline">\(N\)</span>
is the number of people that buy tickets and is usually very large.
However, the number of people that win the lottery oscillates between 0
and 3, which implies the normal approximation does not hold. So why does
CLT not hold? One can explain this mathematically, but the intuition is
that with the sum of successes so close to and also constrained to be
larger than 0, it is impossible for the distribution to be normal. Here
is a quick simulation:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">=</span><span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">7</span> <span class="co">##1 in 10,000,0000 chances of winning</span></span>
<span><span class="va">N</span><span class="op">=</span><span class="fl">5</span><span class="op">*</span><span class="fl">10</span><span class="op">^</span><span class="fl">6</span> <span class="co">##5,000,000 tickets bought</span></span>
<span><span class="va">winners</span><span class="op">=</span><span class="fu">rbinom</span><span class="op">(</span><span class="fl">1000</span>,<span class="va">N</span>,<span class="va">p</span><span class="op">)</span> <span class="co">##1000 is the number of different lotto draws</span></span>
<span><span class="va">tab</span><span class="op">=</span><span class="fu">table</span><span class="op">(</span><span class="va">winners</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tab</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-lottery_winners_outcomes-1.png" alt="Number of people that win the lottery obtained from Monte Carlo simulation." class="figure mx-auto d-block"><figcaption>
Number of people that win the lottery obtained from Monte Carlo
simulation.
</figcaption></figure><div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">prop.table</span><span class="op">(</span><span class="va">tab</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>winners
    0     1     2     3     4
0.618 0.291 0.077 0.012 0.002 </code></pre>
</div>
<p>For cases like this, where <span class="math inline">\(N\)</span> is
very large, but <span class="math inline">\(p\)</span> is small enough
to make <span class="math inline">\(N \times p\)</span> (call it <span class="math inline">\(\lambda\)</span>) a number between 0 and, for
example, 10, then <span class="math inline">\(S\)</span> can be shown to
follow a Poisson distribution, which has a simple parametric form:</p>
<p><span class="math display">\[
\mbox{Pr}(S=k)=\frac{\lambda^k \exp{-\lambda}}{k!}
\]</span></p>
<p>The Poisson distribution is commonly used in RNA-seq analyses.
Because we are sampling thousands of molecules and most genes represent
a very small proportion of the totality of molecules, the Poisson
distribution seems appropriate.</p>
<p>So how does this help us? One way is that it provides insight about
the statistical properties of summaries that are widely used in
practice. For example, let’s say we only have one sample from each of a
case and control RNA-seq experiment and we want to report the genes with
the largest fold-changes. One insight that the Poisson model provides is
that under the null hypothesis of no true differences, the statistical
variability of this quantity depends on the total abundance of the gene.
We can show this mathematically, but here is a quick simulation to
demonstrate the point:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span><span class="op">=</span><span class="fl">10000</span><span class="co">##number of genes</span></span>
<span><span class="va">lambdas</span><span class="op">=</span><span class="fl">2</span><span class="op">^</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">16</span>,len<span class="op">=</span><span class="va">N</span><span class="op">)</span> <span class="co">##these are the true abundances of genes</span></span>
<span><span class="va">y</span><span class="op">=</span><span class="fu">rpois</span><span class="op">(</span><span class="va">N</span>,<span class="va">lambdas</span><span class="op">)</span><span class="co">##note that the null hypothesis is true for all genes</span></span>
<span><span class="va">x</span><span class="op">=</span><span class="fu">rpois</span><span class="op">(</span><span class="va">N</span>,<span class="va">lambdas</span><span class="op">)</span> </span>
<span><span class="va">ind</span><span class="op">=</span><span class="fu">which</span><span class="op">(</span><span class="va">y</span><span class="op">&gt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">x</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span><span class="co">##make sure no 0s due to ratio and log</span></span>
<span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">splot</span><span class="op">(</span><span class="fu">log2</span><span class="op">(</span><span class="va">lambdas</span><span class="op">)</span>,<span class="fu">log2</span><span class="op">(</span><span class="va">y</span><span class="op">/</span><span class="va">x</span><span class="op">)</span>,subset<span class="op">=</span><span class="va">ind</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-rna_seq_simulation-1.png" alt="MA plot of simulated RNA-seq data. Replicated measurements follow a Poisson distribution." class="figure mx-auto d-block"><figcaption>
MA plot of simulated RNA-seq data. Replicated measurements follow a
Poisson distribution.
</figcaption></figure><p>For lower values of <code>lambda</code> there is much more
variability and, if we were to report anything with a fold change of 2
or more, the number of false positives would be quite high for low
abundance genes.</p>
<div class="section level4">
<h4 id="ngs-experiments-and-the-poisson-distribution">NGS experiments and the Poisson distribution<a class="anchor" aria-label="anchor" href="#ngs-experiments-and-the-poisson-distribution"></a>
</h4>
<p>In this section we will use the data stored in this dataset:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">parathyroidSE</span><span class="op">)</span> <span class="co">##available from Bioconductor</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">parathyroidGenesSE</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">parathyroidGenesSE</span></span></code></pre>
</div>
<p>The data is contained in a <code>SummarizedExperiment</code> object,
which we do not describe here. The important thing to know is that it
includes a matrix of data, where each row is a genomic feature and each
column is a sample. We can extract this data using the
<code>assay</code> function. For this dataset, the value of a single
cell in the data matrix is the count of reads which align to a given
gene for a given sample. Thus, a similar plot to the one we simulated
above with technical replicates reveals that the behavior predicted by
the model is present in experimental data:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span>,<span class="fl">23</span><span class="op">]</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span>,<span class="fl">24</span><span class="op">]</span></span>
<span><span class="va">ind</span><span class="op">=</span><span class="fu">which</span><span class="op">(</span><span class="va">y</span><span class="op">&gt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">x</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span><span class="co">##make sure no 0s due to ratio and log</span></span>
<span><span class="fu">splot</span><span class="op">(</span><span class="op">(</span><span class="fu">log2</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">+</span><span class="fu">log2</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>,<span class="fu">log</span><span class="op">(</span><span class="va">x</span><span class="op">/</span><span class="va">y</span><span class="op">)</span>,subset<span class="op">=</span><span class="va">ind</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-RNA-seq_MAplot-1.png" alt="MA plot of replicated RNA-seq data." class="figure mx-auto d-block"><figcaption>
MA plot of replicated RNA-seq data.
</figcaption></figure><p>If we compute the standard deviations across four individuals, it is
quite a bit higher than what is predicted by a Poisson model. Assuming
most genes are differentially expressed across individuals, then, if the
Poisson model is appropriate, there should be a linear relationship in
this plot:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">matrixStats</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vars</span><span class="op">=</span><span class="fu">rowVars</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">8</span>,<span class="fl">16</span>,<span class="fl">21</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="co">##we know these are four</span></span>
<span><span class="va">means</span><span class="op">=</span><span class="fu">rowMeans</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">8</span>,<span class="fl">16</span>,<span class="fl">21</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="co">##different individuals</span></span>
<span></span>
<span><span class="fu">splot</span><span class="op">(</span><span class="va">means</span>,<span class="va">vars</span>,log<span class="op">=</span><span class="st">"xy"</span>,subset<span class="op">=</span><span class="fu">which</span><span class="op">(</span><span class="va">means</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">&amp;</span><span class="va">vars</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="co">##plot a subset of data</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,col<span class="op">=</span><span class="fl">2</span>,lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-var_vs_mean-1.png" alt="Variance versus mean plot. Summaries were obtained from the RNA-seq data." class="figure mx-auto d-block"><figcaption>
Variance versus mean plot. Summaries were obtained from the RNA-seq
data.
</figcaption></figure><p>The reason for this is that the variability plotted here includes
biological variability, which the motivation for the Poisson does not
include. The negative binomial distribution, which combines the sampling
variability of a Poisson and biological variability, is a more
appropriate distribution to model this type of experiment. The negative
binomial has two parameters and permits more flexibility for count data.
For more on the use of the negative binomial to model RNA-seq data you
can read <a href="http://www.ncbi.nlm.nih.gov/pubmed/20979621" class="external-link">this
paper</a>. The Poisson is a special case of the negative binomial
distribution.</p>
</div>
</div>
<div class="section level2">
<h2 id="maximum-likelihood-estimation">Maximum Likelihood Estimation<a class="anchor" aria-label="anchor" href="#maximum-likelihood-estimation"></a>
</h2>
<p>To illustrate the concept of maximum likelihood estimates (MLE), we
use a relatively simple dataset containing palindrome locations in the
HMCV genome. We read in the locations of the palindrome and then count
the number of palindromes in each 4,000 basepair segments.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">=</span><span class="fu">read.csv</span><span class="op">(</span><span class="st">"./data/hcmv.rda"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `[.data.frame`(read.csv("./data/hcmv.rda"), , 2): undefined columns selected</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">breaks</span><span class="op">=</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">4000</span><span class="op">*</span><span class="fu">round</span><span class="op">(</span><span class="fu">max</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">/</span><span class="fl">4000</span><span class="op">)</span>,<span class="fl">4000</span><span class="op">)</span></span>
<span><span class="va">tmp</span><span class="op">=</span><span class="fu">cut</span><span class="op">(</span><span class="va">x</span>,<span class="va">breaks</span><span class="op">)</span></span>
<span><span class="va">counts</span><span class="op">=</span><span class="fu">table</span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-palindrome_count_histogram-1.png" alt="Palindrome count histogram." class="figure mx-auto d-block"><figcaption>
Palindrome count histogram.
</figcaption></figure><p>The counts do appear to follow a Poisson distribution. But what is
the rate <span class="math inline">\(\lambda\)</span> ? The most common
approach to estimating this rate is <em>maximum likelihood
estimation</em>. To find the maximum likelihood estimate (MLE), we note
that these data are independent and the probability of observing the
values we observed is:</p>
<p><span class="math display">\[
\Pr(X_1=k_1,\dots,X_n=k_n;\lambda) = \prod_{i=1}^n \lambda^{k_i} / k_i!
\exp ( -\lambda)
\]</span></p>
<p>The MLE is the value of <span class="math inline">\(\lambda\)</span>
that maximizes the likelihood:.</p>
<p><span class="math display">\[
\mbox{L}(\lambda; X_1=k_1,\dots,X_n=k_1)=\exp\left\{\sum_{i=1}^n \log
\Pr(X_i=k_i;\lambda)\right\}
\]</span></p>
<p>In practice, it is more convenient to maximize the log-likelihood
which is the summation that is exponentiated in the expression above.
Below we write code that computes the log-likelihood for any <span class="math inline">\(\lambda\)</span> and use the function
<code>optimize</code> to find the value that maximizes this function
(the MLE). We show a plot of the log-likelihood along with vertical line
showing the MLE.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span><span class="op">&lt;-</span><span class="kw">function</span><span class="op">(</span><span class="va">lambda</span><span class="op">)</span> <span class="fu">sum</span><span class="op">(</span><span class="fu">dpois</span><span class="op">(</span><span class="va">counts</span>,<span class="va">lambda</span>,log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">lambdas</span><span class="op">&lt;-</span><span class="fu">seq</span><span class="op">(</span><span class="fl">3</span>,<span class="fl">7</span>,len<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">ls</span> <span class="op">&lt;-</span> <span class="fu">exp</span><span class="op">(</span><span class="fu">sapply</span><span class="op">(</span><span class="va">lambdas</span>,<span class="va">l</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">lambdas</span>,<span class="va">ls</span>,type<span class="op">=</span><span class="st">"l"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mle</span><span class="op">=</span><span class="fu">optimize</span><span class="op">(</span><span class="va">l</span>,<span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span>,maximum<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v<span class="op">=</span><span class="va">mle</span><span class="op">$</span><span class="va">maximum</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-mle-1.png" alt="Likelihood versus lambda." class="figure mx-auto d-block"><figcaption>
Likelihood versus lambda.
</figcaption></figure><p>If you work out the math and do a bit of calculus, you realize that
this is a particularly simple example for which the MLE is the
average.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">print</span><span class="op">(</span> <span class="fu">c</span><span class="op">(</span><span class="va">mle</span><span class="op">$</span><span class="va">maximum</span>, <span class="fu">mean</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">)</span> <span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]    9.999944 1279.750000</code></pre>
</div>
<p>Note that a plot of observed counts versus counts predicted by the
Poisson shows that the fit is quite good in this case:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">theoretical</span><span class="op">&lt;-</span><span class="fu">qpois</span><span class="op">(</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">99</span><span class="op">)</span><span class="op">+</span><span class="fl">0.5</span><span class="op">)</span><span class="op">/</span><span class="fl">100</span>,<span class="fu">mean</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">qqplot</span><span class="op">(</span><span class="va">theoretical</span>,<span class="va">counts</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-obs_versus_theoretical_Poisson_count-1.png" alt="Observed counts versus theoretical Poisson counts." class="figure mx-auto d-block"><figcaption>
Observed counts versus theoretical Poisson counts.
</figcaption></figure><p>We therefore can model the palindrome count data with a Poisson with
<span class="math inline">\(\lambda=5.16\)</span>.</p>
</div>
<div class="section level2">
<h2 id="distributions-for-positive-continuous-values">Distributions for Positive Continuous Values<a class="anchor" aria-label="anchor" href="#distributions-for-positive-continuous-values"></a>
</h2>
<p>Different genes vary differently across biological replicates. Later,
in the hierarchical models chapter, we will describe one of the <a href="http://www.ncbi.nlm.nih.gov/pubmed/16646809" class="external-link">most influential
statistical methods</a> in the analysis of genomics data. This method
provides great improvements over naive approaches to detecting
differentially expressed genes. This is achieved by modeling the
distribution of the gene variances. Here we describe the parametric
model used in this method.</p>
<p>We want to model the distribution of the gene-specific standard
errors. Are they normal? Keep in mind that we are modeling the
population standard errors so CLT does not apply, even though we have
thousands of genes.</p>
<p>As an example, we use an experimental data that included both
technical and biological replicates for gene expression measurements on
mice. We can load the data and compute the gene specific sample standard
error for both the technical replicates and the biological
replicates.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">Biobase</span><span class="op">)</span> <span class="co">##available from Bioconductor</span></span>
<span><span class="fu">load</span><span class="op">(</span><span class="st">"data/maPooling.RData"</span><span class="op">)</span></span>
<span><span class="va">pd</span><span class="op">=</span><span class="fu">pData</span><span class="op">(</span><span class="va">maPooling</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##determine which samples are bio reps and which are tech reps</span></span>
<span><span class="va">strain</span><span class="op">=</span><span class="fu">factor</span><span class="op">(</span><span class="fu">as.numeric</span><span class="op">(</span><span class="fu">grepl</span><span class="op">(</span><span class="st">"b"</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">pd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pooled</span><span class="op">=</span><span class="fu">which</span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="va">pd</span><span class="op">)</span><span class="op">==</span><span class="fl">12</span> <span class="op">&amp;</span> <span class="va">strain</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">techreps</span><span class="op">=</span><span class="fu">exprs</span><span class="op">(</span><span class="va">maPooling</span><span class="op">[</span>,<span class="va">pooled</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">individuals</span><span class="op">=</span><span class="fu">which</span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="va">pd</span><span class="op">)</span><span class="op">==</span><span class="fl">1</span> <span class="op">&amp;</span> <span class="va">strain</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##remove replicates</span></span>
<span><span class="va">individuals</span><span class="op">=</span><span class="va">individuals</span><span class="op">[</span><span class="op">-</span><span class="fu">grep</span><span class="op">(</span><span class="st">"tr"</span>, <span class="fu">names</span><span class="op">(</span><span class="va">individuals</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">bioreps</span><span class="op">=</span><span class="fu">exprs</span><span class="op">(</span><span class="va">maPooling</span><span class="op">)</span><span class="op">[</span>,<span class="va">individuals</span><span class="op">]</span></span>
<span></span>
<span><span class="co">###now compute the gene specific standard deviations</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">matrixStats</span><span class="op">)</span></span>
<span><span class="va">techsds</span><span class="op">=</span><span class="fu">rowSds</span><span class="op">(</span><span class="va">techreps</span><span class="op">)</span></span>
<span><span class="va">biosds</span><span class="op">=</span><span class="fu">rowSds</span><span class="op">(</span><span class="va">bioreps</span><span class="op">)</span></span></code></pre>
</div>
<p>We can now explore the sample standard deviation:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">###now plot</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">shist</span><span class="op">(</span><span class="va">biosds</span>,unit<span class="op">=</span><span class="fl">0.1</span>,col<span class="op">=</span><span class="fl">1</span>,xlim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">shist</span><span class="op">(</span><span class="va">techsds</span>,unit<span class="op">=</span><span class="fl">0.1</span>,col<span class="op">=</span><span class="fl">2</span>,add<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">legend</span><span class="op">(</span><span class="st">"topright"</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"Biological"</span>,<span class="st">"Technical"</span><span class="op">)</span>, col<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,lty<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-bio_sd_versus_tech_sd-1.png" alt="Histograms of biological variance and technical variance." class="figure mx-auto d-block"><figcaption>
Histograms of biological variance and technical variance.
</figcaption></figure><p>An important observation here is that the biological variability is
substantially higher than the technical variability. This provides
strong evidence that genes do in fact have gene-specific biological
variability.</p>
<p>If we want to model this variability, we first notice that the normal
distribution is not appropriate here since the right tail is rather
large. Also, because SDs are strictly positive, there is a limitation to
how symmetric this distribution can be. A qqplot shows this very
clearly:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">biosds</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">biosds</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-sd_qqplot-1.png" alt="Normal qq-plot for sample standard deviations." class="figure mx-auto d-block"><figcaption>
Normal qq-plot for sample standard deviations.
</figcaption></figure><p>There are parametric distributions that possess these properties
(strictly positive and <em>heavy</em> right tails). Two examples are the
<em>gamma</em> and <em>F</em> distributions. The density of the gamma
distribution is defined by:</p>
<p><span class="math display">\[
f(x;\alpha,\beta)=\frac{\beta^\alpha x^{\alpha-1}\exp{-\beta
x}}{\Gamma(\alpha)}
\]</span></p>
<p>It is defined by two parameters <span class="math inline">\(\alpha\)</span> and <span class="math inline">\(\beta\)</span> that can, indirectly, control
location and scale. They also control the shape of the distribution. For
more on this distribution please refer to <a href="https://www.stat.berkeley.edu/~rice/Book3ed/index.html" class="external-link">this
book</a>.</p>
<p>Two special cases of the gamma distribution are the chi-squared and
exponential distribution. We used the chi-squared earlier to analyze a
2x2 table data. For chi-square, we have <span class="math inline">\(\alpha=\nu/2\)</span> and <span class="math inline">\(\beta=2\)</span> with <span class="math inline">\(\nu\)</span> the degrees of freedom. For
exponential, we have <span class="math inline">\(\alpha=1\)</span> and
<span class="math inline">\(\beta=\lambda\)</span> the rate.</p>
<p>The F-distribution comes up in analysis of variance (ANOVA). It is
also always positive and has large right tails. Two parameters control
its shape:</p>
<p><span class="math display">\[
f(x,d_1,d_2)=\frac{1}{B\left( \frac{d_1}{2},\frac{d_2}{2}\right)}
  \left(\frac{d_1}{d_2}\right)^{\frac{d_1}{2}}
  x^{\frac{d_1}{2}-1}\left(1+\frac{d1}{d2}x\right)^{-\frac{d_1+d_2}{2}}
\]</span></p>
<p>with <span class="math inline">\(B\)</span> the <em>beta
function</em> and <span class="math inline">\(d_1\)</span> and <span class="math inline">\(d_2\)</span> are called the degrees of freedom for
reasons having to do with how it arises in ANOVA. A third parameter is
sometimes used with the F-distribution, which is a scale parameter.</p>
<div class="section level4">
<h4 id="modeling-the-variance">Modeling the variance<a class="anchor" aria-label="anchor" href="#modeling-the-variance"></a>
</h4>
<p>In a later section we will learn about a hierarchical model approach
to improve estimates of variance. In these cases it is mathematically
convenient to model the distribution of the variance <span class="math inline">\(\sigma^2\)</span>. The hierarchical model used <a href="http://www.ncbi.nlm.nih.gov/pubmed/16646809" class="external-link">here</a> implies that
the sample standard deviation of genes follows scaled F-statistics:</p>
<p><span class="math display">\[
s^2 \sim s_0^2 F_{d,d_0}
\]</span></p>
<p>with <span class="math inline">\(d\)</span> the degrees of freedom
involved in computing <span class="math inline">\(s^2\)</span>. For
example, in a case comparing 3 versus 3, the degrees of freedom would be
4. This leaves two free parameters to adjust to the data. Here <span class="math inline">\(d\)</span> will control the location and <span class="math inline">\(s_0\)</span> will control the scale. Below are
some examples of <span class="math inline">\(F\)</span> distribution
plotted on top of the histogram from the sample variances:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">3</span>,<span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">sds</span><span class="op">=</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">2</span>,len<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">d</span> <span class="kw">in</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">5</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">s0</span> <span class="kw">in</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">tmp</span><span class="op">=</span><span class="fu">hist</span><span class="op">(</span><span class="va">biosds</span>,main<span class="op">=</span><span class="fu">paste</span><span class="op">(</span><span class="st">"s_0 ="</span>,<span class="va">s0</span>,<span class="st">"d ="</span>,<span class="va">d</span><span class="op">)</span>,</span>
<span>      xlab<span class="op">=</span><span class="st">"sd"</span>,ylab<span class="op">=</span><span class="st">"density"</span>,freq<span class="op">=</span><span class="cn">FALSE</span>,nc<span class="op">=</span><span class="fl">100</span>,xlim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">dd</span><span class="op">=</span><span class="fu">df</span><span class="op">(</span><span class="va">sds</span><span class="op">^</span><span class="fl">2</span><span class="op">/</span><span class="va">s0</span><span class="op">^</span><span class="fl">2</span>,<span class="fl">11</span>,<span class="va">d</span><span class="op">)</span></span>
<span>    <span class="co">##multiply by normalizing constant to assure same range on plot</span></span>
<span>    <span class="va">k</span><span class="op">=</span><span class="fu">sum</span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">density</span><span class="op">)</span><span class="op">/</span><span class="fu">sum</span><span class="op">(</span><span class="va">dd</span><span class="op">)</span> </span>
<span>    <span class="fu">lines</span><span class="op">(</span><span class="va">sds</span>,<span class="va">dd</span><span class="op">*</span><span class="va">k</span>,type<span class="op">=</span><span class="st">"l"</span>,col<span class="op">=</span><span class="fl">2</span>,lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-modeling_variance-1.png" alt="Histograms of sample standard deviations and densities of estimated distributions." class="figure mx-auto d-block"><figcaption>
Histograms of sample standard deviations and densities of estimated
distributions.
</figcaption></figure><p>Now which <span class="math inline">\(s_0\)</span> and <span class="math inline">\(d\)</span> fit our data best? This is a rather
advanced topic as the MLE does not perform well for this particular
distribution (we refer to Smyth (2004)). The Bioconductor
<code>limma</code> package provides a function to estimate these
parameters:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">limma</span><span class="op">)</span></span>
<span><span class="va">estimates</span><span class="op">=</span><span class="fu">fitFDist</span><span class="op">(</span><span class="va">biosds</span><span class="op">^</span><span class="fl">2</span>, <span class="fl">11</span><span class="op">)</span></span>
<span></span>
<span><span class="va">theoretical</span><span class="op">&lt;-</span> <span class="fu">sqrt</span><span class="op">(</span><span class="fu">qf</span><span class="op">(</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">999</span><span class="op">)</span> <span class="op">+</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">/</span><span class="fl">1000</span>, <span class="fl">11</span>, <span class="va">estimates</span><span class="op">$</span><span class="va">df2</span><span class="op">)</span> <span class="op">*</span> <span class="va">estimates</span><span class="op">$</span><span class="va">scale</span><span class="op">)</span></span>
<span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="va">biosds</span></span></code></pre>
</div>
<p>The fitted models do appear to provide a reasonable approximation, as
demonstrated by the qq-plot and histogram:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">qqplot</span><span class="op">(</span><span class="va">theoretical</span>,<span class="va">observed</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tmp</span><span class="op">=</span><span class="fu">hist</span><span class="op">(</span><span class="va">biosds</span>,main<span class="op">=</span><span class="fu">paste</span><span class="op">(</span><span class="st">"s_0 ="</span>, <span class="fu">signif</span><span class="op">(</span><span class="va">estimates</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>                  <span class="st">"d ="</span>, <span class="fu">signif</span><span class="op">(</span><span class="va">estimates</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  xlab<span class="op">=</span><span class="st">"sd"</span>, ylab<span class="op">=</span><span class="st">"density"</span>, freq<span class="op">=</span><span class="cn">FALSE</span>, nc<span class="op">=</span><span class="fl">100</span>, xlim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">9</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dd</span><span class="op">=</span><span class="fu">df</span><span class="op">(</span><span class="va">sds</span><span class="op">^</span><span class="fl">2</span><span class="op">/</span><span class="va">estimates</span><span class="op">$</span><span class="va">scale</span>,<span class="fl">11</span>,<span class="va">estimates</span><span class="op">$</span><span class="va">df2</span><span class="op">)</span></span>
<span><span class="va">k</span><span class="op">=</span><span class="fu">sum</span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">density</span><span class="op">)</span><span class="op">/</span><span class="fu">sum</span><span class="op">(</span><span class="va">dd</span><span class="op">)</span> <span class="co">##a normalizing constant to assure same area in plot</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">sds</span>, <span class="va">dd</span><span class="op">*</span><span class="va">k</span>, type<span class="op">=</span><span class="st">"l"</span>, col<span class="op">=</span><span class="fl">2</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-variance_model_fit-1.png" alt="qq-plot (left) and density (right) demonstrate that model fits data well." class="figure mx-auto d-block"><figcaption>
qq-plot (left) and density (right) demonstrate that model fits data
well.
</figcaption></figure>
</div>
</div>
<div class="section level2">
<h2 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<p>Suppose you have an urn with blue and red balls. If N balls are
selected at random with replacement (you put the ball back after you
pick it), we can denote the outcomes as random variables X1,…,XN that
are 1 or 0. If the proportion of red balls is p , then the distribution
of each of these is Pr(Xi=1)=p.</p>
<p>These are also called Bernoulli trials. These random variables are
independent because we replace the balls. Flipping a coin is an example
of this with p=0.5.</p>
<p>You can show that the mean and variance are p and p(1−p)
respectively. The binomial distribution gives us the distribution of the
sum SN of these random variables. The probability that we see k red
balls is given by: Pr(SN=k)=(Nk)pk(1−p)N−k</p>
<p>In R, the function dbimom gives you this result. The function pbinom
gives us Pr(SN≤k).</p>
<p>This equation has many uses in the life sciences. We give some
examples below.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>The probability of conceiving a girl is 0.49. What is the probability
that a family with 4 children has 2 girls and 2 boys (you can assume
that the outcomes are independent)?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p><code>g &lt;- sampleInfo$group</code><br><code>pvals &lt;- rowttests(geneExpression, factor(g))$p.value</code><br><code>sum(pvals &lt; 0.05)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>Use what you learned in Question 1 to answer these questions: What is
the probability that a family with 10 children has 6 girls and 4 boys
(assume no twins)?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p><code>p &lt;- 0.49</code><br><code>dbinom(6,10,p=p)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>The genome has 3 billion bases. About 20% are C, 20% are G, 30% are
T, and 30% are A. Suppose you take a random interval of 20 bases, what
is the probability that the GC-content (proportion of Gs of Cs) is
strictly above 0.5 in this interval?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p><code>p &lt;- 0.4</code><br><code>pbinom(20,20,p) - pbinom(10,20,p)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>The probability of winning the lottery is 1 in 175,223,510. If
20,000,000 people buy a ticket, what is the probability that more than
one person wins?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p><code>p &lt;- 1/175233510</code><br><code>N &lt;- 20000000</code><br><code>1 - ppois(1,N*p)</code><br><code>1 - pbinom(1,N,p)</code><br>
Since the poisson distribution is a type of binomial distribution, both
distributions give the same values.</p>
</div>
</div>
</div>
</div>
<div id="exercise-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-5" class="callout-inner">
<h3 class="callout-title">Exercise 5</h3>
<div class="callout-content">
<p>We can show that the binomial approximation is approximately normal
when N is large and p is not too close to 0 or 1. This means that:</p>
<p>SN−E(SN)Var(SN)‾‾‾‾‾‾‾‾√</p>
<p>is approximately normal with mean 0 and SD 1. Using the results for
sums of independent random variables, we can show that E(SN)=Np and
Var(Sn)=Np(1−p). The genome has 3 billion bases. About 20% are C, 20%
are G, 30% are T, and 30% are A. Suppose you take a random interval of
20 bases, what is the exact probability that the GC-content (proportion
of Gs of Cs) is greater than 0.35 and smaller or equal to 0.45 in this
interval?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<p><code>p &lt;- 0.4</code><br><code>pbinom(0.45*20, 20, p) - pbinom(0.35*20,20,p)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-6" class="callout-inner">
<h3 class="callout-title">Exercise 6</h3>
<div class="callout-content">
<p>For the question above, what is the normal approximation to the
probability?</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<p><code>p &lt;- 0.4</code><br><code>N &lt;- 20</code><br><code>a &lt;- (0.45*20 - N*p) / sqrt(N*p*(1-p))</code><br><code>b &lt;- (0.35*20 - N*p) / sqrt(N*p*(1-p))</code></p>
<p><code>approx &lt;- pnorm(a) - pnorm(b)</code><br><code>approx</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-7" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-7" class="callout-inner">
<h3 class="callout-title">Exercise 7</h3>
<div class="callout-content">
<p>Repeat exercise 5, but using an interval of 1000 bases. What is the
difference (in absolute value) between the normal approximation and the
exact distribution of the GC-content being greater than 0.35 and lesser
or equal to 0.45?</p>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<p><code>N &lt;- 1000</code><br><code>p &lt;- 0.4</code><br><code>a &lt;- (0.45*N - N*p) / sqrt(N*p*(1-p))</code><br><code>b &lt;- (0.35*N - N*p) / sqrt(N*p*(1-p))</code><br><code>approx &lt;- pnorm(a) - pnorm(b)</code></p>
<p><code>exact &lt;- pbinom(0.45*N,N,p) - pbinom(0.35*N,N,p)</code><br><code>exact - approx</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-8" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-8" class="callout-inner">
<h3 class="callout-title">Exercise 8</h3>
<div class="callout-content">
<p>The Cs in our genomes can be methylated or unmethylated. Suppose we
have a large (millions) group of cells in which a proportion p of the Cs
of interest are methylated. We break up the DNA of these cells and
randomly select pieces and end up with N pieces that contain the C we
care about. This means that the probability of seeing k methylated Cs is
binomial: exact = dbinom(k,N,p) We can approximate this with the normal
distribution: <code>a &lt;- (k+0.5 - N*p)/sqrt(N*p*(1-p))</code><br><code>b &lt;- (k-0.5 - N*p)/sqrt(N*p*(1-p))</code><br><code>approx = pnorm(a) - pnorm(b)</code><br>
Compute the difference approx - exact for:<br><code>N &lt;- c(5,10,50,100,500)</code><br><code>p &lt;- seq(0,1,0.25)</code><br>
Compare the approximation and exact probability of the proportion of Cs
being p, k=1,…,N−1 plotting the exact versus the approximation for each
p and N combination. A) The normal approximation works well when p is
close to 0.5 even for small N=10<br>
B) The normal approximation breaks down when p is close to 0 or 1 even
for large N<br>
C) When N is 100 all approximations are spot on.<br>
D) When p=0.01 the approximation are terrible for N=5,10,30 and only OK
for N=100</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<p><code>Ns &lt;- c(5,10,50,100,500)</code><br><code>Ps &lt;- seq(0,1,0.25)</code><br><code>mypar(5,5)</code><br><code>for (i in seq_along(Ns)) {</code><br><code>n &lt;- Ns[[i]]</code><br><code>k &lt;- seq(1:n-1)</code><br><code>for (j in seq_along(Ps)) {</code><br><code>p &lt;- Ps[[j]]</code><br><code>exact = dbinom(k, n, p)</code><br><code>a = (k+0.5- n*p)/sqrt(n*p*(1-p))</code><br><code>b = (k-0.5- n*p)/sqrt(n*p*(1-p))</code><br><code>approx = pnorm(a) - pnorm(b)</code><br><code>qqplot(exact,approx,xlab='exact',ylab='approx',</code><br><code>main = paste0('N=',n,' ','p=',p))</code><br><code>abline(0,1)</code><br><code>}</code><br><code>}</code><br>
The answer is C: When N is 100 all approximations are spot on. When p is
close to 0 or 1, the normal distribution breaks down even at N =
100.</p>
</div>
</div>
</div>
</div>
<div id="exercise-9" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-9" class="callout-inner">
<h3 class="callout-title">Exercise 9</h3>
<div class="callout-content">
<p>We saw in the previous question that when p is very small, the normal
approximation breaks down. If N is very large, then we can use the
Poisson approximation.<br>
Earlier we computed 1 or more people winning the lottery when the
probability of winning was 1 in 175,223,510 and 20,000,000 people bought
a ticket. Using the binomial, we can compute the probability of exactly
two people winning to be:</p>
<p><code>N &lt;- 20000000</code> <code>p &lt;- 1/175223510</code>
<code>dbinom(2,N,p)</code></p>
<p>If we were to use the normal approximation, we would greatly
underestimate this:</p>
<p><code>a &lt;- (2+0.5 - N*p)/sqrt(N*p*(1-p))</code>
<code>b &lt;- (2-0.5 - N*p)/sqrt(N*p*(1-p))</code>
<code>pnorm(a) - pnorm(b)</code></p>
<p>To use the Poisson approximation here, use the rate λ=Np</p>
<p>representing the number of people per 20,000,000 that win the
lottery. Note how much better the approximation is:</p>
<p><code>dpois(2,N*p)</code></p>
<p>In this case. it is practically the same because N is very large and
Np is not 0. These are the assumptions needed for the Poisson to work.
What is the Poisson approximation for more than one person winning?</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" aria-labelledby="headingSolution9" data-bs-parent="#accordionSolution9">
<div class="accordion-body">
<p><code>1 - ppois(1,N*p)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-10" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-10" class="callout-inner">
<h3 class="callout-title">Exercise 10</h3>
<div class="callout-content">
<p>Now we are going to explore if palindromes are over-represented in
some part of the HCMV genome. Make sure you have the latest version of
the dagdata, load the palindrome data from the Human cytomegalovirus
genome, and plot locations of palindromes on the genome for this
virus:</p>
<p><code>library(dagdata)</code><br><code>data(hcmv)</code><br><code>plot(locations,rep(1,length(locations)),ylab="",yaxt="n")</code></p>
<p>These palindromes are quite rare, and therefore p is very small. If
we break the genome into bins of 4000 basepairs, then we have Np not so
small and we might be able to use Poisson to model the number of
palindromes in each bin:</p>
<p><code>breaks=seq(0,4000*round(max(locations)/4000),4000)</code>
<code>tmp=cut(locations,breaks)</code>
<code>counts=as.numeric(table(tmp))</code></p>
<p>So if our model is correct, counts should follow a Poisson
distribution. The distribution seems about right:</p>
<p><code>hist(counts)</code></p>
<p>So let X1,…,Xn be the random variables representing counts then
Pr(Xi=k)=λk/k!exp(−λ) and to fully describe this distribution, we need
to know λ. For this we will use MLE. We can write the likelihood
described in book in R. For example, for λ=4 we have:</p>
<p><code>probs &lt;- dpois(counts,4)</code>
<code>likelihood &lt;- prod(probs)</code> <code>likelihood</code></p>
<p>Notice that it’s a tiny number. It is usually more convenient to
compute log-likelihoods:</p>
<p><code>logprobs &lt;- dpois(counts,4,log=TRUE)</code><br><code>loglikelihood &lt;- sum(logprobs)</code><br><code>loglikelihood</code></p>
<p>Now write a function that takes λ and the vector of counts as input
and returns the log-likelihood. Compute this log-likelihood for lambdas
= <code>seq(0,15,len=300)</code> and make a plot. What value of lambdas
maximizes the log-likelihood?</p>
</div>
</div>
</div>
<div id="accordionSolution10" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution10" aria-expanded="false" aria-controls="collapseSolution10">
  <h4 class="accordion-header" id="headingSolution10"> Show me the solution </h4>
</button>
<div id="collapseSolution10" class="accordion-collapse collapse" aria-labelledby="headingSolution10" data-bs-parent="#accordionSolution10">
<div class="accordion-body">
<p><code>loglike = function(lambdas) {</code><br><code>logprobs &lt;- dpois(counts,lambdas,log=T)</code><br><code>loglikelihood &lt;- sum(logprobs)</code><br><code>return(loglikelihood)</code><br><code>}</code><br><code>lambdas &lt;- seq(0,15,len=300)</code><br><code>log_res &lt;- exp(sapply(lambdas,loglike))</code><br><code>optim_lambda &lt;- lambdas[which(log_res == max(log_res))]</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-11" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-11" class="callout-inner">
<h3 class="callout-title">Exercise 11</h3>
<div class="callout-content">
<p>The point of collecting this dataset was to try to determine if there
is a region of the genome that has a higher palindrome rate than
expected. We can create a plot and see the counts per location:</p>
<p><code>library(dagdata)</code><br><code>data(hcmv)</code><br><code>breaks=seq(0,4000*round(max(locations)/4000),4000)</code><br><code>tmp=cut(locations,breaks)</code><br><code>counts=as.numeric(table(tmp))</code><br><code>binLocation=(breaks[-1]+breaks[-length(breaks)])/2</code><br><code>plot(binLocation,counts,type="l",xlab=)</code></p>
<p>What is the center of the bin with the highest count?</p>
</div>
</div>
</div>
<div id="accordionSolution11" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution11" aria-expanded="false" aria-controls="collapseSolution11">
  <h4 class="accordion-header" id="headingSolution11"> Show me the solution </h4>
</button>
<div id="collapseSolution11" class="accordion-collapse collapse" aria-labelledby="headingSolution11" data-bs-parent="#accordionSolution11">
<div class="accordion-body">
<p><code>binLocation[which(counts == max(counts))]</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-12" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-12" class="callout-inner">
<h3 class="callout-title">Exercise 12</h3>
<div class="callout-content">
<p>What is the maximum count?</p>
</div>
</div>
</div>
<div id="accordionSolution12" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution12" aria-expanded="false" aria-controls="collapseSolution12">
  <h4 class="accordion-header" id="headingSolution12"> Show me the solution </h4>
</button>
<div id="collapseSolution12" class="accordion-collapse collapse" aria-labelledby="headingSolution12" data-bs-parent="#accordionSolution12">
<div class="accordion-body">
<p><code>max(counts)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-13" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-13" class="callout-inner">
<h3 class="callout-title">Exercise 13</h3>
<div class="callout-content">
<p>Once we have identified the location with the largest palindrome
count, we want to know if we could see a value this big by chance. If X
is a Poisson random variable with rate:</p>
<p><code>lambda = mean(counts[ - which.max(counts) ])</code></p>
<p>What is the probability of seeing a count of 14 or more?</p>
</div>
</div>
</div>
<div id="accordionSolution13" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution13" aria-expanded="false" aria-controls="collapseSolution13">
  <h4 class="accordion-header" id="headingSolution13"> Show me the solution </h4>
</button>
<div id="collapseSolution13" class="accordion-collapse collapse" aria-labelledby="headingSolution13" data-bs-parent="#accordionSolution13">
<div class="accordion-body">
<p><code>1 - ppois(13, lambda)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-14" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-14" class="callout-inner">
<h3 class="callout-title">Exercise 14</h3>
<div class="callout-content">
<p>So we obtain a p-value smaller than 0.001 for a count of 14. Why is
it problematic to report this p-value as strong evidence of a location
that is different?</p>
<ol style="list-style-type: upper-alpha">
<li>Poisson in only an approximation.<br>
</li>
<li>We selected the highest region out of 57 and need to adjust for
multiple testing.<br>
</li>
<li>λ is an estimate, a random variable, and we didn’t take into account
its variability.<br>
</li>
<li>We don’t know the effect size.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution14" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution14" aria-expanded="false" aria-controls="collapseSolution14">
  <h4 class="accordion-header" id="headingSolution14"> Show me the solution </h4>
</button>
<div id="collapseSolution14" class="accordion-collapse collapse" aria-labelledby="headingSolution14" data-bs-parent="#accordionSolution14">
<div class="accordion-body">
<p>The answer is B: We selected the highest region out of 57 and need to
adjust for multiple testing. Answer A is wrong because we do use normal
approximation in t-test to get a p-value, so there is nothing wrong with
using approximation. Answer B is correct because the p-value that we
obtained is from a comparison against the sample mean (z score = 0)
rather than all other counts. Therefore, the p-value must be corrected
(ex. Bonferroni’s procedure). Answer C is wrong because p value can also
be a random variable, but this answer choice implies that p-value is not
a random variable. Answer D is wrong because effect size is
irrelevant.</p>
</div>
</div>
</div>
</div>
<div id="exercise-15" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-15" class="callout-inner">
<h3 class="callout-title">Exercise 15</h3>
<div class="callout-content">
<p>Use the Bonferonni correction to determine the p-value cut-off that
guarantees a FWER of 0.05. What is this p-value cutoff?</p>
</div>
</div>
</div>
<div id="accordionSolution15" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution15" aria-expanded="false" aria-controls="collapseSolution15">
  <h4 class="accordion-header" id="headingSolution15"> Show me the solution </h4>
</button>
<div id="collapseSolution15" class="accordion-collapse collapse" aria-labelledby="headingSolution15" data-bs-parent="#accordionSolution15">
<div class="accordion-body">
<p><code>0.05/length(counts)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-16" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-16" class="callout-inner">
<h3 class="callout-title">Exercise 16</h3>
<div class="callout-content">
<p>Create a qq-plot to see if our Poisson model is a good fit:</p>
<p><code>ps &lt;- (seq(along=counts) - 0.5)/length(counts)</code><br><code>lambda &lt;- mean( counts[ -which.max(counts)])</code><br><code>poisq &lt;- qpois(ps,lambda)</code><br><code>plot(poisq,sort(counts))</code><br><code>abline(0,1)</code></p>
<p>How would you characterize this qq-plot? A) Poisson is a terrible
approximation.<br>
B) Poisson is a very good approximation except for one point that we
actually think is a region of interest.<br>
C) There are too many 1s in the data.<br>
D) A normal distribution provides a better approximation.</p>
</div>
</div>
</div>
<div id="accordionSolution16" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution16" aria-expanded="false" aria-controls="collapseSolution16">
  <h4 class="accordion-header" id="headingSolution16"> Show me the solution </h4>
</button>
<div id="collapseSolution16" class="accordion-collapse collapse" aria-labelledby="headingSolution16" data-bs-parent="#accordionSolution16">
<div class="accordion-body">
<p>The answer is B. You can check whether the palindrome counts are well
approximated by the normal distribution.<br><code>qqnorm(sort(counts))</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-17" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-17" class="callout-inner">
<h3 class="callout-title">Exercise 17</h3>
<div class="callout-content">
<p>Load the <code>tissuesGeneExpression</code> data library</p>
<p><code>load("data/tissuesGeneExpression.rda")</code></p>
<p>Now load this data and select the columns related to endometrium:</p>
<p><code>library(genefilter)</code><br><code>y = e[,which(tissue=="endometrium")]</code></p>
<p>This will give you a matrix y with 15 samples. Compute the across
sample variance for the first three samples. Then make a qq-plot to see
if the data follow a normal distribution. Which of the following is
true? A) With the exception of a handful of outliers, the data follow a
normal distribution.<br>
B) The variance does not follow a normal distribution, but taking the
square root fixes this.<br>
C) The normal distribution is not usable here: the left tail is over
estimated and the right tail is underestimated.<br>
D) The normal distribution fits the data almost perfectly.</p>
</div>
</div>
</div>
<div id="accordionSolution17" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution17" aria-expanded="false" aria-controls="collapseSolution17">
  <h4 class="accordion-header" id="headingSolution17"> Show me the solution </h4>
</button>
<div id="collapseSolution17" class="accordion-collapse collapse" aria-labelledby="headingSolution17" data-bs-parent="#accordionSolution17">
<div class="accordion-body">
<p><code>vars &lt;- rowVars(y[,1:3])</code><br><code>length(vars)</code><br><code>mypar(1,2)</code><br><code>qqnorm(sqrt(vars))</code> # choice B is false<br><code>qqnorm(vars)</code> # choice A and D are false<br>
The answer is C.</p>
</div>
</div>
</div>
</div>
<div id="exercise-18" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-18" class="callout-inner">
<h3 class="callout-title">Exercise 18</h3>
<div class="callout-content">
<p>Now fit an F-distribution with 14 degrees of freedom using the
fitFDist function in the limma package:</p>
</div>
</div>
</div>
<div id="accordionSolution18" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution18" aria-expanded="false" aria-controls="collapseSolution18">
  <h4 class="accordion-header" id="headingSolution18"> Show me the solution </h4>
</button>
<div id="collapseSolution18" class="accordion-collapse collapse" aria-labelledby="headingSolution18" data-bs-parent="#accordionSolution18">
<div class="accordion-body">
<p><code>library(limma)</code><br><code>res &lt;- fitFDist(vars,14)</code><br><code>res</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-19" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-19" class="callout-inner">
<h3 class="callout-title">Exercise 19</h3>
<div class="callout-content">
<p>Now create a qq-plot of the observed sample variances versus the
F-distribution quantiles. Which of the following best describes the
qq-plot? A) The fitted F-distribution provides a perfect fit.<br>
B) If we exclude the lowest 0.1% of the data, the F-distribution
provides a good fit.<br>
C) The normal distribution provided a better fit.<br>
D) If we exclude the highest 0.1% of the data, the F-distribution
provides a good fit.</p>
</div>
</div>
</div>
<div id="accordionSolution19" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution19" aria-expanded="false" aria-controls="collapseSolution19">
  <h4 class="accordion-header" id="headingSolution19"> Show me the solution </h4>
</button>
<div id="collapseSolution19" class="accordion-collapse collapse" aria-labelledby="headingSolution19" data-bs-parent="#accordionSolution19">
<div class="accordion-body">
<p><code>pf &lt;- (seq(along=vars)-0.5)/length(vars)</code><br><code>theory &lt;- qf(pf,14,res$df2) # theoretical quantiles from F distribution</code><br><code>mypar(1,2)</code><br><code>qqplot(theory, sort(vars), xlab = 'theory', ylab ='obs') # F approximation</code>
<code>vs variance from the data</code>
<code>qqnorm(sort(vars)) # normal approximation vs variance from the data</code><br>
The answer is D: If we exclude the highest 0.1% of the data, the
F-distribution provides a good fit. Actually I do not think answer is
entirely correct but this is the most appropriate one. Even if we
exclude the top 0.1% there are still more outliers to remove. Here is a
demonstration.</p>
<p><code>vars_sort &lt;- sort(vars)</code><br><code>vars_excl &lt;- vars_sort[1:18000]</code> # 18000 out of 22215
kept = 81%<br><code>pf_excl &lt;- (seq(along=vars_excl)-0.5)/length(vars_excl)</code><br><code>theory &lt;- qf(pf_excl,14,res$df2)</code></p>
<p><code>mypar(2,2)</code><br><code>qqplot(theory,vars_excl, xlab = 'theory',ylab='obs',</code><br><code>main = 'all data compared with F-approximation')</code><br><code>qqplot(vars_excl,theory, xlim = c(0,0.2),</code><br><code>ylim = c(0,100), main = 'zoomed in')</code><br><code>qqnorm(vars_excl, main = 'normal qqplot') # comparing with normal</code>
<code>approximation with filtered variance (81% kept)</code><br>
Even if I keep up to 81% of the values, F-distribution</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Not every dataset follows a normal distribution. It is important to
be aware of other parametric distributions underlying the data
collected. This improves the technical precision of study results.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</div></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/smcclatchy/inference-high-dimensional-data/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/smcclatchy/inference-high-dimensional-data/" class="external-link">Source</a></p>
				<p><a href="https://github.com/smcclatchy/inference-high-dimensional-data/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:susan.mcclatchy@jax.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://smcclatchy.github.io/inference-high-dimensional-data/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries, inference, high-dimensional data, biomedical science",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://smcclatchy.github.io/inference-high-dimensional-data/instructor/aio.html",
  "identifier": "https://smcclatchy.github.io/inference-high-dimensional-data/instructor/aio.html",
  "dateCreated": "2025-07-14",
  "dateModified": "2025-08-11",
  "datePublished": "2025-08-11"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

