<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Inference for High-dimensional Data: Statistical Models</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png"><link rel="manifest" href="../favicons/incubator/site.webmanifest"><link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-warning">
          <abbr title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="border-radius: 5px"></i>
              Alpha
            </a>
            <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../statistical-models.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Inference for High-dimensional Data
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Inference for High-dimensional Data
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Inference for High-dimensional Data
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 88%" class="percentage">
    88%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 88%" aria-valuenow="88" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../statistical-models.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Using RMarkdown</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="example-gene-expression-data.html">2. Example Gene Expression Datasets</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="inference-for-highthroughput.html">3. Basic inference for high-throughput data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="procedures-for-multiple-testing.html">4. Procedures for Multiple Comparisons</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="error-rates.html">5. Error Rates</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="bonferroni-correction.html">6. The Bonferroni Correction</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="false-discovery-rate.html">7. False Discovery Rate</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="direct-approach-false-discovery-rate.html">8. Direct Approach to FDR and q-values</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="basic-exploratory-data-analysis.html">9. Basic EDA for high-throughput data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="principal-component-analysis.html">10. Principal Components Analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        11. Statistical Models
        </span>
      
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/principal-component-analysis.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/principal-component-analysis.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Principal Components
        </a>
        <a class="chapter-link float-end" href="../instructor/index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Statistical Models</h1>
        <p>Last updated on 2025-08-11 |

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/statistical-models.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What are some of the most widely used parametric distributions used
in the life sciences besides the normal?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Define the Poisson distribution.</li>
<li>Explain its relevance in RNA sequence analysis.</li>
<li>Explain the purpose of maximum likelihood estimation.</li>
<li>Generate some research questions that maximum likelihood estimation
can address.</li>
<li>Model distributions of gene variances.</li>
</ul></div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="statistical-models">Statistical Models<a class="anchor" aria-label="anchor" href="#statistical-models"></a></h1>
<blockquote>
<p>All models are wrong, but some are useful. -George E. P. Box</p>
</blockquote>
<p>When we see a p-value in the literature, it means a probability
distribution of some sort was used to quantify the null hypothesis. Many
times deciding which probability distribution to use is relatively
straightforward. For example, in the tea tasting challenge we can use
simple probability calculations to determine the null distribution. Most
p-values in the scientific literature are based on sample averages or
least squares estimates from a linear model and make use of the CLT to
approximate the null distribution of their statistic as normal.</p>
<p>The CLT is backed by theoretical results that guarantee that the
approximation is accurate. However, we cannot always use this
approximation, such as when our sample size is too small. Previously, we
described how the sample average can be approximated as t-distributed
when the population data is approximately normal. However, there is no
theoretical backing for this assumption. We are now <em>modeling</em>.
In the case of height, we know from experience that this turns out to be
a very good model.</p>
<p>But this does not imply that every dataset we collect will follow a
normal distribution. Some examples are: coin tosses, the number of
people who win the lottery, and US incomes. The normal distribution is
not the only parametric distribution that is available for modeling.
Here we provide a very brief introduction to some of the most widely
used parametric distributions and some of their uses in the life
sciences. We focus on the models and concepts needed to understand the
techniques currently used to perform statistical inference on
high-throughput data. To do this we also need to introduce the basics of
Bayesian statistics. For more in-depth description of probability models
and parametric distributions please consult a Statistics textbook such
as <a href="https://www.stat.berkeley.edu/~rice/Book3ed/index.html" class="external-link">this
one</a>.</p>
<div class="section level2">
<h2 id="the-binomial-distribution">The Binomial Distribution<a class="anchor" aria-label="anchor" href="#the-binomial-distribution"></a></h2>
<p>The first distribution we will describe is the binomial distribution.
It reports the probability of observing <span class="math inline">\(S=k\)</span> successes in <span class="math inline">\(N\)</span> trials as</p>
<p><span class="math display">\[
\mbox{Pr}(S=k) = {N \choose k}p^k (1-p)^{N-k}
\]</span></p>
<p>with <span class="math inline">\(p\)</span> the probability of
success. The best known example is coin tosses with <span class="math inline">\(S\)</span> the number of heads when tossing <span class="math inline">\(N\)</span> coins. In this example <span class="math inline">\(p=0.5\)</span>.</p>
<p>Note that <span class="math inline">\(S/N\)</span> is the average of
independent random variables and thus the CLT tells us that <span class="math inline">\(S\)</span> is approximately normal when <span class="math inline">\(N\)</span> is large. This distribution has many
applications in the life sciences. Recently, it has been used by the
variant callers and genotypers applied to next generation sequencing. A
special case of this distribution is approximated by the Poisson
distribution which we describe next.</p>
</div>
<div class="section level2">
<h2 id="the-poisson-distribution">The Poisson Distribution<a class="anchor" aria-label="anchor" href="#the-poisson-distribution"></a></h2>
<p>Since it is the sum of binary outcomes, the number of people that win
the lottery follows a binomial distribution (we assume each person buys
one ticket). The number of trials <span class="math inline">\(N\)</span>
is the number of people that buy tickets and is usually very large.
However, the number of people that win the lottery oscillates between 0
and 3, which implies the normal approximation does not hold. So why does
CLT not hold? One can explain this mathematically, but the intuition is
that with the sum of successes so close to and also constrained to be
larger than 0, it is impossible for the distribution to be normal. Here
is a quick simulation:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">=</span><span class="fl">10</span><span class="op">^</span><span class="op">-</span><span class="fl">7</span> <span class="co">##1 in 10,000,0000 chances of winning</span></span>
<span><span class="va">N</span><span class="op">=</span><span class="fl">5</span><span class="op">*</span><span class="fl">10</span><span class="op">^</span><span class="fl">6</span> <span class="co">##5,000,000 tickets bought</span></span>
<span><span class="va">winners</span><span class="op">=</span><span class="fu">rbinom</span><span class="op">(</span><span class="fl">1000</span>,<span class="va">N</span>,<span class="va">p</span><span class="op">)</span> <span class="co">##1000 is the number of different lotto draws</span></span>
<span><span class="va">tab</span><span class="op">=</span><span class="fu">table</span><span class="op">(</span><span class="va">winners</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">tab</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-lottery_winners_outcomes-1.png" alt="Number of people that win the lottery obtained from Monte Carlo simulation." class="figure mx-auto d-block"><figcaption>
Number of people that win the lottery obtained from Monte Carlo
simulation.
</figcaption></figure><div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">prop.table</span><span class="op">(</span><span class="va">tab</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>winners
    0     1     2     3     4
0.618 0.291 0.077 0.012 0.002 </code></pre>
</div>
<p>For cases like this, where <span class="math inline">\(N\)</span> is
very large, but <span class="math inline">\(p\)</span> is small enough
to make <span class="math inline">\(N \times p\)</span> (call it <span class="math inline">\(\lambda\)</span>) a number between 0 and, for
example, 10, then <span class="math inline">\(S\)</span> can be shown to
follow a Poisson distribution, which has a simple parametric form:</p>
<p><span class="math display">\[
\mbox{Pr}(S=k)=\frac{\lambda^k \exp{-\lambda}}{k!}
\]</span></p>
<p>The Poisson distribution is commonly used in RNA-seq analyses.
Because we are sampling thousands of molecules and most genes represent
a very small proportion of the totality of molecules, the Poisson
distribution seems appropriate.</p>
<p>So how does this help us? One way is that it provides insight about
the statistical properties of summaries that are widely used in
practice. For example, let’s say we only have one sample from each of a
case and control RNA-seq experiment and we want to report the genes with
the largest fold-changes. One insight that the Poisson model provides is
that under the null hypothesis of no true differences, the statistical
variability of this quantity depends on the total abundance of the gene.
We can show this mathematically, but here is a quick simulation to
demonstrate the point:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span><span class="op">=</span><span class="fl">10000</span><span class="co">##number of genes</span></span>
<span><span class="va">lambdas</span><span class="op">=</span><span class="fl">2</span><span class="op">^</span><span class="fu">seq</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">16</span>,len<span class="op">=</span><span class="va">N</span><span class="op">)</span> <span class="co">##these are the true abundances of genes</span></span>
<span><span class="va">y</span><span class="op">=</span><span class="fu">rpois</span><span class="op">(</span><span class="va">N</span>,<span class="va">lambdas</span><span class="op">)</span><span class="co">##note that the null hypothesis is true for all genes</span></span>
<span><span class="va">x</span><span class="op">=</span><span class="fu">rpois</span><span class="op">(</span><span class="va">N</span>,<span class="va">lambdas</span><span class="op">)</span> </span>
<span><span class="va">ind</span><span class="op">=</span><span class="fu">which</span><span class="op">(</span><span class="va">y</span><span class="op">&gt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">x</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span><span class="co">##make sure no 0s due to ratio and log</span></span>
<span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">splot</span><span class="op">(</span><span class="fu">log2</span><span class="op">(</span><span class="va">lambdas</span><span class="op">)</span>,<span class="fu">log2</span><span class="op">(</span><span class="va">y</span><span class="op">/</span><span class="va">x</span><span class="op">)</span>,subset<span class="op">=</span><span class="va">ind</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-rna_seq_simulation-1.png" alt="MA plot of simulated RNA-seq data. Replicated measurements follow a Poisson distribution." class="figure mx-auto d-block"><figcaption>
MA plot of simulated RNA-seq data. Replicated measurements follow a
Poisson distribution.
</figcaption></figure><p>For lower values of <code>lambda</code> there is much more
variability and, if we were to report anything with a fold change of 2
or more, the number of false positives would be quite high for low
abundance genes.</p>
<div class="section level4">
<h4 id="ngs-experiments-and-the-poisson-distribution">NGS experiments and the Poisson distribution<a class="anchor" aria-label="anchor" href="#ngs-experiments-and-the-poisson-distribution"></a></h4>
<p>In this section we will use the data stored in this dataset:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">parathyroidSE</span><span class="op">)</span> <span class="co">##available from Bioconductor</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">parathyroidGenesSE</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">parathyroidGenesSE</span></span></code></pre>
</div>
<p>The data is contained in a <code>SummarizedExperiment</code> object,
which we do not describe here. The important thing to know is that it
includes a matrix of data, where each row is a genomic feature and each
column is a sample. We can extract this data using the
<code>assay</code> function. For this dataset, the value of a single
cell in the data matrix is the count of reads which align to a given
gene for a given sample. Thus, a similar plot to the one we simulated
above with technical replicates reveals that the behavior predicted by
the model is present in experimental data:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span>,<span class="fl">23</span><span class="op">]</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span>,<span class="fl">24</span><span class="op">]</span></span>
<span><span class="va">ind</span><span class="op">=</span><span class="fu">which</span><span class="op">(</span><span class="va">y</span><span class="op">&gt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">x</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span><span class="co">##make sure no 0s due to ratio and log</span></span>
<span><span class="fu">splot</span><span class="op">(</span><span class="op">(</span><span class="fu">log2</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">+</span><span class="fu">log2</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>,<span class="fu">log</span><span class="op">(</span><span class="va">x</span><span class="op">/</span><span class="va">y</span><span class="op">)</span>,subset<span class="op">=</span><span class="va">ind</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-RNA-seq_MAplot-1.png" alt="MA plot of replicated RNA-seq data." class="figure mx-auto d-block"><figcaption>
MA plot of replicated RNA-seq data.
</figcaption></figure><p>If we compute the standard deviations across four individuals, it is
quite a bit higher than what is predicted by a Poisson model. Assuming
most genes are differentially expressed across individuals, then, if the
Poisson model is appropriate, there should be a linear relationship in
this plot:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">matrixStats</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vars</span><span class="op">=</span><span class="fu">rowVars</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">8</span>,<span class="fl">16</span>,<span class="fl">21</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="co">##we know these are four</span></span>
<span><span class="va">means</span><span class="op">=</span><span class="fu">rowMeans</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">8</span>,<span class="fl">16</span>,<span class="fl">21</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="co">##different individuals</span></span>
<span></span>
<span><span class="fu">splot</span><span class="op">(</span><span class="va">means</span>,<span class="va">vars</span>,log<span class="op">=</span><span class="st">"xy"</span>,subset<span class="op">=</span><span class="fu">which</span><span class="op">(</span><span class="va">means</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">&amp;</span><span class="va">vars</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="co">##plot a subset of data</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,col<span class="op">=</span><span class="fl">2</span>,lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-var_vs_mean-1.png" alt="Variance versus mean plot. Summaries were obtained from the RNA-seq data." class="figure mx-auto d-block"><figcaption>
Variance versus mean plot. Summaries were obtained from the RNA-seq
data.
</figcaption></figure><p>The reason for this is that the variability plotted here includes
biological variability, which the motivation for the Poisson does not
include. The negative binomial distribution, which combines the sampling
variability of a Poisson and biological variability, is a more
appropriate distribution to model this type of experiment. The negative
binomial has two parameters and permits more flexibility for count data.
For more on the use of the negative binomial to model RNA-seq data you
can read <a href="http://www.ncbi.nlm.nih.gov/pubmed/20979621" class="external-link">this
paper</a>. The Poisson is a special case of the negative binomial
distribution.</p>
</div>
</div>
<div class="section level2">
<h2 id="maximum-likelihood-estimation">Maximum Likelihood Estimation<a class="anchor" aria-label="anchor" href="#maximum-likelihood-estimation"></a></h2>
<p>To illustrate the concept of maximum likelihood estimates (MLE), we
use a relatively simple dataset containing palindrome locations in the
HMCV genome. We read in the locations of the palindrome and then count
the number of palindromes in each 4,000 basepair segments.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">=</span><span class="fu">read.csv</span><span class="op">(</span><span class="st">"./data/hcmv.rda"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `[.data.frame`(read.csv("./data/hcmv.rda"), , 2): undefined columns selected</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">breaks</span><span class="op">=</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">4000</span><span class="op">*</span><span class="fu">round</span><span class="op">(</span><span class="fu">max</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">/</span><span class="fl">4000</span><span class="op">)</span>,<span class="fl">4000</span><span class="op">)</span></span>
<span><span class="va">tmp</span><span class="op">=</span><span class="fu">cut</span><span class="op">(</span><span class="va">x</span>,<span class="va">breaks</span><span class="op">)</span></span>
<span><span class="va">counts</span><span class="op">=</span><span class="fu">table</span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-palindrome_count_histogram-1.png" alt="Palindrome count histogram." class="figure mx-auto d-block"><figcaption>
Palindrome count histogram.
</figcaption></figure><p>The counts do appear to follow a Poisson distribution. But what is
the rate <span class="math inline">\(\lambda\)</span> ? The most common
approach to estimating this rate is <em>maximum likelihood
estimation</em>. To find the maximum likelihood estimate (MLE), we note
that these data are independent and the probability of observing the
values we observed is:</p>
<p><span class="math display">\[
\Pr(X_1=k_1,\dots,X_n=k_n;\lambda) = \prod_{i=1}^n \lambda^{k_i} / k_i!
\exp ( -\lambda)
\]</span></p>
<p>The MLE is the value of <span class="math inline">\(\lambda\)</span>
that maximizes the likelihood:.</p>
<p><span class="math display">\[
\mbox{L}(\lambda; X_1=k_1,\dots,X_n=k_1)=\exp\left\{\sum_{i=1}^n \log
\Pr(X_i=k_i;\lambda)\right\}
\]</span></p>
<p>In practice, it is more convenient to maximize the log-likelihood
which is the summation that is exponentiated in the expression above.
Below we write code that computes the log-likelihood for any <span class="math inline">\(\lambda\)</span> and use the function
<code>optimize</code> to find the value that maximizes this function
(the MLE). We show a plot of the log-likelihood along with vertical line
showing the MLE.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span><span class="op">&lt;-</span><span class="kw">function</span><span class="op">(</span><span class="va">lambda</span><span class="op">)</span> <span class="fu">sum</span><span class="op">(</span><span class="fu">dpois</span><span class="op">(</span><span class="va">counts</span>,<span class="va">lambda</span>,log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">lambdas</span><span class="op">&lt;-</span><span class="fu">seq</span><span class="op">(</span><span class="fl">3</span>,<span class="fl">7</span>,len<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">ls</span> <span class="op">&lt;-</span> <span class="fu">exp</span><span class="op">(</span><span class="fu">sapply</span><span class="op">(</span><span class="va">lambdas</span>,<span class="va">l</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">lambdas</span>,<span class="va">ls</span>,type<span class="op">=</span><span class="st">"l"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mle</span><span class="op">=</span><span class="fu">optimize</span><span class="op">(</span><span class="va">l</span>,<span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">10</span><span class="op">)</span>,maximum<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v<span class="op">=</span><span class="va">mle</span><span class="op">$</span><span class="va">maximum</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-mle-1.png" alt="Likelihood versus lambda." class="figure mx-auto d-block"><figcaption>
Likelihood versus lambda.
</figcaption></figure><p>If you work out the math and do a bit of calculus, you realize that
this is a particularly simple example for which the MLE is the
average.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">print</span><span class="op">(</span> <span class="fu">c</span><span class="op">(</span><span class="va">mle</span><span class="op">$</span><span class="va">maximum</span>, <span class="fu">mean</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">)</span> <span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1]    9.999944 1279.750000</code></pre>
</div>
<p>Note that a plot of observed counts versus counts predicted by the
Poisson shows that the fit is quite good in this case:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">theoretical</span><span class="op">&lt;-</span><span class="fu">qpois</span><span class="op">(</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">99</span><span class="op">)</span><span class="op">+</span><span class="fl">0.5</span><span class="op">)</span><span class="op">/</span><span class="fl">100</span>,<span class="fu">mean</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">qqplot</span><span class="op">(</span><span class="va">theoretical</span>,<span class="va">counts</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-obs_versus_theoretical_Poisson_count-1.png" alt="Observed counts versus theoretical Poisson counts." class="figure mx-auto d-block"><figcaption>
Observed counts versus theoretical Poisson counts.
</figcaption></figure><p>We therefore can model the palindrome count data with a Poisson with
<span class="math inline">\(\lambda=5.16\)</span>.</p>
</div>
<div class="section level2">
<h2 id="distributions-for-positive-continuous-values">Distributions for Positive Continuous Values<a class="anchor" aria-label="anchor" href="#distributions-for-positive-continuous-values"></a></h2>
<p>Different genes vary differently across biological replicates. Later,
in the hierarchical models chapter, we will describe one of the <a href="http://www.ncbi.nlm.nih.gov/pubmed/16646809" class="external-link">most influential
statistical methods</a> in the analysis of genomics data. This method
provides great improvements over naive approaches to detecting
differentially expressed genes. This is achieved by modeling the
distribution of the gene variances. Here we describe the parametric
model used in this method.</p>
<p>We want to model the distribution of the gene-specific standard
errors. Are they normal? Keep in mind that we are modeling the
population standard errors so CLT does not apply, even though we have
thousands of genes.</p>
<p>As an example, we use an experimental data that included both
technical and biological replicates for gene expression measurements on
mice. We can load the data and compute the gene specific sample standard
error for both the technical replicates and the biological
replicates.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">Biobase</span><span class="op">)</span> <span class="co">##available from Bioconductor</span></span>
<span><span class="fu">load</span><span class="op">(</span><span class="st">"data/maPooling.RData"</span><span class="op">)</span></span>
<span><span class="va">pd</span><span class="op">=</span><span class="fu">pData</span><span class="op">(</span><span class="va">maPooling</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##determine which samples are bio reps and which are tech reps</span></span>
<span><span class="va">strain</span><span class="op">=</span><span class="fu">factor</span><span class="op">(</span><span class="fu">as.numeric</span><span class="op">(</span><span class="fu">grepl</span><span class="op">(</span><span class="st">"b"</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">pd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pooled</span><span class="op">=</span><span class="fu">which</span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="va">pd</span><span class="op">)</span><span class="op">==</span><span class="fl">12</span> <span class="op">&amp;</span> <span class="va">strain</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">techreps</span><span class="op">=</span><span class="fu">exprs</span><span class="op">(</span><span class="va">maPooling</span><span class="op">[</span>,<span class="va">pooled</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">individuals</span><span class="op">=</span><span class="fu">which</span><span class="op">(</span><span class="fu">rowSums</span><span class="op">(</span><span class="va">pd</span><span class="op">)</span><span class="op">==</span><span class="fl">1</span> <span class="op">&amp;</span> <span class="va">strain</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##remove replicates</span></span>
<span><span class="va">individuals</span><span class="op">=</span><span class="va">individuals</span><span class="op">[</span><span class="op">-</span><span class="fu">grep</span><span class="op">(</span><span class="st">"tr"</span>, <span class="fu">names</span><span class="op">(</span><span class="va">individuals</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">bioreps</span><span class="op">=</span><span class="fu">exprs</span><span class="op">(</span><span class="va">maPooling</span><span class="op">)</span><span class="op">[</span>,<span class="va">individuals</span><span class="op">]</span></span>
<span></span>
<span><span class="co">###now compute the gene specific standard deviations</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">matrixStats</span><span class="op">)</span></span>
<span><span class="va">techsds</span><span class="op">=</span><span class="fu">rowSds</span><span class="op">(</span><span class="va">techreps</span><span class="op">)</span></span>
<span><span class="va">biosds</span><span class="op">=</span><span class="fu">rowSds</span><span class="op">(</span><span class="va">bioreps</span><span class="op">)</span></span></code></pre>
</div>
<p>We can now explore the sample standard deviation:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">###now plot</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">shist</span><span class="op">(</span><span class="va">biosds</span>,unit<span class="op">=</span><span class="fl">0.1</span>,col<span class="op">=</span><span class="fl">1</span>,xlim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">shist</span><span class="op">(</span><span class="va">techsds</span>,unit<span class="op">=</span><span class="fl">0.1</span>,col<span class="op">=</span><span class="fl">2</span>,add<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">legend</span><span class="op">(</span><span class="st">"topright"</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"Biological"</span>,<span class="st">"Technical"</span><span class="op">)</span>, col<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,lty<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-bio_sd_versus_tech_sd-1.png" alt="Histograms of biological variance and technical variance." class="figure mx-auto d-block"><figcaption>
Histograms of biological variance and technical variance.
</figcaption></figure><p>An important observation here is that the biological variability is
substantially higher than the technical variability. This provides
strong evidence that genes do in fact have gene-specific biological
variability.</p>
<p>If we want to model this variability, we first notice that the normal
distribution is not appropriate here since the right tail is rather
large. Also, because SDs are strictly positive, there is a limitation to
how symmetric this distribution can be. A qqplot shows this very
clearly:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">qqnorm</span><span class="op">(</span><span class="va">biosds</span><span class="op">)</span></span>
<span><span class="fu">qqline</span><span class="op">(</span><span class="va">biosds</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-sd_qqplot-1.png" alt="Normal qq-plot for sample standard deviations." class="figure mx-auto d-block"><figcaption>
Normal qq-plot for sample standard deviations.
</figcaption></figure><p>There are parametric distributions that possess these properties
(strictly positive and <em>heavy</em> right tails). Two examples are the
<em>gamma</em> and <em>F</em> distributions. The density of the gamma
distribution is defined by:</p>
<p><span class="math display">\[
f(x;\alpha,\beta)=\frac{\beta^\alpha x^{\alpha-1}\exp{-\beta
x}}{\Gamma(\alpha)}
\]</span></p>
<p>It is defined by two parameters <span class="math inline">\(\alpha\)</span> and <span class="math inline">\(\beta\)</span> that can, indirectly, control
location and scale. They also control the shape of the distribution. For
more on this distribution please refer to <a href="https://www.stat.berkeley.edu/~rice/Book3ed/index.html" class="external-link">this
book</a>.</p>
<p>Two special cases of the gamma distribution are the chi-squared and
exponential distribution. We used the chi-squared earlier to analyze a
2x2 table data. For chi-square, we have <span class="math inline">\(\alpha=\nu/2\)</span> and <span class="math inline">\(\beta=2\)</span> with <span class="math inline">\(\nu\)</span> the degrees of freedom. For
exponential, we have <span class="math inline">\(\alpha=1\)</span> and
<span class="math inline">\(\beta=\lambda\)</span> the rate.</p>
<p>The F-distribution comes up in analysis of variance (ANOVA). It is
also always positive and has large right tails. Two parameters control
its shape:</p>
<p><span class="math display">\[
f(x,d_1,d_2)=\frac{1}{B\left( \frac{d_1}{2},\frac{d_2}{2}\right)}
  \left(\frac{d_1}{d_2}\right)^{\frac{d_1}{2}}
  x^{\frac{d_1}{2}-1}\left(1+\frac{d1}{d2}x\right)^{-\frac{d_1+d_2}{2}}
\]</span></p>
<p>with <span class="math inline">\(B\)</span> the <em>beta
function</em> and <span class="math inline">\(d_1\)</span> and <span class="math inline">\(d_2\)</span> are called the degrees of freedom for
reasons having to do with how it arises in ANOVA. A third parameter is
sometimes used with the F-distribution, which is a scale parameter.</p>
<div class="section level4">
<h4 id="modeling-the-variance">Modeling the variance<a class="anchor" aria-label="anchor" href="#modeling-the-variance"></a></h4>
<p>In a later section we will learn about a hierarchical model approach
to improve estimates of variance. In these cases it is mathematically
convenient to model the distribution of the variance <span class="math inline">\(\sigma^2\)</span>. The hierarchical model used <a href="http://www.ncbi.nlm.nih.gov/pubmed/16646809" class="external-link">here</a> implies that
the sample standard deviation of genes follows scaled F-statistics:</p>
<p><span class="math display">\[
s^2 \sim s_0^2 F_{d,d_0}
\]</span></p>
<p>with <span class="math inline">\(d\)</span> the degrees of freedom
involved in computing <span class="math inline">\(s^2\)</span>. For
example, in a case comparing 3 versus 3, the degrees of freedom would be
4. This leaves two free parameters to adjust to the data. Here <span class="math inline">\(d\)</span> will control the location and <span class="math inline">\(s_0\)</span> will control the scale. Below are
some examples of <span class="math inline">\(F\)</span> distribution
plotted on top of the histogram from the sample variances:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">3</span>,<span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">sds</span><span class="op">=</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">2</span>,len<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">d</span> <span class="kw">in</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">5</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">s0</span> <span class="kw">in</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">tmp</span><span class="op">=</span><span class="fu">hist</span><span class="op">(</span><span class="va">biosds</span>,main<span class="op">=</span><span class="fu">paste</span><span class="op">(</span><span class="st">"s_0 ="</span>,<span class="va">s0</span>,<span class="st">"d ="</span>,<span class="va">d</span><span class="op">)</span>,</span>
<span>      xlab<span class="op">=</span><span class="st">"sd"</span>,ylab<span class="op">=</span><span class="st">"density"</span>,freq<span class="op">=</span><span class="cn">FALSE</span>,nc<span class="op">=</span><span class="fl">100</span>,xlim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">dd</span><span class="op">=</span><span class="fu">df</span><span class="op">(</span><span class="va">sds</span><span class="op">^</span><span class="fl">2</span><span class="op">/</span><span class="va">s0</span><span class="op">^</span><span class="fl">2</span>,<span class="fl">11</span>,<span class="va">d</span><span class="op">)</span></span>
<span>    <span class="co">##multiply by normalizing constant to assure same range on plot</span></span>
<span>    <span class="va">k</span><span class="op">=</span><span class="fu">sum</span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">density</span><span class="op">)</span><span class="op">/</span><span class="fu">sum</span><span class="op">(</span><span class="va">dd</span><span class="op">)</span> </span>
<span>    <span class="fu">lines</span><span class="op">(</span><span class="va">sds</span>,<span class="va">dd</span><span class="op">*</span><span class="va">k</span>,type<span class="op">=</span><span class="st">"l"</span>,col<span class="op">=</span><span class="fl">2</span>,lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-modeling_variance-1.png" alt="Histograms of sample standard deviations and densities of estimated distributions." class="figure mx-auto d-block"><figcaption>
Histograms of sample standard deviations and densities of estimated
distributions.
</figcaption></figure><p>Now which <span class="math inline">\(s_0\)</span> and <span class="math inline">\(d\)</span> fit our data best? This is a rather
advanced topic as the MLE does not perform well for this particular
distribution (we refer to Smyth (2004)). The Bioconductor
<code>limma</code> package provides a function to estimate these
parameters:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">limma</span><span class="op">)</span></span>
<span><span class="va">estimates</span><span class="op">=</span><span class="fu">fitFDist</span><span class="op">(</span><span class="va">biosds</span><span class="op">^</span><span class="fl">2</span>, <span class="fl">11</span><span class="op">)</span></span>
<span></span>
<span><span class="va">theoretical</span><span class="op">&lt;-</span> <span class="fu">sqrt</span><span class="op">(</span><span class="fu">qf</span><span class="op">(</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">999</span><span class="op">)</span> <span class="op">+</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">/</span><span class="fl">1000</span>, <span class="fl">11</span>, <span class="va">estimates</span><span class="op">$</span><span class="va">df2</span><span class="op">)</span> <span class="op">*</span> <span class="va">estimates</span><span class="op">$</span><span class="va">scale</span><span class="op">)</span></span>
<span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="va">biosds</span></span></code></pre>
</div>
<p>The fitted models do appear to provide a reasonable approximation, as
demonstrated by the qq-plot and histogram:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">qqplot</span><span class="op">(</span><span class="va">theoretical</span>,<span class="va">observed</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tmp</span><span class="op">=</span><span class="fu">hist</span><span class="op">(</span><span class="va">biosds</span>,main<span class="op">=</span><span class="fu">paste</span><span class="op">(</span><span class="st">"s_0 ="</span>, <span class="fu">signif</span><span class="op">(</span><span class="va">estimates</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>                  <span class="st">"d ="</span>, <span class="fu">signif</span><span class="op">(</span><span class="va">estimates</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  xlab<span class="op">=</span><span class="st">"sd"</span>, ylab<span class="op">=</span><span class="st">"density"</span>, freq<span class="op">=</span><span class="cn">FALSE</span>, nc<span class="op">=</span><span class="fl">100</span>, xlim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">9</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dd</span><span class="op">=</span><span class="fu">df</span><span class="op">(</span><span class="va">sds</span><span class="op">^</span><span class="fl">2</span><span class="op">/</span><span class="va">estimates</span><span class="op">$</span><span class="va">scale</span>,<span class="fl">11</span>,<span class="va">estimates</span><span class="op">$</span><span class="va">df2</span><span class="op">)</span></span>
<span><span class="va">k</span><span class="op">=</span><span class="fu">sum</span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">density</span><span class="op">)</span><span class="op">/</span><span class="fu">sum</span><span class="op">(</span><span class="va">dd</span><span class="op">)</span> <span class="co">##a normalizing constant to assure same area in plot</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">sds</span>, <span class="va">dd</span><span class="op">*</span><span class="va">k</span>, type<span class="op">=</span><span class="st">"l"</span>, col<span class="op">=</span><span class="fl">2</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/statistical-models-rendered-variance_model_fit-1.png" alt="qq-plot (left) and density (right) demonstrate that model fits data well." class="figure mx-auto d-block"><figcaption>
qq-plot (left) and density (right) demonstrate that model fits data
well.
</figcaption></figure></div>
</div>
<div class="section level2">
<h2 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a></h2>
<p>Suppose you have an urn with blue and red balls. If N balls are
selected at random with replacement (you put the ball back after you
pick it), we can denote the outcomes as random variables X1,…,XN that
are 1 or 0. If the proportion of red balls is p , then the distribution
of each of these is Pr(Xi=1)=p.</p>
<p>These are also called Bernoulli trials. These random variables are
independent because we replace the balls. Flipping a coin is an example
of this with p=0.5.</p>
<p>You can show that the mean and variance are p and p(1−p)
respectively. The binomial distribution gives us the distribution of the
sum SN of these random variables. The probability that we see k red
balls is given by: Pr(SN=k)=(Nk)pk(1−p)N−k</p>
<p>In R, the function dbimom gives you this result. The function pbinom
gives us Pr(SN≤k).</p>
<p>This equation has many uses in the life sciences. We give some
examples below.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>The probability of conceiving a girl is 0.49. What is the probability
that a family with 4 children has 2 girls and 2 boys (you can assume
that the outcomes are independent)?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p><code>g &lt;- sampleInfo$group</code><br><code>pvals &lt;- rowttests(geneExpression, factor(g))$p.value</code><br><code>sum(pvals &lt; 0.05)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>Use what you learned in Question 1 to answer these questions: What is
the probability that a family with 10 children has 6 girls and 4 boys
(assume no twins)?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p><code>p &lt;- 0.49</code><br><code>dbinom(6,10,p=p)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>The genome has 3 billion bases. About 20% are C, 20% are G, 30% are
T, and 30% are A. Suppose you take a random interval of 20 bases, what
is the probability that the GC-content (proportion of Gs of Cs) is
strictly above 0.5 in this interval?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p><code>p &lt;- 0.4</code><br><code>pbinom(20,20,p) - pbinom(10,20,p)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>The probability of winning the lottery is 1 in 175,223,510. If
20,000,000 people buy a ticket, what is the probability that more than
one person wins?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p><code>p &lt;- 1/175233510</code><br><code>N &lt;- 20000000</code><br><code>1 - ppois(1,N*p)</code><br><code>1 - pbinom(1,N,p)</code><br>
Since the poisson distribution is a type of binomial distribution, both
distributions give the same values.</p>
</div>
</div>
</div>
</div>
<div id="exercise-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-5" class="callout-inner">
<h3 class="callout-title">Exercise 5</h3>
<div class="callout-content">
<p>We can show that the binomial approximation is approximately normal
when N is large and p is not too close to 0 or 1. This means that:</p>
<p>SN−E(SN)Var(SN)‾‾‾‾‾‾‾‾√</p>
<p>is approximately normal with mean 0 and SD 1. Using the results for
sums of independent random variables, we can show that E(SN)=Np and
Var(Sn)=Np(1−p). The genome has 3 billion bases. About 20% are C, 20%
are G, 30% are T, and 30% are A. Suppose you take a random interval of
20 bases, what is the exact probability that the GC-content (proportion
of Gs of Cs) is greater than 0.35 and smaller or equal to 0.45 in this
interval?</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<p><code>p &lt;- 0.4</code><br><code>pbinom(0.45*20, 20, p) - pbinom(0.35*20,20,p)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-6" class="callout-inner">
<h3 class="callout-title">Exercise 6</h3>
<div class="callout-content">
<p>For the question above, what is the normal approximation to the
probability?</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<p><code>p &lt;- 0.4</code><br><code>N &lt;- 20</code><br><code>a &lt;- (0.45*20 - N*p) / sqrt(N*p*(1-p))</code><br><code>b &lt;- (0.35*20 - N*p) / sqrt(N*p*(1-p))</code></p>
<p><code>approx &lt;- pnorm(a) - pnorm(b)</code><br><code>approx</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-7" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-7" class="callout-inner">
<h3 class="callout-title">Exercise 7</h3>
<div class="callout-content">
<p>Repeat exercise 5, but using an interval of 1000 bases. What is the
difference (in absolute value) between the normal approximation and the
exact distribution of the GC-content being greater than 0.35 and lesser
or equal to 0.45?</p>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" data-bs-parent="#accordionSolution7" aria-labelledby="headingSolution7">
<div class="accordion-body">
<p><code>N &lt;- 1000</code><br><code>p &lt;- 0.4</code><br><code>a &lt;- (0.45*N - N*p) / sqrt(N*p*(1-p))</code><br><code>b &lt;- (0.35*N - N*p) / sqrt(N*p*(1-p))</code><br><code>approx &lt;- pnorm(a) - pnorm(b)</code></p>
<p><code>exact &lt;- pbinom(0.45*N,N,p) - pbinom(0.35*N,N,p)</code><br><code>exact - approx</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-8" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-8" class="callout-inner">
<h3 class="callout-title">Exercise 8</h3>
<div class="callout-content">
<p>The Cs in our genomes can be methylated or unmethylated. Suppose we
have a large (millions) group of cells in which a proportion p of the Cs
of interest are methylated. We break up the DNA of these cells and
randomly select pieces and end up with N pieces that contain the C we
care about. This means that the probability of seeing k methylated Cs is
binomial: exact = dbinom(k,N,p) We can approximate this with the normal
distribution: <code>a &lt;- (k+0.5 - N*p)/sqrt(N*p*(1-p))</code><br><code>b &lt;- (k-0.5 - N*p)/sqrt(N*p*(1-p))</code><br><code>approx = pnorm(a) - pnorm(b)</code><br>
Compute the difference approx - exact for:<br><code>N &lt;- c(5,10,50,100,500)</code><br><code>p &lt;- seq(0,1,0.25)</code><br>
Compare the approximation and exact probability of the proportion of Cs
being p, k=1,…,N−1 plotting the exact versus the approximation for each
p and N combination. A) The normal approximation works well when p is
close to 0.5 even for small N=10<br>
B) The normal approximation breaks down when p is close to 0 or 1 even
for large N<br>
C) When N is 100 all approximations are spot on.<br>
D) When p=0.01 the approximation are terrible for N=5,10,30 and only OK
for N=100</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" data-bs-parent="#accordionSolution8" aria-labelledby="headingSolution8">
<div class="accordion-body">
<p><code>Ns &lt;- c(5,10,50,100,500)</code><br><code>Ps &lt;- seq(0,1,0.25)</code><br><code>mypar(5,5)</code><br><code>for (i in seq_along(Ns)) {</code><br><code>n &lt;- Ns[[i]]</code><br><code>k &lt;- seq(1:n-1)</code><br><code>for (j in seq_along(Ps)) {</code><br><code>p &lt;- Ps[[j]]</code><br><code>exact = dbinom(k, n, p)</code><br><code>a = (k+0.5- n*p)/sqrt(n*p*(1-p))</code><br><code>b = (k-0.5- n*p)/sqrt(n*p*(1-p))</code><br><code>approx = pnorm(a) - pnorm(b)</code><br><code>qqplot(exact,approx,xlab='exact',ylab='approx',</code><br><code>main = paste0('N=',n,' ','p=',p))</code><br><code>abline(0,1)</code><br><code>}</code><br><code>}</code><br>
The answer is C: When N is 100 all approximations are spot on. When p is
close to 0 or 1, the normal distribution breaks down even at N =
100.</p>
</div>
</div>
</div>
</div>
<div id="exercise-9" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-9" class="callout-inner">
<h3 class="callout-title">Exercise 9</h3>
<div class="callout-content">
<p>We saw in the previous question that when p is very small, the normal
approximation breaks down. If N is very large, then we can use the
Poisson approximation.<br>
Earlier we computed 1 or more people winning the lottery when the
probability of winning was 1 in 175,223,510 and 20,000,000 people bought
a ticket. Using the binomial, we can compute the probability of exactly
two people winning to be:</p>
<p><code>N &lt;- 20000000</code> <code>p &lt;- 1/175223510</code>
<code>dbinom(2,N,p)</code></p>
<p>If we were to use the normal approximation, we would greatly
underestimate this:</p>
<p><code>a &lt;- (2+0.5 - N*p)/sqrt(N*p*(1-p))</code>
<code>b &lt;- (2-0.5 - N*p)/sqrt(N*p*(1-p))</code>
<code>pnorm(a) - pnorm(b)</code></p>
<p>To use the Poisson approximation here, use the rate λ=Np</p>
<p>representing the number of people per 20,000,000 that win the
lottery. Note how much better the approximation is:</p>
<p><code>dpois(2,N*p)</code></p>
<p>In this case. it is practically the same because N is very large and
Np is not 0. These are the assumptions needed for the Poisson to work.
What is the Poisson approximation for more than one person winning?</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9"> Show me the solution </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" data-bs-parent="#accordionSolution9" aria-labelledby="headingSolution9">
<div class="accordion-body">
<p><code>1 - ppois(1,N*p)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-10" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-10" class="callout-inner">
<h3 class="callout-title">Exercise 10</h3>
<div class="callout-content">
<p>Now we are going to explore if palindromes are over-represented in
some part of the HCMV genome. Make sure you have the latest version of
the dagdata, load the palindrome data from the Human cytomegalovirus
genome, and plot locations of palindromes on the genome for this
virus:</p>
<p><code>library(dagdata)</code><br><code>data(hcmv)</code><br><code>plot(locations,rep(1,length(locations)),ylab="",yaxt="n")</code></p>
<p>These palindromes are quite rare, and therefore p is very small. If
we break the genome into bins of 4000 basepairs, then we have Np not so
small and we might be able to use Poisson to model the number of
palindromes in each bin:</p>
<p><code>breaks=seq(0,4000*round(max(locations)/4000),4000)</code>
<code>tmp=cut(locations,breaks)</code>
<code>counts=as.numeric(table(tmp))</code></p>
<p>So if our model is correct, counts should follow a Poisson
distribution. The distribution seems about right:</p>
<p><code>hist(counts)</code></p>
<p>So let X1,…,Xn be the random variables representing counts then
Pr(Xi=k)=λk/k!exp(−λ) and to fully describe this distribution, we need
to know λ. For this we will use MLE. We can write the likelihood
described in book in R. For example, for λ=4 we have:</p>
<p><code>probs &lt;- dpois(counts,4)</code>
<code>likelihood &lt;- prod(probs)</code> <code>likelihood</code></p>
<p>Notice that it’s a tiny number. It is usually more convenient to
compute log-likelihoods:</p>
<p><code>logprobs &lt;- dpois(counts,4,log=TRUE)</code><br><code>loglikelihood &lt;- sum(logprobs)</code><br><code>loglikelihood</code></p>
<p>Now write a function that takes λ and the vector of counts as input
and returns the log-likelihood. Compute this log-likelihood for lambdas
= <code>seq(0,15,len=300)</code> and make a plot. What value of lambdas
maximizes the log-likelihood?</p>
</div>
</div>
</div>
<div id="accordionSolution10" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution10" aria-expanded="false" aria-controls="collapseSolution10">
  <h4 class="accordion-header" id="headingSolution10"> Show me the solution </h4>
</button>
<div id="collapseSolution10" class="accordion-collapse collapse" data-bs-parent="#accordionSolution10" aria-labelledby="headingSolution10">
<div class="accordion-body">
<p><code>loglike = function(lambdas) {</code><br><code>logprobs &lt;- dpois(counts,lambdas,log=T)</code><br><code>loglikelihood &lt;- sum(logprobs)</code><br><code>return(loglikelihood)</code><br><code>}</code><br><code>lambdas &lt;- seq(0,15,len=300)</code><br><code>log_res &lt;- exp(sapply(lambdas,loglike))</code><br><code>optim_lambda &lt;- lambdas[which(log_res == max(log_res))]</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-11" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-11" class="callout-inner">
<h3 class="callout-title">Exercise 11</h3>
<div class="callout-content">
<p>The point of collecting this dataset was to try to determine if there
is a region of the genome that has a higher palindrome rate than
expected. We can create a plot and see the counts per location:</p>
<p><code>library(dagdata)</code><br><code>data(hcmv)</code><br><code>breaks=seq(0,4000*round(max(locations)/4000),4000)</code><br><code>tmp=cut(locations,breaks)</code><br><code>counts=as.numeric(table(tmp))</code><br><code>binLocation=(breaks[-1]+breaks[-length(breaks)])/2</code><br><code>plot(binLocation,counts,type="l",xlab=)</code></p>
<p>What is the center of the bin with the highest count?</p>
</div>
</div>
</div>
<div id="accordionSolution11" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution11" aria-expanded="false" aria-controls="collapseSolution11">
  <h4 class="accordion-header" id="headingSolution11"> Show me the solution </h4>
</button>
<div id="collapseSolution11" class="accordion-collapse collapse" data-bs-parent="#accordionSolution11" aria-labelledby="headingSolution11">
<div class="accordion-body">
<p><code>binLocation[which(counts == max(counts))]</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-12" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-12" class="callout-inner">
<h3 class="callout-title">Exercise 12</h3>
<div class="callout-content">
<p>What is the maximum count?</p>
</div>
</div>
</div>
<div id="accordionSolution12" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution12" aria-expanded="false" aria-controls="collapseSolution12">
  <h4 class="accordion-header" id="headingSolution12"> Show me the solution </h4>
</button>
<div id="collapseSolution12" class="accordion-collapse collapse" data-bs-parent="#accordionSolution12" aria-labelledby="headingSolution12">
<div class="accordion-body">
<p><code>max(counts)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-13" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-13" class="callout-inner">
<h3 class="callout-title">Exercise 13</h3>
<div class="callout-content">
<p>Once we have identified the location with the largest palindrome
count, we want to know if we could see a value this big by chance. If X
is a Poisson random variable with rate:</p>
<p><code>lambda = mean(counts[ - which.max(counts) ])</code></p>
<p>What is the probability of seeing a count of 14 or more?</p>
</div>
</div>
</div>
<div id="accordionSolution13" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution13" aria-expanded="false" aria-controls="collapseSolution13">
  <h4 class="accordion-header" id="headingSolution13"> Show me the solution </h4>
</button>
<div id="collapseSolution13" class="accordion-collapse collapse" data-bs-parent="#accordionSolution13" aria-labelledby="headingSolution13">
<div class="accordion-body">
<p><code>1 - ppois(13, lambda)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-14" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-14" class="callout-inner">
<h3 class="callout-title">Exercise 14</h3>
<div class="callout-content">
<p>So we obtain a p-value smaller than 0.001 for a count of 14. Why is
it problematic to report this p-value as strong evidence of a location
that is different?</p>
<ol style="list-style-type: upper-alpha"><li>Poisson in only an approximation.<br></li>
<li>We selected the highest region out of 57 and need to adjust for
multiple testing.<br></li>
<li>λ is an estimate, a random variable, and we didn’t take into account
its variability.<br></li>
<li>We don’t know the effect size.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution14" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution14" aria-expanded="false" aria-controls="collapseSolution14">
  <h4 class="accordion-header" id="headingSolution14"> Show me the solution </h4>
</button>
<div id="collapseSolution14" class="accordion-collapse collapse" data-bs-parent="#accordionSolution14" aria-labelledby="headingSolution14">
<div class="accordion-body">
<p>The answer is B: We selected the highest region out of 57 and need to
adjust for multiple testing. Answer A is wrong because we do use normal
approximation in t-test to get a p-value, so there is nothing wrong with
using approximation. Answer B is correct because the p-value that we
obtained is from a comparison against the sample mean (z score = 0)
rather than all other counts. Therefore, the p-value must be corrected
(ex. Bonferroni’s procedure). Answer C is wrong because p value can also
be a random variable, but this answer choice implies that p-value is not
a random variable. Answer D is wrong because effect size is
irrelevant.</p>
</div>
</div>
</div>
</div>
<div id="exercise-15" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-15" class="callout-inner">
<h3 class="callout-title">Exercise 15</h3>
<div class="callout-content">
<p>Use the Bonferonni correction to determine the p-value cut-off that
guarantees a FWER of 0.05. What is this p-value cutoff?</p>
</div>
</div>
</div>
<div id="accordionSolution15" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution15" aria-expanded="false" aria-controls="collapseSolution15">
  <h4 class="accordion-header" id="headingSolution15"> Show me the solution </h4>
</button>
<div id="collapseSolution15" class="accordion-collapse collapse" data-bs-parent="#accordionSolution15" aria-labelledby="headingSolution15">
<div class="accordion-body">
<p><code>0.05/length(counts)</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-16" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-16" class="callout-inner">
<h3 class="callout-title">Exercise 16</h3>
<div class="callout-content">
<p>Create a qq-plot to see if our Poisson model is a good fit:</p>
<p><code>ps &lt;- (seq(along=counts) - 0.5)/length(counts)</code><br><code>lambda &lt;- mean( counts[ -which.max(counts)])</code><br><code>poisq &lt;- qpois(ps,lambda)</code><br><code>plot(poisq,sort(counts))</code><br><code>abline(0,1)</code></p>
<p>How would you characterize this qq-plot? A) Poisson is a terrible
approximation.<br>
B) Poisson is a very good approximation except for one point that we
actually think is a region of interest.<br>
C) There are too many 1s in the data.<br>
D) A normal distribution provides a better approximation.</p>
</div>
</div>
</div>
<div id="accordionSolution16" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution16" aria-expanded="false" aria-controls="collapseSolution16">
  <h4 class="accordion-header" id="headingSolution16"> Show me the solution </h4>
</button>
<div id="collapseSolution16" class="accordion-collapse collapse" data-bs-parent="#accordionSolution16" aria-labelledby="headingSolution16">
<div class="accordion-body">
<p>The answer is B. You can check whether the palindrome counts are well
approximated by the normal distribution.<br><code>qqnorm(sort(counts))</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-17" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-17" class="callout-inner">
<h3 class="callout-title">Exercise 17</h3>
<div class="callout-content">
<p>Load the <code>tissuesGeneExpression</code> data library</p>
<p><code>load("data/tissuesGeneExpression.rda")</code></p>
<p>Now load this data and select the columns related to endometrium:</p>
<p><code>library(genefilter)</code><br><code>y = e[,which(tissue=="endometrium")]</code></p>
<p>This will give you a matrix y with 15 samples. Compute the across
sample variance for the first three samples. Then make a qq-plot to see
if the data follow a normal distribution. Which of the following is
true? A) With the exception of a handful of outliers, the data follow a
normal distribution.<br>
B) The variance does not follow a normal distribution, but taking the
square root fixes this.<br>
C) The normal distribution is not usable here: the left tail is over
estimated and the right tail is underestimated.<br>
D) The normal distribution fits the data almost perfectly.</p>
</div>
</div>
</div>
<div id="accordionSolution17" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution17" aria-expanded="false" aria-controls="collapseSolution17">
  <h4 class="accordion-header" id="headingSolution17"> Show me the solution </h4>
</button>
<div id="collapseSolution17" class="accordion-collapse collapse" data-bs-parent="#accordionSolution17" aria-labelledby="headingSolution17">
<div class="accordion-body">
<p><code>vars &lt;- rowVars(y[,1:3])</code><br><code>length(vars)</code><br><code>mypar(1,2)</code><br><code>qqnorm(sqrt(vars))</code> # choice B is false<br><code>qqnorm(vars)</code> # choice A and D are false<br>
The answer is C.</p>
</div>
</div>
</div>
</div>
<div id="exercise-18" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-18" class="callout-inner">
<h3 class="callout-title">Exercise 18</h3>
<div class="callout-content">
<p>Now fit an F-distribution with 14 degrees of freedom using the
fitFDist function in the limma package:</p>
</div>
</div>
</div>
<div id="accordionSolution18" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution18" aria-expanded="false" aria-controls="collapseSolution18">
  <h4 class="accordion-header" id="headingSolution18"> Show me the solution </h4>
</button>
<div id="collapseSolution18" class="accordion-collapse collapse" data-bs-parent="#accordionSolution18" aria-labelledby="headingSolution18">
<div class="accordion-body">
<p><code>library(limma)</code><br><code>res &lt;- fitFDist(vars,14)</code><br><code>res</code></p>
</div>
</div>
</div>
</div>
<div id="exercise-19" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-19" class="callout-inner">
<h3 class="callout-title">Exercise 19</h3>
<div class="callout-content">
<p>Now create a qq-plot of the observed sample variances versus the
F-distribution quantiles. Which of the following best describes the
qq-plot? A) The fitted F-distribution provides a perfect fit.<br>
B) If we exclude the lowest 0.1% of the data, the F-distribution
provides a good fit.<br>
C) The normal distribution provided a better fit.<br>
D) If we exclude the highest 0.1% of the data, the F-distribution
provides a good fit.</p>
</div>
</div>
</div>
<div id="accordionSolution19" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution19" aria-expanded="false" aria-controls="collapseSolution19">
  <h4 class="accordion-header" id="headingSolution19"> Show me the solution </h4>
</button>
<div id="collapseSolution19" class="accordion-collapse collapse" data-bs-parent="#accordionSolution19" aria-labelledby="headingSolution19">
<div class="accordion-body">
<p><code>pf &lt;- (seq(along=vars)-0.5)/length(vars)</code><br><code>theory &lt;- qf(pf,14,res$df2) # theoretical quantiles from F distribution</code><br><code>mypar(1,2)</code><br><code>qqplot(theory, sort(vars), xlab = 'theory', ylab ='obs') # F approximation</code>
<code>vs variance from the data</code>
<code>qqnorm(sort(vars)) # normal approximation vs variance from the data</code><br>
The answer is D: If we exclude the highest 0.1% of the data, the
F-distribution provides a good fit. Actually I do not think answer is
entirely correct but this is the most appropriate one. Even if we
exclude the top 0.1% there are still more outliers to remove. Here is a
demonstration.</p>
<p><code>vars_sort &lt;- sort(vars)</code><br><code>vars_excl &lt;- vars_sort[1:18000]</code> # 18000 out of 22215
kept = 81%<br><code>pf_excl &lt;- (seq(along=vars_excl)-0.5)/length(vars_excl)</code><br><code>theory &lt;- qf(pf_excl,14,res$df2)</code></p>
<p><code>mypar(2,2)</code><br><code>qqplot(theory,vars_excl, xlab = 'theory',ylab='obs',</code><br><code>main = 'all data compared with F-approximation')</code><br><code>qqplot(vars_excl,theory, xlim = c(0,0.2),</code><br><code>ylim = c(0,100), main = 'zoomed in')</code><br><code>qqnorm(vars_excl, main = 'normal qqplot') # comparing with normal</code>
<code>approximation with filtered variance (81% kept)</code><br>
Even if I keep up to 81% of the values, F-distribution</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>Not every dataset follows a normal distribution. It is important to
be aware of other parametric distributions underlying the data
collected. This improves the technical precision of study results.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/principal-component-analysis.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/principal-component-analysis.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Principal Components
        </a>
        <a class="chapter-link float-end" href="../instructor/index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/smcclatchy/inference-high-dimensional-data/edit/main/episodes/statistical-models.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/smcclatchy/inference-high-dimensional-data/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/smcclatchy/inference-high-dimensional-data/" class="external-link">Source</a></p>
				<p><a href="https://github.com/smcclatchy/inference-high-dimensional-data/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:susan.mcclatchy@jax.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://smcclatchy.github.io/inference-high-dimensional-data/instructor/statistical-models.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries, inference, high-dimensional data, biomedical science",
  "name": "Statistical Models",
  "creativeWorkStatus": "active",
  "url": "https://smcclatchy.github.io/inference-high-dimensional-data/instructor/statistical-models.html",
  "identifier": "https://smcclatchy.github.io/inference-high-dimensional-data/instructor/statistical-models.html",
  "dateCreated": "2025-07-14",
  "dateModified": "2025-08-11",
  "datePublished": "2025-08-12"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

